---
title: "SCLC_DLL3_scRNA_cohortB"
output: html_document
date: "2025-04-03"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is document of data analyses of scRNA-seq of 10 SCLC patients collected at MGH.


## Step 1: setup

```{r}
# Load libraries
library(SingleCellExperiment)
library(Seurat)
library(tidyverse)
library(Matrix)
library(scales)
library(cowplot)
library(RCurl)
library(here)
library(SoupX)
library(DoubletFinder)
library(openxlsx)
library(celldex)
library(SingleR)

library(ggrepel)
library(openxlsx)

library(msigdbr)
library(clusterProfiler)
library(ggplot2)
library(stringr)
library(infercnv)

library(GeneNMF)
library(fgsea)
library(UCell)

library(ggpubr)
library(data.table)

library(SeuratExtend)
library(webr)
library(ggvenn)
library(eulerr)
library(gridExtra)


library(corrplot)
#library(ggVennDiagram)
```


## Setp 2: preprocessing

```{r}
# Step 2.1: Remove ambient RNA by SoupX

inDir <- "../../data"

all_files <- list.files(inDir, pattern= "_raw_feature_bc_matrix.h5",full.names = FALSE)

samples <- gsub("_raw_feature_bc_matrix.h5","",all_files)


data.10x = list()
for (sample in samples){
  filt.matrix <- Read10X_h5(paste0(inDir,"/",sample,"_filtered_feature_bc_matrix.h5"), use.names = T)
  raw.matrix <- Read10X_h5(paste0(inDir,"/",sample,"_raw_feature_bc_matrix.h5"), use.names = T)
  ## issue for cell ranger multi generated singleplex data from the fixed samples.
  filt_genes <- rownames(filt.matrix)

  # Subset raw.matrix to keep only the genes in filt.matrix
  raw.matrix_subset <- raw.matrix[rownames(raw.matrix) %in% filt_genes, ]
  srat <- CreateSeuratObject(counts = filt.matrix)
  #soup.channel <- SoupChannel(raw.matrix, filt.matrix) raw.matrix_subset
  soup.channel <- SoupChannel(raw.matrix_subset, filt.matrix)
  srat <- SCTransform(srat, verbose = F)
  srat <- RunPCA(srat, verbose = F)
  srat <- RunUMAP(srat, dims = 1:30, verbose = F)
  srat <- FindNeighbors(srat, dims = 1:30, verbose = F)
  srat <- FindClusters(srat, verbose = T)
  meta <- srat@meta.data
  umap <- srat@reductions$umap@cell.embeddings
  soup.channel <- setClusters(soup.channel, setNames(meta$seurat_clusters, rownames(meta)))
  soup.channel <- setDR(soup.channel, umap)
  soup.channel <- autoEstCont(soup.channel)
  data.10x[[sample]] <- adjustCounts(soup.channel, roundToInt = T)
}

# Create Seurat object after SoupX
scrna.list = list()
for (sample in samples) {
    scrna.list[[sample]] = CreateSeuratObject(counts = data.10x[[sample]], min.cells=3, project=sample)
}

# Remove raw data to save memory
rm(data.10x)
                    

# add sample name
for(i in 1:length(samples)){
  sample=samples[i]
  scrna.list[[sample]]$sample <- sample
}

### Step 2.2: Add percent.mt cell level metadata
for (sample in samples) {
  scrna.list[[sample]][["percent.mito"]] <- PercentageFeatureSet(scrna.list[[sample]], pattern = "^MT:|MT-|mt:|mt-") 
}

# Save scrna.lsit
saveRDS(scrna.list, paste0("../../rdata/","scrna.list.raw.rds"))

# merge list of prefiltered Seurat
scrna.combined_prefilter <- merge(x = scrna.list[[1]], 
                       y = scrna.list[-1], 
                       add.cell.id = names(scrna.list))
# metadata variable
metadata_prefilter <- scrna.combined_prefilter@meta.data
# Save Prefilter
saveRDS(scrna.combined_prefilter, paste0("../../rdata/","scrna.combined_prefilter.seurat.rds"))


### Step 2.3: prefilter QC plot

# Visualize the number of cell counts per sample
metadata_prefilter %>%
  	ggplot(aes(x=sample, fill=sample)) +
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("Prefilter NCells ")

ggsave(paste0("../../plots/",Sys.Date(),"_Prefilter_NCell.pdf"))


# mito
metadata_prefilter %>% ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mito)) + 
geom_point() + 
scale_colour_gradient(low = "gray90", high = "black") +
stat_smooth(method=lm) +
scale_x_log10() +
scale_y_log10() + 
theme_classic() +
geom_vline(xintercept = 500) +
geom_hline(yintercept = 300) +
facet_wrap(~sample) + 
ggtitle("Prefilter Mitochondrial Percentage")

ggsave(paste0("../../plots/",Sys.Date(),"_Prefilter_Mito_Percentage.pdf"))



# mito simple

metadata_prefilter %>% 
  	ggplot(aes(color=sample, x=percent.mito, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 10) + 
    ggtitle("Prefilter mito percent Per Cell")

ggsave(paste0("../../plots/",Sys.Date(),"_Prefilter_Mito_simple.pdf"))

# UMI
metadata_prefilter %>% 
  	ggplot(aes(color=sample, x=nCount_RNA, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500) + 
    ggtitle("Prefilter UMIs Per Cell")

ggsave(paste0("../../plots/",Sys.Date(),"_Prefilter_UMI.pdf"))

# gene

metadata_prefilter %>% 
  	ggplot(aes(color=sample, x=nCount_RNA, fill=sample )) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300) + 
    ggtitle("Prefilter Genes Per Cell")


ggsave(paste0("../../plots/",Sys.Date(),"_Prefilter_NGene.pdf"))


### Step 2.4 Filter cells based on nCount and nFeature, percent of mito
gene_cutoff <- 500
umi_cutoff <-  1000
mito_cutoff <- 10
for (sample in samples){
  scrna.list[[sample]] <- subset(scrna.list[[sample]], subset = nCount_RNA > umi_cutoff & nFeature_RNA > gene_cutoff & percent.mito < mito_cutoff )
}

saveRDS(scrna.list, paste0("../../rdata/","scrna.list.before.remove.doublet.rds"))
scrna.list <- readRDS(paste0("../../rdata/","scrna.list.before.remove.doublet.rds"))

### Step 2.5 remove doublets

### Doublet Finder
# grab number of cells
n_cells = list()
for (sample in samples){
n_cells[[sample]] <- length(colnames(scrna.list[[sample]]))
}
multiplet_rate = list()
for (sample in samples){
    if (n_cells[[sample]] <= 500){
        multiplet_rate[[sample]] = 0.004
    }else if (n_cells[[sample]] > 500 & n_cells[[sample]] < 2000){
             multiplet_rate[[sample]] = 0.008 
    } else if (n_cells[[sample]] >= 2000 & n_cells[[sample]] < 3000){
      multiplet_rate[[sample]] = 0.016 
    } else if (n_cells[[sample]] >= 3000 & n_cells[[sample]] < 4000){
      multiplet_rate[[sample]] = 0.024
    } else if (n_cells[[sample]] >= 4000 & n_cells[[sample]] < 5000){
      multiplet_rate[[sample]] = 0.032
    } else if (n_cells[[sample]] >= 5000 & n_cells[[sample]] < 6000){
      multiplet_rate[[sample]] = 0.040
    } else if (n_cells[[sample]] >= 6000 & n_cells[[sample]] < 7000){
      multiplet_rate[[sample]] = 0.048
    } else if (n_cells[[sample]] >= 7000 & n_cells[[sample]] < 8000){
      multiplet_rate[[sample]] = 0.056
    } else if (n_cells[[sample]] >= 8000 & n_cells[[sample]] < 9000){
      multiplet_rate[[sample]] = 0.064
    } else if (n_cells[[sample]] >= 9000 & n_cells[[sample]] < 10000){
      multiplet_rate[[sample]] = 0.072
    }else{
     multiplet_rate[[sample]] = 0.080
    }
}
nExp = list()
for (sample in samples){
scrna.list[[sample]] <- scrna.list[[sample]] %>% NormalizeData()
scrna.list[[sample]] = FindVariableFeatures(scrna.list[[sample]], verbose = F)
scrna.list[[sample]] = ScaleData(scrna.list[[sample]],verbose = F)
scrna.list[[sample]] = RunPCA(scrna.list[[sample]], verbose = F)
# Find significant PCs
stdv <- scrna.list[[sample]][["pca"]]@stdev
sum.stdv <- sum(scrna.list[[sample]][["pca"]]@stdev)
percent.stdv <- (stdv / sum.stdv) * 100
cumulative <- cumsum(percent.stdv)
co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] - 
                       percent.stdv[2:length(percent.stdv)]) > 0.1), 
              decreasing = T)[1] + 1
min.pc <- min(co1, co2)

scrna.list[[sample]] = RunUMAP(scrna.list[[sample]], dims = 1:min.pc, verbose = F)

## pK Identification (no ground-truth) ---------------------------------------------------------------------------------------
sweep.res <- paramSweep(scrna.list[[sample]], PCs = 1:min.pc, sct = FALSE)
sweep.stats <- summarizeSweep(sweep.res, GT = FALSE)
bcmvn <- find.pK(sweep.stats)
# Optimal pK is the max of the bomodality coefficent (BCmvn) distribution
bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
optimal.pk <- bcmvn.max$pK
optimal.pk <- as.numeric(levels(optimal.pk))[optimal.pk]

nExp[[sample]] <- round(ncol(scrna.list[[sample]]) * multiplet_rate[[sample]])  # expected doublets
scrna.list[[sample]] <- suppressMessages(doubletFinder(scrna.list[[sample]], pN = 0.25, pK = optimal.pk, nExp = nExp[[sample]], PCs = 1:min.pc))
}


DF.name = list()
for (sample in samples){
# name of the DF prediction can change, so extract the correct column name.
DF.name[[sample]] = colnames(scrna.list[[sample]]@meta.data)[grepl("^DF.classification", colnames(scrna.list[[sample]]@meta.data))]
}
# Plot the Doublet Finder results
UMAP_plots <- list()
for (sample in samples){
    #UMAP_plots[[sample]] <-  cowplot::plot_grid(ncol = 2, DimPlot(scrna.list[[sample]], group.by = "sample") + NoAxes(),
    #DimPlot(scrna.list[[sample]], group.by = DF.name[[sample]]) + NoAxes())
    UMAP_plots[[sample]] <- DimPlot(scrna.list[[sample]], group.by = DF.name[[sample]])
}

for (sample in samples){
    pdf(paste0("../../plots/",sample,"_DoubletFinder_UMAP_Plot.pdf"))
    print(UMAP_plots[[sample]])
    dev.off()
}

VlnPlots = list()
for (sample in samples){
VlnPlots[[sample]] <- VlnPlot(scrna.list[[sample]], features = "nFeature_RNA", group.by = DF.name[[sample]], pt.size = 0.1)
}

for (sample in samples){
    pdf(paste0("../../plots/",sample,"_DoubletFinder_VlnPlot.pdf"))
    print(VlnPlots[[sample]])
    dev.off()
}

# Remove the Doublet Cells
cells.use = list()
for (sample in samples){
     cells.use[[sample]] <- colnames(scrna.list[[sample]])[which(scrna.list[[sample]][[]][DF.name[[sample]]] == "Singlet")]
     scrna.list[[sample]] <- subset(scrna.list[[sample]], cells = cells.use[[sample]])
}

## Step 2.6 cell cycle scoring

load("/Volumes/rama/labMembers/mao20/reference/cycle.rda")

# Score cells for cell cycle
for (sample in samples){
    scrna.list[[sample]] <- CellCycleScoring(scrna.list[[sample]], 
                                 g2m.features = g2m_genes, 
                                 s.features = s_genes)
}
# Save scrna.lsit
saveRDS(scrna.list, paste0("../../rdata/","scrna.list.after.filters.rds"))

# merge list of postfiltered Seurat
scrna.combined_postfilter <- merge(x = scrna.list[[1]], 
                       y = scrna.list[-1], 
                       add.cell.id = names(scrna.list))
# metadata variable

metadata_postfilter <- scrna.combined_postfilter@meta.data

# Save Postfilter
saveRDS(scrna.combined_postfilter, paste0("../../rdata/","scrna.combined_postfilter.seurat.rds"))

### Step 2.7: postfilter QC plot


# Visualize the number of cell counts per sample
metadata_postfilter %>%
  	ggplot(aes(x=sample, fill=sample)) +
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("Postfilter NCells ")

ggsave(paste0("../../plots/",Sys.Date(),"_Postfilter_NCell.pdf"))



## mito 

metadata_postfilter %>% ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mito)) + 
geom_point() + 
scale_colour_gradient(low = "gray90", high = "black") +
stat_smooth(method=lm) +
scale_x_log10() +
scale_y_log10() + 
theme_classic() +
geom_vline(xintercept = 1000) +
geom_hline(yintercept = 500) +
facet_wrap(~sample) + 
ggtitle("Postfilter Mitochondrial Percentage")

ggsave(paste0("../../plots/",Sys.Date(),"_Postfilter_Mito_Percentage.pdf"))

## mito simple
metadata_postfilter %>% 
  	ggplot(aes(color=sample, x=percent.mito, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 10) + 
    ggtitle("Postfilter mito percent Per Cell")

ggsave(paste0("../../plots/",Sys.Date(),"_Postfilter_mito_simple.pdf"))
## UMI

metadata_postfilter %>% 
  	ggplot(aes(color=sample, x=nCount_RNA, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 1000) + 
    ggtitle("Postfilter UMIs Transcripts Per Cell")

ggsave(paste0("../../plots/",Sys.Date(),"_Postfilter_UMI.pdf"))
## NGene

metadata_postfilter %>% 
  	ggplot(aes(color=sample, x=nFeature_RNA, fill=sample )) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 500) + 
    ggtitle("Postfilter Genes Per Cell")

ggsave(paste0("../../plots/",Sys.Date(),"_Postfilter_NGene.pdf"))


```

## Step 3: Normalization and regress out unwanted variation


```{r}

# Normalize the counts
seurat_phase <- NormalizeData(scrna.combined_postfilter)

# Evaluating effects of cell cycle

# Identify the most variable genes
seurat_phase <- FindVariableFeatures(seurat_phase, 
                     selection.method = "vst",
                     nfeatures = 2000, 
                     verbose = FALSE)
		     
# Scale the counts
seurat_phase <- ScaleData(seurat_phase)

# Perform PCA
seurat_phase <- RunPCA(seurat_phase)

# Plot the PCA colored by cell cycle phase
DimPlot(seurat_phase,
        reduction = "pca",
        group.by= "Phase",
        split.by = "Phase")

ggsave("../../plots/cell_cycle_PCA.pdf")
# need to regress out cell cycle phase

## Evaluating effects of mitochondrial expression

# Check quartile values
summary(seurat_phase@meta.data$percent.mito)

# Check quartile values
mitoqtl <- summary(seurat_phase@meta.data$percent.mito)

# Turn mitoRatio into categorical factor vector based on quartile values
seurat_phase@meta.data$mitoFr <- cut(seurat_phase@meta.data$percent.mito, 
                   breaks=c(-Inf, as.numeric(mitoqtl[2]), as.numeric(mitoqtl[3]), as.numeric(mitoqtl[5]), Inf), 
                   labels=c("Low","Medium","Medium high", "High"))
# Plot the PCA colored by cell cycle phase
DimPlot(seurat_phase,
        reduction = "pca",
        group.by= "mitoFr",
        split.by = "mitoFr")

ggsave("../../plots/mitoFr_PCA.pdf")
				
## need to regress out mito but not cell cycle

rm(seurat_phase)
```

## Step 4: clustering


```{r}

## tried harmony integration but overcorrected the intertumoral heterogeneity, and other cell types from different samples have already clustered together without integration. thus no integration was needed to keep biological variation.
scrna.unintegrated <- readRDS(paste0("../../rdata/","scrna.combined_postfilter.seurat.rds"))

scrna.unintegrated <- NormalizeData(scrna.unintegrated)
scrna.unintegrated <- FindVariableFeatures(scrna.unintegrated)
scrna.unintegrated <- ScaleData(scrna.unintegrated,vars.to.regress = c("percent.mito"))
scrna.unintegrated <- RunPCA(scrna.unintegrated, npcs = 30, verbose = F)

scrna.unintegrated <- FindNeighbors(scrna.unintegrated, reduction = "pca", dims = 1:30)
scrna.unintegrated <- FindClusters(scrna.unintegrated, resolution = 0.2)

scrna.unintegrated <- RunUMAP(scrna.unintegrated, reduction = "pca",dims = 1:30, reduction.name = "umap")


DimPlot(
  scrna.unintegrated,
  reduction = "umap",
  group.by = c("sample"),
  combine = FALSE, label.size = 2
)

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_by_sample.pdf"))

DimPlot(
  scrna.unintegrated,
  reduction = "umap",
  label.size = 6,
  label=TRUE
) + theme(legend.position = "none")


ggsave(here(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_by_clusters.pdf")))


## epithelial

FeaturePlot(scrna.unintegrated, features = c("EPCAM","KRT8","KRT18","KRT19"),label=TRUE)
ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_epithelial_markers_cluster.pdf"))

## SCLC markers
FeaturePlot(scrna.unintegrated, features = c("ASCL1","NEUROD1","POU2F3","YAP1"))

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_SCLC_markers_cluster.pdf"))

# Save integrated object
saveRDS(scrna.unintegrated, paste0("../../rdata/","scrna.unintegrated.seurat.rds"))

##NEXT: run NMF for classification

```
## Step 5: NMF for annotation

```{r}
## run NMF

scrna.list <- readRDS(paste0("../../rdata/","scrna.list.after.filters.rds"))

geneNMF.programs <- multiNMF(scrna.list, assay="RNA", k=4:9, min.exp = 0.05)

## combine the gene programs identified over multiple samples and numbers of k into metaprograms

geneNMF.metaprograms <- getMetaPrograms(geneNMF.programs,
                                        metric = "cosine",
                                        weight.explained = 0.5,
                                        nMP=10)
## visualize pairwise similarity (in terms of cosine similarity or Jaccard index) between individual gene programs that compose meta-programs. 
pdf(paste0("../../plots/",Sys.Date(),"_NMF_pheatmap_MPs.pdf"))
plotMetaPrograms(geneNMF.metaprograms,
                       similarity.cutoff = c(0.1,1))
dev.off()


# inspect useful statistics, such as the “sample coverage” (in what fraction of samples the MP was detected); or the silhouette coefficient (how similar are individual programs in a MP relative to programs in other MPs - the higher the better).

geneNMF.metaprograms$metaprograms.metrics


##Interpretation of gene programs

lapply(geneNMF.metaprograms$metaprograms.genes, head)

geneNMF.metaprograms$metaprograms.genes.weights$MP6


## run GSEA to understand MP programs

top_p <- lapply(geneNMF.metaprograms$metaprograms.genes, function(program) {
  runGSEA(program, universe=rownames(scrna.unintegrated), category = "H")
})


head(top_p$MP1)


## Signature scores for gene programs

mp.genes <- geneNMF.metaprograms$metaprograms.genes
scrna.unintegrated <- AddModuleScore_UCell(scrna.unintegrated, features = mp.genes, ncores=4, name = "")



VlnPlot(scrna.unintegrated, features=names(mp.genes), pt.size = 0, ncol=5)

ggsave((paste0("../../plots/",Sys.Date(),"_NMF_MPs_score_by_cluster.pdf")))


# Rename all identities
scrna.unintegrated <- RenameIdents(object = scrna.unintegrated, 
                                  "0"="T",
                                  "1"="epithelial",
                                  "2"="epithelial",
                                  "3"="epithelial",
                                  "4"="epithelial",
                                  "5"="epithelial",
                                  "6"="B",
                                  "7"="epithelial",
                                  "8"="myeloid1",
                                  "9"="epithelial",
                                  "10"="fibroblast",
                                  "11"="epithelial",
                                  "12"="epithelial",
                                  "13"="NK",
                                  "14"="endothelial",
                                  "15"="B",
                                  "16"="epithelial",
                                  "17"="epithelial",
                                  "18"="myeloid2",
                                  "19"="epithelial")

scrna.unintegrated$final_anno <- scrna.unintegrated@active.ident

DimPlot(
  scrna.unintegrated,
  reduction = "umap",
  group.by = "final_anno",
  label.size = 4,
  label=TRUE) + theme(legend.position = "none")

ggsave(paste0("../../plots/",Sys.Date(),"_unintegrated_UMAP_by_final_anno.pdf"))

# Save integrated object
saveRDS(scrna.unintegrated, paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))


## plot heatmap with marker genes

scrna.unintegrated <-  readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))

## change factor level of final annotation

scrna.unintegrated$final_anno <- factor(scrna.unintegrated$final_anno, levels = c("epithelial", "endothelial","fibroblast","NK","T","B","myeloid1","myeloid2"))
### get marker genes

topgenes <- c("ASCL1","KRT8","KRT18","KRT19","EPCAM","PECAM1","FN1","CALD1","NKG7","GNLY","GZMA","GZMB","CD3E","CD3D","CD8A","CD8B","CD74","HLA-DRA","HLA-DRB1","CD68","LYZ","S100A8","S100A9")

DoHeatmap(scrna.unintegrated, features=topgenes,group.by="final_anno",size=3.5,angle=60,hjust=0.3,vjust=0)+ theme(axis.text.y = element_text(size = 8)) + NoLegend()

ggsave(paste0("../../plots/",Sys.Date(),"_heatmap_by_top5_marker_genes_by_cluster.pdf"))

# Save integrated object
saveRDS(scrna.unintegrated, paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))

```

## Step 6: infercnv

```{r}
## use fibroblast and endothelial as normal (C10 and C14); check epi malignancy(C1,2,3,4,5,7,9,11,12,16,17,19)

scrna.unintegrated <-  readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))

dir.create("../../inferCNV/inputs",,recursive =TRUE)

## download gene order file from 

##prepare input files

## input 1: gene order file

url <- "https://data.broadinstitute.org/Trinity/CTAT/cnv/hg38_gencode_v27.txt"

destfile <- "../../inferCNV/inputs/hg38_gencode_v27.txt"

download.file(url, destfile)

##Input 2: count matrix


## subset

scrna.unintegrated <- JoinLayers(scrna.unintegrated)


### use T, B, endothelial and fibroblast as control


epi_sub <- subset(scrna.unintegrated,subset= final_anno %in% c("T","B","endothelial","fibroblast","epithelial"))


counts <- epi_sub@assays$RNA@layers$counts

rownames(counts) <- rownames(epi_sub)
colnames(counts) <- colnames(epi_sub)

## input 3: sample annotation file

annotation <- as.data.frame(epi_sub@meta.data[,c("seurat_clusters","final_anno")])
annotation$cell_type <- as.character(annotation$final_anno)

idx <- which(annotation$final_anno %in% c("epithelial"))
annotation$cell_type[idx] <- paste0(as.character(annotation$final_anno[idx]),"_",as.character(annotation$seurat_clusters[idx]))


annotation <- annotation[3]

write.table(annotation, file="../../inferCNV/inputs/singleCell.annotation.4normal.txt",quote=FALSE,col.names = FALSE,row.names=TRUE,sep="\t")

 
# create the infercnv object
infercnv_obj = CreateInfercnvObject(raw_counts_matrix=counts,
                                    annotations_file="../../inferCNV/inputs/singleCell.annotation.4normal.txt",
                                    delim="\t",
                                    gene_order_file="../../inferCNV/inputs/hg38_gencode_v27.txt",
                                    chr_exclude = c("chrX","chrY","chrM"),
                                    ref_group_names=c("T","B","endothelial","fibroblast"))

options(scipen = 100)
# perform infercnv operations to reveal cnv signal
infercnv_obj = infercnv::run(infercnv_obj,
                             cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
                             out_dir="../../inferCNV/output_rv",  # dir is auto-created for storing outputs
                             cluster_by_groups=T,   # cluster
                             analysis_mode= "samples",
                             denoise=T,
                             HMM=T,
                             num_threads=4,
                             output_format = "pdf"
                             )


## quantify CNV scores
infercnv_obj <- readRDS("../../inferCNV/output_rv/run.final.infercnv_obj")

### centered at 0
cnv <- infercnv_obj@expr.data -1 

cnv_score <- colSums(abs(cnv))

## get index of subset in whole object

idx <- match(colnames(cnv),colnames(scrna.unintegrated))

scrna.unintegrated$totCNV <- 0

scrna.unintegrated$totCNV[idx] <- cnv_score


FeaturePlot(scrna.unintegrated,features="totCNV",col=c("lightgrey","red"),label=TRUE)

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_totCNV_allin.pdf")))
VlnPlot(scrna.unintegrated,features="totCNV",group.by="final_anno",pt.size=0)+ NoLegend()

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_Violin_totCNV_allin.pdf")))

scrna.unintegrated$combine_anno <- paste0(scrna.unintegrated$final_anno,"_",scrna.unintegrated$seurat_clusters)
VlnPlot(scrna.unintegrated,features="totCNV",group.by="combine_anno",pt.size=0)+ NoLegend()

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_Violin_totCNV_allin_combined_anno.pdf")))


cnv_sd <- apply(abs(cnv),2,sd)

scrna.unintegrated$sdCNV <- 0

scrna.unintegrated$sdCNV[idx] <- cnv_sd
FeaturePlot(scrna.unintegrated,features="sdCNV",col=c("lightgrey","red"),label=TRUE)

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_sdCNV_allin.pdf")))
VlnPlot(scrna.unintegrated,features="sdCNV",group.by="final_anno",pt.size=0) + NoLegend()

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_Violin_sdCNV_allin.pdf")))

VlnPlot(scrna.unintegrated,features="sdCNV",group.by="combine_anno",pt.size=0) + NoLegend()

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_Violin_sdCNV_allin_combined_anno.pdf")))

### epithelial cluster 17 are likely to be normal
```

## Step 7: additional checking on epithelials 

```{r}
scrna.unintegrated <-  readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))

epi_sub <- subset(scrna.unintegrated,subset= final_anno=="epithelial")


epi_sub <- NormalizeData(epi_sub)
epi_sub <- FindVariableFeatures(epi_sub)
epi_sub <- ScaleData(epi_sub,vars.to.regress = c("percent.mito"))
epi_sub <- RunPCA(epi_sub, npcs = 30, verbose = F)

epi_sub <- FindNeighbors(epi_sub, reduction = "pca", dims = 1:30)
epi_sub <- FindClusters(epi_sub, resolution = 0.1)

epi_sub <- RunUMAP(epi_sub, reduction = "pca",dims = 1:30, reduction.name = "umap")


DimPlot(
  epi_sub,
  reduction = "umap",
  group.by = c("sample"),
  combine = FALSE, label.size = 2
)

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_by_sample_epi.pdf"))

DimPlot(
  epi_sub,
  reduction = "umap",
  label.size = 6,
  label=TRUE
) + theme(legend.position = "none")


ggsave(here(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_by_clusters_epi.pdf")))


### perform integration to see if can separate out normal epithelial

epi.integrated <- IntegrateLayers(
  object = epi_sub, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "integrated.harmony",
  verbose = FALSE
)

epi.integrated <- FindNeighbors(epi.integrated, reduction = "integrated.harmony", dims = 1:30)
epi.integrated <- FindClusters(epi.integrated, resolution = 0.1, cluster.name = "harmony_clusters")

epi.integrated <- RunUMAP(epi.integrated, reduction = "integrated.harmony",dims = 1:30, reduction.name = "umap.harmony")



DimPlot(
  epi.integrated,
  reduction = "umap.harmony",
  group.by = c("sample"),
  combine = FALSE, label.size = 2
)

ggsave(paste0("../../plots/",Sys.Date(),"_integration_UMAP_by_sample_epi.pdf"))

DimPlot(
  epi.integrated,
  reduction = "umap.harmony",
  label.size = 6,
  label=TRUE
) + theme(legend.position = "none")


ggsave(here(paste0("../../plots/",Sys.Date(),"_integration_UMAP_by_clusters_epi.pdf")))

## definitely no integration needed for tumors

####NMF

epi_sub <- JoinLayers(epi_sub)

seu.list <- SplitObject(epi_sub, split.by = "sample")


geneNMF.programs <- multiNMF(seu.list, assay="RNA", k=4:9, min.exp = 0.05)

## combine the gene programs identified over multiple samples and numbers of k into metaprograms

geneNMF.metaprograms <- getMetaPrograms(geneNMF.programs,
                                        metric = "cosine",
                                        weight.explained = 0.5,
                                        nMP=10)

## visualize pairwise similarity (in terms of cosine similarity or Jaccard index) between individual gene programs that compose meta-programs. 
pdf(paste0("../../plots/",Sys.Date(),"_NMF_pheatmap_MPs_epi.pdf"))
plotMetaPrograms(geneNMF.metaprograms,
                       similarity.cutoff = c(0.1,1))
dev.off()


# inspect useful statistics, such as the “sample coverage” (in what fraction of samples the MP was detected); or the silhouette coefficient (how similar are individual programs in a MP relative to programs in other MPs - the higher the better).

geneNMF.metaprograms$metaprograms.metrics


##Interpretation of gene programs

lapply(geneNMF.metaprograms$metaprograms.genes, head)

geneNMF.metaprograms$metaprograms.genes.weights$MP6


## run GSEA to understand MP programs

top_p <- lapply(geneNMF.metaprograms$metaprograms.genes, function(program) {
  runGSEA(program, universe=rownames(epi_sub), category = "H")
})

head(top_p$MP1)


## Signature scores for gene programs

mp.genes <- geneNMF.metaprograms$metaprograms.genes
epi_sub <- AddModuleScore_UCell(epi_sub, features = mp.genes, ncores=4, name = "")



VlnPlot(epi_sub, features=names(mp.genes), pt.size = 0, ncol=5)

ggsave((paste0("../../plots/",Sys.Date(),"_NMF_MPs_score_by_cluster_epi.pdf")))

## add known marker for SCLC subtype


gene_sets <- list(
  pASCL1 = c("ASCL1", "SOX4", "STMN2", "DOC2A"),
  pNEUROD1 = c("NEUROD1", "ADCYAP1","NRXN1", "SSTR2", "ID1", "ID3", "SST", "DLK1"),
  pPOU2F3 = c("POU2F3", "ASCL2", "CD44", "MYC", "KIT", "YBX1"),
  pYAP1=c("YAP1"),
  pInflamed= c("CD274","PDCD1","CD80","CD86","CTLA4","CD38","IDO1","TIGIT","ICOS","LAG3","CCL5","CXCL10",
              "HLA-A","HLA-B","HLA-C","HLA-E","HLA-F","HLA-G","HLA-DRA","HLA-DRB1","HLA-DQA1","HLA-DMA","HLA-DPA1",
              "HLA-DMB","HLA-DPB1","HLA-DQB1","HLA-DOA","HLA-DOB")
)

epi_sub <- AddModuleScore_UCell(epi_sub, features = gene_sets, ncores=4, name = "")

VlnPlot(epi_sub, features=names(gene_sets), pt.size = 0, ncol=2)

ggsave((paste0("../../plots/",Sys.Date(),"_SCLC_MP_score_epi.pdf")))


FeaturePlot(epi_sub,features=c("ASCL1","NEUROD1","POU2F3","YAP1","HLA-A","HLA-B","HLA-DRA","HLA-DRB1","CTLA4"))

ggsave((paste0("../../plots/",Sys.Date(),"_SCLC_marker_UMAP_epi.pdf")))


### classify individual cells for max

# Get the score matrix (remove the "1" suffix added by AddModuleScore)
score_matrix <- FetchData(epi_sub, vars = names(gene_sets))
colnames(score_matrix) <- c("SCLC-A","SCLC-N","SCLC-P","SCLC-Y","SCLC-I")

# Assign cell type based on highest score
epi_sub$SCLC_subtype <- colnames(score_matrix)[max.col(score_matrix)]

# Visualize results
DimPlot(epi_sub, group.by = "SCLC_subtype")

ggsave(paste0("../../plots/",Sys.Date(),"_UMAP_SCLC_sybtype_per_cell_using_addmodulescore.pdf"))


###plot stackedbar for subtype per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SCLC_subtype))
colnames(samType) <- c("Sample","cellType","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(cellType == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = cellType)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_SCLC_per_sample.pdf")))
```

## Step 8: DLL3+ vs DLL3- cells 

```{r}
## calculate DLL3+ fraction per sample

## DLL3 pos vs neg
epi_sub$DLL3_group <- "neg"
 
epi_sub$DLL3_group[which(epi_sub[['RNA']]$data["DLL3",] > 0)] <- "pos" 
epi_sub$DLL3_group <- factor(epi_sub$DLL3_group)

## show UMAP
UMAPPlot(epi_sub, group.by="DLL3_group",cols=c("blue","red")) 
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_DLL3_pos_neg.pdf")))

### plot fraction of DLL3+ DLL3-

###plot stackedbar for subtype per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample,FUN = sum) * 100


## sort sample

reordered_data <- samType %>%
  filter(DLL3 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(text = element_text(size = 8),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_by_sample.pdf")))

## add treatment

epi_sub$treatment <- "naive"

epi_sub$treatment[which(epi_sub$sample %in% c("LP21","LP28","LP33","LP85"))] <- "treated"

## add stage

epi_sub$stage <- "metastatic"

epi_sub$stage[which(epi_sub$sample %in% c("LP19","LP21","LP91"))] <- "primary"

# Save integrated object
saveRDS(epi_sub, paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))


## separate by stage and treatment

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## DLL3+ vs DLL3- by treatment

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$treatment))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by Treatment", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_by_treatment.pdf")))

### boxplot

df <- reordered_data %>%
   filter(DLL3 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%DLL3+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_per_sample_by_treatment.pdf")))


### violin plot

VlnPlot(epi_sub,features="DLL3",group.by="treatment",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_treatment.pdf")))


## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=DLL3_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "DLL3",
  cells=cells,
  group.by = "treatment",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_treatment_withBox.pdf")))


### violin plot with enhanced vlnplots for subtype

cells <- WhichCells(epi_sub, expression=(DLL3_group=="pos") & (SCLC_subtype %in% c("SCLC-A","SCLC-N","SCLC-P")))
  
VlnPlot2(
  epi_sub,
  features = "DLL3",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("#F8766D", "#00BF7D", "#00B0F6"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox.pdf")))


## DLL3+ vs DLL3- by stage

## DLL3+ vs DLL3- by treatment

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$stage))


colnames(samTreat) <- c("Sample","Stage","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Stage,desc(Value)) %>%
   select(Stage,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Stage","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by Stage", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_by_stage.pdf")))

### boxplot

df <- reordered_data %>%
   filter(DLL3 == "pos") %>%
   select(Stage, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Stage", y = "Value.y",
          color = "Stage", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%DLL3+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_per_sample_by_stage.pdf")))


##violin plot for DLL3 expression

VlnPlot(epi_sub,features="DLL3",group.by="stage",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_stage.pdf")))

```

## Step 9:  CK genes and NOTCH pathway genes for dropout estimation

```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## CKs
## by treatment
ck8t <- VlnPlot(epi_sub,features="KRT8",group.by="treatment",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck18t <- VlnPlot(epi_sub,features="KRT18",group.by="treatment",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck19t <- VlnPlot(epi_sub,features="KRT19",group.by="treatment",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ggarrange(ck8t,ck18t,ck19t,nrow=3,ncol=1,font.label = list(size = 5))

ggsave(here(paste0("../../plots/",Sys.Date(),"_CK_violinplot_by_DLL3_by_treatment.pdf")))


## by stage


ck8t <- VlnPlot(epi_sub,features="KRT8",group.by="stage",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck18t <- VlnPlot(epi_sub,features="KRT18",group.by="stage",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck19t <- VlnPlot(epi_sub,features="KRT19",group.by="stage",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ggarrange(ck8t,ck18t,ck19t,nrow=3,ncol=1,font.label = list(size = 5))

ggsave(here(paste0("../../plots/",Sys.Date(),"_CK_violinplot_by_DLL3_by_stage.pdf")))


## check cell cycle

### stage
## S.Score
VlnPlot(epi_sub,features="S.Score",group.by="stage",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ggsave(here(paste0("../../plots/",Sys.Date(),"_CellCycle_S_violinplot_by_DLL3_by_stage.pdf")))

## G2M.Score

VlnPlot(epi_sub,features="G2M.Score",group.by="stage",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ggsave(here(paste0("../../plots/",Sys.Date(),"_CellCycle_G2M_violinplot_by_DLL3_by_stage.pdf")))

### treatment

## S.Score
VlnPlot(epi_sub,features="S.Score",group.by="treatment",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ggsave(here(paste0("../../plots/",Sys.Date(),"_CellCycle_S_violinplot_by_DLL3_by_treatment.pdf")))

## G2M.Score

VlnPlot(epi_sub,features="G2M.Score",group.by="treatment",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ggsave(here(paste0("../../plots/",Sys.Date(),"_CellCycle_G2M_violinplot_by_DLL3_by_treatment.pdf")))



## NOTCH pathway as baits

## reported co-expressed gene in Notch pathway

cogenes <- c("NOTCH1", "NOTCH2", "NOTCH3", "NOTCH4", "HES1", "HEY1", "JAG1","JAG2")

## count noon_zeros in this gene list

# Extract the count matrix (raw counts)
count_matrix <- GetAssayData(epi_sub, layer = "counts")

# Subset the count matrix to include only the genes in your list
gene_counts <- count_matrix[rownames(count_matrix) %in% cogenes, ]

# Count how many cells have more than zero counts
cells_with_nonzero_counts <- colSums(gene_counts > 0)

pdf(here(paste0("../../plots/",Sys.Date(),"_hist_DLL3_coexpressed_gene_per_cell.pdf")))
hist(cells_with_nonzero_counts, breaks = seq(-0.5, 8.5, by = 1), main="Counts of DLL3 co-expressed genes per cell")

dev.off()

##create a group for DLL3

epi_sub$DLL3_coNum <- cells_with_nonzero_counts

for (nco in 0:8){
epi_sub$DLL3_co <- as.character(epi_sub$DLL3_group)
 
epi_sub$DLL3_co[which(epi_sub$DLL3_group=="neg" & epi_sub$DLL3_coNum <= nco)] <- "true_neg" 

epi_sub$DLL3_co[which(epi_sub$DLL3_group=="neg" & epi_sub$DLL3_coNum >nco)] <- "false_neg" 

epi_sub$DLL3_co <- factor(epi_sub$DLL3_co, levels = c("true_neg","false_neg","pos"))


## plot

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_co))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(DLL3 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = paste0("DLL3 expression per sample (false_neg: >= ",nco+1," co-exp genes)"), y = "Proportion", x = "")


ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_using_",nco,"_coexp_by_subtype.pdf")))

}


```

## Step 10: run NMF on tumors check any differences between DLL3+ and DLL3- populations

```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

seu.list <- SplitObject(epi_sub, split.by = "sample")


geneNMF.programs <- multiNMF(seu.list, assay="RNA", k=4:9, min.exp = 0.05)

## combine the gene programs identified over multiple samples and numbers of k into metaprograms

geneNMF.metaprograms <- getMetaPrograms(geneNMF.programs,
                                        metric = "cosine",
                                        weight.explained = 0.5,
                                        nMP=10)

## visualize pairwise similarity (in terms of cosine similarity or Jaccard index) between individual gene programs that compose meta-programs. 
pdf(paste0("../../plots/",Sys.Date(),"_NMF_pheatmap_MPs_epi.pdf"))
plotMetaPrograms(geneNMF.metaprograms,
                       similarity.cutoff = c(0.1,1))
dev.off()


# inspect useful statistics, such as the “sample coverage” (in what fraction of samples the MP was detected); or the silhouette coefficient (how similar are individual programs in a MP relative to programs in other MPs - the higher the better).

geneNMF.metaprograms$metaprograms.metrics


##Interpretation of gene programs

lapply(geneNMF.metaprograms$metaprograms.genes, head)

geneNMF.metaprograms$metaprograms.genes.weights$MP6




## run GSEA to understand MP programs

top_p <- lapply(geneNMF.metaprograms$metaprograms.genes, function(program) {
  runGSEA(program, universe=rownames(epi_sub), category = "H")
})


head(top_p$MP1)


## Signature scores for gene programs

mp.genes <- geneNMF.metaprograms$metaprograms.genes
epi_sub <- AddModuleScore_UCell(epi_sub, features = mp.genes, ncores=4, name = "")



VlnPlot(epi_sub, features=names(mp.genes), cols=c("blue","red"),group.by = "DLL3_group", pt.size = 0, ncol=5)

ggsave((paste0("../../plots/",Sys.Date(),"_NMF_MPs_score_by_DLL3_epi.pdf")))


### let's plot runGSEA results


library(data.table)

## MP3

result <- as.data.frame(top_p$MP3)
ggplot(result, aes(x = reorder(pathway, foldEnrichment), y = foldEnrichment, fill = padj)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  scale_fill_gradient(low = "red", high = "blue") +
  labs(title = "GSEA Results for MP3",
       x = "Pathway",
       y = "FoldEnrichment") +
  theme_minimal()


ggsave((paste0("../../plots/",Sys.Date(),"_NMF_GSEA_MP3.pdf")))

for (num in 1:10) {
  result <- as.data.frame(top_p[[num]])
ggplot(result, aes(x = reorder(pathway, foldEnrichment), y = foldEnrichment, fill = padj)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  scale_fill_gradient(low = "red", high = "blue") +
  labs(title = paste0("GSEA Results for MP",num),
       x = "Pathway",
       y = "FoldEnrichment") +
  theme_minimal()


ggsave((paste0("../../plots/",Sys.Date(),"_NMF_GSEA_MP",num,".pdf")))
  
}

```

## Step 11: scDEG

```{r}

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))
Idents(epi_sub) <- 'DLL3_group'
dll3.deg.markers <- FindMarkers(epi_sub, ident.1 = "pos", ident.2 = "neg")
head(dll3.deg.markers,n=10)

#library(ggrepel)
dll3.deg.markers$Gene <- rownames(dll3.deg.markers)
# The significantly differentially expressed genes are the ones found in the upper-left and upper-right corners.
# Add a column to the data frame to specify if they are UP- or DOWN- regulated (log2FoldChange respectively positive or negative)
# add a column of NAs
dll3.deg.markers$diffexpressed <- "NO"
fcthresh <- 2
pthresh <- 0.05
# if avg_log2FC > FCthreshold and p_val_adj < pthreshold, set as "UP"
dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC > log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "UP"
# if avg_log2FC < -FCthreshold and p_val_adj < pthreshold, set as "DOWN"
dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC < -log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "DOWN"
# Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
dll3.deg.markers$delabel <- NA
dll3.deg.markers$delabel[dll3.deg.markers$diffexpressed != "NO"] <- dll3.deg.markers$Gene[dll3.deg.markers$diffexpressed != "NO"]

pdf(here(paste0("../../plots/",Sys.Date(),"_volcano_DLL3_pos_vs_neg_fc",fcthresh,"_p_",pthresh,".pdf")))
# plot adding up all layers we have seen so far
p <-
  ggplot(data=dll3.deg.markers, aes(x=avg_log2FC, y=-log10(p_val_adj), col=diffexpressed, label=delabel)) +
  geom_point() +
  theme_minimal() +
  geom_text_repel(size=2,max.overlaps = 40) +
  scale_color_manual(values=c("blue", "black", "red")) +
  geom_vline(xintercept=c(-log2(fcthresh), log2(fcthresh)), col="red") +
  geom_hline(yintercept=-log10(pthresh), col="red") +
  ggtitle("DLL3 pos vs neg") + xlab("log2FoldChange")
print(p)
dev.off()

#library(msigdbr)
#library(clusterProfiler)
#library(stringr)
# only HALLMARK set
TERM2GENE <- msigdbr(species = "Homo sapiens", category = "H")%>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()
emsigdbup <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="UP")], TERM2GENE = TERM2GENE)
dotplot(emsigdbup, showCategory=25, font.size = 7, title=paste0("MSigDB (UP)"))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))

ggsave(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_pos_vs_neg_fc",fcthresh,"_p",pthresh,".pdf")))

## no upregulated genes

emsigdbdn <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="DOWN")], TERM2GENE = TERM2GENE)
dotplot(emsigdbdn, showCategory=25, font.size = 7, title=paste0("MSigDB (DOWN)"))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))
ggsave(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_pos_vs_neg_DOWN_fc",fcthresh,"_p",pthresh,".pdf")))

```

## step 12: additional target gene checking

```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## by sample

DimPlot(
  epi_sub,
  reduction = "umap",
  group.by = c("sample"),
  combine = FALSE, label.size = 2
)

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_by_sample_epi.pdf"))



## SEZ6 and B7H3(CD276)

## SCLC markers
FeaturePlot(epi_sub, features = c("SEZ6"))+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_SEZ6.pdf"))

FeaturePlot(epi_sub, features = c("CD276"))+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_CD276.pdf"))

FeaturePlot(epi_sub, features = c("DLL3"))+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_DLL3.pdf"))


### correlation

FeatureScatter(epi_sub, feature1 = "DLL3", feature2 = "SEZ6",group.by = "sample")+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_scatter_SEZ6_vs_DLL3.pdf"))


FeatureScatter(epi_sub, feature1 = "DLL3", feature2 = "CD276",group.by = "sample")+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_scatter_CD276_vs_DLL3.pdf"))


FeatureScatter(epi_sub, feature1 = "SEZ6", feature2 = "CD276",group.by = "sample")+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_scatter_CD276_vs_SEZ6.pdf"))


## multiple on the same plot

FeaturePlot3(epi_sub, feature.1 = "SEZ6", feature.2 = "CD276", feature.3 = "DLL3", pt.size = 1)

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_SEZ6_CD276_DLL3.pdf"))


### SEZ6+ vs SEZ6-


## SEZ6 pos vs neg
epi_sub$SEZ6_group <- "neg"
 
epi_sub$SEZ6_group[which(epi_sub[['RNA']]$data["SEZ6",] > 0)] <- "pos" 
epi_sub$SEZ6_group <- factor(epi_sub$SEZ6_group)

## show UMAP
UMAPPlot(epi_sub, group.by="SEZ6_group",cols=c("blue","red")) + coord_fixed()
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_SEZ6_pos_neg.pdf")))

### plot fraction of SEZ6+ SEZ6-

###plot stackedbar per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample,FUN = sum) * 100


## sort sample

reordered_data <- samType %>%
  filter(SEZ6 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(text = element_text(size = 8),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_by_sample.pdf")))



### violin plot with enhanced vlnplots for subtype

cells <- WhichCells(epi_sub, expression=(SEZ6_group=="pos") & (SCLC_subtype %in% c("SCLC-A","SCLC-N","SCLC-P")))
  
VlnPlot2(
  epi_sub,
  features = "SEZ6",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("#F8766D", "#00BF7D", "#00B0F6"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_SCLC_subtype_withBox.pdf")))



## SEZ6+ vs SEZ6- by treatment

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$treatment))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(SEZ6 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample by Treatment", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_per_sample_by_treatment.pdf")))

### boxplot

df <- reordered_data %>%
   filter(SEZ6 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%SEZ6+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_per_sample_by_treatment.pdf")))

## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "SEZ6",
  cells=cells,
  group.by = "treatment",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_treatment_withBox.pdf")))




## SEZ6+ vs SEZ6- by stage

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$stage))


colnames(samTreat) <- c("Sample","Stage","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(SEZ6 == "pos") %>%
   arrange(Stage,desc(Value)) %>%
   select(Stage,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Stage","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample by Stage", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_per_sample_by_stage.pdf")))

### boxplot

df <- reordered_data %>%
   filter(SEZ6 == "pos") %>%
   select(Stage, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Stage", y = "Value.y",
          color = "Stage", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%SEZ6+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_per_sample_by_stage.pdf")))



## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "SEZ6",
  cells=cells,
  group.by = "stage",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_stage_withBox.pdf")))


### CD276+ vs CD276-


## CD276 pos vs neg
epi_sub$CD276_group <- "neg"
 
epi_sub$CD276_group[which(epi_sub[['RNA']]$data["CD276",] > 0)] <- "pos" 
epi_sub$CD276_group <- factor(epi_sub$CD276_group)

## show UMAP
UMAPPlot(epi_sub, group.by="CD276_group",cols=c("blue","red")) + coord_fixed()
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_CD276_pos_neg.pdf")))

### plot fraction of CD276+ CD276-

###plot stackedbar per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample,FUN = sum) * 100


## sort sample

reordered_data <- samType %>%
  filter(CD276 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(text = element_text(size = 8),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "CD276 expression per sample", y = "Proportion", x = "Sample")+ coord_fixed()

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_by_sample.pdf")))



### violin plot with enhanced vlnplots for subtype

cells <- WhichCells(epi_sub, expression=(CD276_group=="pos") & (SCLC_subtype %in% c("SCLC-A","SCLC-N","SCLC-P")))
  
VlnPlot2(
  epi_sub,
  features = "CD276",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("#F8766D", "#00BF7D", "#00B0F6"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_SCLC_subtype_withBox.pdf")))



## CD276+ vs CD276- by treatment

samType <- as.data.frame(table(epi_sub$sample,epi_sub$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$treatment))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(CD276 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "CD276 expression per sample by Treatment", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_per_sample_by_treatment.pdf")))

### boxplot

df <- reordered_data %>%
   filter(CD276 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%CD276+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_CD276_per_sample_by_treatment.pdf")))

## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=CD276_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "CD276",
  cells=cells,
  group.by = "treatment",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_treatment_withBox.pdf")))




## CD276+ vs CD276- by stage

samType <- as.data.frame(table(epi_sub$sample,epi_sub$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$stage))


colnames(samTreat) <- c("Sample","Stage","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(CD276 == "pos") %>%
   arrange(Stage,desc(Value)) %>%
   select(Stage,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Stage","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "CD276 expression per sample by Stage", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_per_sample_by_stage.pdf")))

### boxplot

df <- reordered_data %>%
   filter(CD276 == "pos") %>%
   select(Stage, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Stage", y = "Value.y",
          color = "Stage", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%CD276+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_CD276_per_sample_by_stage.pdf")))



## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=CD276_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "CD276",
  cells=cells,
  group.by = "stage",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_stage_withBox.pdf")))

## save seurat object
saveRDS(epi_sub,paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))
```

## Step 13: additional tumor pathways checking
```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## pathways

# proliferation: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/CELL_PROLIFERATION_GO_0008283

#Stemness : https://www.gsea-msigdb.org/gsea/msigdb/cards/MALTA_CURATED_STEMNESS_MARKERS.html

#Quiescence: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GRAHAM_NORMAL_QUIESCENT_VS_NORMAL_DIVIDING_UP.html

#Metastasis promoting pathways: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/RAMASWAMY_METASTASIS_UP.html

#Tumor aggressiveness: did not find, use MYC and E2F instead

##E2F: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/HALLMARK_E2F_TARGETS.html

#Invasion : https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/WANG_TUMOR_INVASIVENESS_UP.html?ex=1

## p53 pathway:https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/HALLMARK_P53_PATHWAY.html 

### load pathway signatures

proliferation <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="proliferation",colNames = FALSE)
stemness <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="stemness",colNames = FALSE)
quiescence <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="quiescence",colNames = FALSE)
metastasis <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="metastasis",colNames = FALSE)
MYC <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="MYC",colNames = FALSE)
E2F <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="E2F",colNames = FALSE)
invasion <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="invasion",colNames = FALSE)
p53 <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="p53",colNames = FALSE)
EMT <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="EMT",colNames = FALSE)

gene_sets <- list(
  proliferation = proliferation,
  stemness=stemness,
  quiescence=quiescence,
  metastasis=metastasis,
  pMYC=MYC,
  pE2F=E2F,
  invasion=invasion,
  pP53=p53,
  EMT=EMT
)

## save tumor aggressiveness pathway with name of gene_sets
save(gene_sets, file="/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/tumor.aggressiveness.list.RData")

epi_sub <- AddModuleScore_UCell(epi_sub, features = gene_sets, ncores=4, name = "")

VlnPlot2(epi_sub, features=names(gene_sets), group.by = "DLL3_group", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("blue","red"))

ggsave((paste0("../../plots/",Sys.Date(),"_SCLC_SMpathway_score_epi.pdf")))
```

## Step 14: coexpression of DLL3, SEZ6 and CD276

```{r}
## check SEZ6 and CD276 expression between DLL3+ and DLL3- cells
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

VlnPlot2(
  epi_sub,
  features = "CD276",
  group.by = "DLL3_group",
  pt=F,
  box=F,
  width=0.5,
  show.mean = TRUE,
  cols=c("blue","red"),
  lab_fill = "DLL3",
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_DLL3_withBox.pdf")))


VlnPlot2(
  epi_sub,
  features = "SEZ6",
  group.by = "DLL3_group",
  pt=F,
  box=F,
  width=0.5,
  show.mean = TRUE,
  cols=c("blue","red"),
  lab_fill = "DLL3",
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_DLL3_withBox.pdf")))


### overlap among three genes' positive cells

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## only keep top 3 SCLC-subtypes

cells <- WhichCells(epi_sub, expression=SCLC_subtype %in% c("SCLC-A", "SCLC-N", "SCLC-P"))

epi_sub1 <- subset(epi_sub,cells=cells) 

## Q1: What fraction of all SCLC are DLL3+ vs DLL3-, and what are the numbers across the classical different histological subsets?

data <- as.data.frame(epi_sub1@meta.data[,c("SCLC_subtype","DLL3_group")])

df <- data %>% group_by(SCLC_subtype,DLL3_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(DLL3=DLL3_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_DLL3_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =DLL3 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,labelpositionThreshold=0.01,showRatioThreshold = 0.005,
		ratioByGroup = FALSE)

dev.off()


#Same question for SEZ6, plus how much overlap is there for individual cells expressing DLL3 and SEZ6?

data <- as.data.frame(epi_sub1@meta.data[,c("SCLC_subtype","SEZ6_group")])

df <- data %>% group_by(SCLC_subtype,SEZ6_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(SEZ6=SEZ6_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_SEZ6_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =SEZ6 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,labelpositionThreshold=0.01,showRatioThreshold = 0.003,
		ratioByGroup = FALSE)

dev.off()

#Same question for B7H3, plus how much overlap is there for individual cells expressing DLL3, SEZ6 or B7H3?

data <- as.data.frame(epi_sub1@meta.data[,c("SCLC_subtype","CD276_group")])

df <- data %>% group_by(SCLC_subtype,CD276_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(CD276=CD276_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_CD276_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =CD276 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,labelpositionThreshold=0.005,showRatioThreshold = 0.0015,
		ratioByGroup = FALSE)

dev.off()


## cells with DLL3

DLL3_pos <- WhichCells(epi_sub1, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub1, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub1, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

ggvenn(a)

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3.pdf")))


## venn diagram has the equal sizes and misleading

## try install.packages("eulerr") or upset plot
pdf((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler.pdf")))
plot(euler(a, shape = "ellipse"), quantities = TRUE)
dev.off()



epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

cells <- WhichCells(epi_sub, expression=SCLC_subtype %in% c("SCLC-A", "SCLC-N", "SCLC-P"))

epi_sub1 <- subset(epi_sub,cells=cells) 

### split by subtype
subtypes <- unique(epi_sub1$SCLC_subtype)
genes <- c("DLL3","SEZ6","CD276")
### loop through three subtypes
for (st in subtypes){
  cells <- WhichCells(epi_sub1,expression=SCLC_subtype==st)
  
  epi <- subset(epi_sub1,cells=cells)
  
  ## loop through three genes
  for (gene in genes){
    group <- paste0(gene,"_group")
    data <- as.data.frame(epi@meta.data[,c(group,"SCLC_subtype")])
    
    df <- data %>% group_by(across(all_of(group))) %>% summarise(count=n(),.groups="drop") 
    
    fill_col <- sym(group)
    
    ggplot(df, aes(x = "", y = count, fill = !!fill_col)) +
      geom_col(color = "black") +
      geom_text(aes(label = count),
                position = position_stack(vjust = 0.5)) +
      coord_polar(theta = "y") +
      scale_fill_manual(values = c("blue","red")) + 
      labs(fill=gene) + ggtitle(st) + 
      theme_void() + theme(plot.title = element_text(hjust = 0.5,vjust = -6))
    
    ggsave(here(paste0("../../plots/",Sys.Date(),"_pie_",gene,"_",st,".pdf")))
  }
  ## venn diagram
  
  ## cells with DLL3
  
  DLL3_pos <- WhichCells(epi, expression=DLL3_group=="pos")
  SEZ6_pos <- WhichCells(epi, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(epi, expression=CD276_group=="pos")
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
  
  p <- plot(euler(a, shape = "ellipse",main=st), quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=st)
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,".pdf")),g)
  #pdf((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,".pdf")))
  #print(p)
  #dev.off()

}


### split by treatment and subtype

### split by subtype
subtypes <- unique(epi_sub1$SCLC_subtype)
treat <- unique(epi_sub1$treatment)
### loop through three subtypes
for (st in subtypes){
  for (tr in treat){ 
    cells <- WhichCells(epi_sub1,expression=SCLC_subtype==st & treatment==tr)
    
    epi <- subset(epi_sub1,cells=cells)
  
    ## venn diagram
    
    ## positive cells
    
    DLL3_pos <- WhichCells(epi, expression=DLL3_group=="pos")
    SEZ6_pos <- WhichCells(epi, expression=SEZ6_group=="pos")
    B7H3_pos <- WhichCells(epi, expression=CD276_group=="pos")
    
    a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
    
    p <- plot(euler(a, shape = "ellipse",main=paste0(st,":",tr)), quantities = TRUE)
    g <- grid.arrange(grobs=list(p),top=paste0(st,":",tr))
    ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,"_",tr,".pdf")),g)
   
  }
    
  
}


### fill in summary table

#table(epi_sub1$SCLC_subtype,epi_sub1$SEZ6_group)/rowSums(table(epi_sub1$SCLC_subtype,epi_sub1$SEZ6_group))

sub <- subset(epi_sub1,cells=WhichCells(epi_sub1,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

sub1 <- subset(sub,subset=SCLC_subtype=="SCLC-P")
count <- colSums(GetAssayData(object = sub1, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## triple negative

sub <- subset(epi_sub,subset=SCLC_subtype=="SCLC-A")

length(WhichCells(sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

### NAIVE/treated
epi_sub2 <- subset(epi_sub1,subset=treatment=="treated")

table(epi_sub2$SCLC_subtype,epi_sub2$SEZ6_group)/rowSums(table(epi_sub2$SCLC_subtype,epi_sub2$SEZ6_group))

sub <- subset(epi_sub2,cells=WhichCells(epi_sub2,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))


sub1 <- subset(sub,subset=SCLC_subtype=="SCLC-P")
count <- colSums(GetAssayData(object = sub1, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## triple negative

sub <- subset(epi_sub2,subset=SCLC_subtype=="SCLC-P")

length(WhichCells(sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))


```

## Step 15: finer annotation

```{r}
## fine-annotation of cluster 6 and 15 (both B), 8 and 18 (both myeloid)

scrna.unintegrated <-  readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))


## join layers for RNA assay

scrna.unintegrated <- JoinLayers(scrna.unintegrated)


### get marker genes
Idents(scrna.unintegrated) <- "seurat_clusters"
relaxed.markers <- FindAllMarkers(scrna.unintegrated,logfc.threshold=0.25)

# Extract top 10 markers per cluster

top20 <- relaxed.markers %>% 
  group_by(cluster) %>% 
  top_n(n = 20, 
        wt = avg_log2FC)


write.xlsx(top20,paste0("../../tables/",Sys.Date(),"_unintegration_top20_relaxed_marker_genes_per_cluster_res0.2.xlsx"))


### SCLC markers

FeaturePlot(scrna.unintegrated, features = c("ASCL1","NEUROD1","POU2F3","YAP1","DLL3","SEZ6"))

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_SCLC_markers_cluster.pdf"))



### SCLC_subtype marker

FeaturePlot(scrna.unintegrated, features = c("ASCL1","NEUROD1","POU2F3","YAP1"))

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_SCLC_subtype_markers_cluster.pdf"))



## check a few cluster

cmarker <- FindMarkers(scrna.unintegrated,ident.1="8",ident.2="18")

cmarker %>%  top_n(n = 20,  wt = avg_log2FC)

### reannotate

# Rename all identities
scrna.unintegrated <- RenameIdents(object = scrna.unintegrated, 
                                  "0"="T",
                                  "1"="epithelial",
                                  "2"="epithelial",
                                  "3"="epithelial",
                                  "4"="epithelial",
                                  "5"="epithelial",
                                  "6"="B",
                                  "7"="epithelial",
                                  "8"="monocyte",
                                  "9"="epithelial",
                                  "10"="fibroblast",
                                  "11"="epithelial",
                                  "12"="epithelial",
                                  "13"="NK",
                                  "14"="endothelial",
                                  "15"="plasma",
                                  "16"="epithelial",
                                  "17"="epithelial",
                                  "18"="pDC",
                                  "19"="epithelial")

scrna.unintegrated$final_anno <- scrna.unintegrated@active.ident

DimPlot(
  scrna.unintegrated,
  reduction = "umap",
  group.by = "final_anno",
  label.size = 4,
  label=TRUE) + theme(legend.position = "none")

ggsave(paste0("../../plots/",Sys.Date(),"_unintegrated_UMAP_by_final_anno.pdf"))

# Save integrated object
saveRDS(scrna.unintegrated, paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))

## change factor level of final annotation

scrna.unintegrated$final_anno <- factor(scrna.unintegrated$final_anno, levels = c("epithelial", "endothelial","fibroblast","NK","T","B","plasma","monocyte","pDC"))
### get marker genes

topgenes <- c("ASCL1","EPCAM","DLL3","KRT8","KRT18","KRT19","PECAM1","VWF","ECSCR","COL3A1","COL1A1","DCN","NKG7","GNLY","GZMA","GZMB","CD3E","CD3D","CD8A","CD8B","CD19","MS4A1","BANK1","CD38","IGHG1","CD68","LYZ","CD14","S100A8","S100A9","LILRA4","CLEC4C","PTCRA")

scrna.unintegrated <- FindVariableFeatures(scrna.unintegrated,nfeatures=10000)

scrna.unintegrated <- ScaleData(scrna.unintegrated,vars.to.regress = c("percent.mito"))

DoHeatmap(scrna.unintegrated, features=topgenes,group.by="final_anno",raster=FALSE,size=3.5,angle=60,hjust=0.3,vjust=0)+ theme(axis.text.y = element_text(size = 8)) + NoLegend()

ggsave(paste0("../../plots/",Sys.Date(),"_heatmap_by_marker_genes_by_cluster.pdf"))

DoHeatmap(scrna.unintegrated, features=topgenes,group.by="final_anno",size=3.5,angle=60,hjust=0.3,vjust=0)+ theme(axis.text.y = element_text(size = 8)) + NoLegend()

ggsave(paste0("../../plots/",Sys.Date(),"_heatmap_by_marker_genes_by_cluster_raster.pdf"))


#### epithelial: exclude likely normal epithelial cells
#### visualize infercnv score for 17 and 19


infercnv_obj <- readRDS("../../inferCNV/output_rv/run.final.infercnv_obj")

### centered at 1
cnv <- infercnv_obj@expr.data -1 ## make it centered at 0

cnv_score <- colSums(abs(cnv))

## get index of subset in whole object

idx <- match(colnames(cnv),colnames(scrna.unintegrated))

scrna.unintegrated$totCNV <- 0

scrna.unintegrated$totCNV[idx] <- cnv_score


FeaturePlot(scrna.unintegrated,features="totCNV",col=c("lightgrey","red"),label=TRUE)

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_totCNV_allin.pdf")))
VlnPlot(scrna.unintegrated,features="totCNV",group.by="final_anno",pt.size=0)+ NoLegend()

ggsave(here(paste0("../../plots/",Sys.Date(),"_Violin_totCNV_allin.pdf")))

scrna.unintegrated$combine_anno <- paste0(scrna.unintegrated$final_anno,"_",scrna.unintegrated$seurat_clusters)
VlnPlot(scrna.unintegrated,features="totCNV",group.by="combine_anno",pt.size=0)+ NoLegend()

ggsave(here(paste0("../../plots/",Sys.Date(),"_Violin_totCNV_allin_combined_anno.pdf")))

## epithelial_17 is very likely to be normal

cnv_sd <- apply(abs(cnv),2,sd)

scrna.unintegrated$sdCNV <- 0

scrna.unintegrated$sdCNV[idx] <- cnv_sd
FeaturePlot(scrna.unintegrated,features="sdCNV",col=c("lightgrey","red"),label=TRUE)

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_sdCNV_allin.pdf")))
VlnPlot(scrna.unintegrated,features="sdCNV",group.by="final_anno",pt.size=0) + NoLegend()

ggsave(here(paste0("../../plots/",Sys.Date(),"_Violin_sdCNV_allin.pdf")))

VlnPlot(scrna.unintegrated,features="sdCNV",group.by="combine_anno",pt.size=0) + NoLegend()

ggsave(here(paste0("../../plots/",Sys.Date(),"_Violin_sdCNV_allin_combined_anno.pdf")))


scrna.unintegrated@meta.data <- scrna.unintegrated@meta.data[, -c(which(startsWith(colnames(scrna.unintegrated@meta.data),"DF")),which(startsWith(colnames(scrna.unintegrated@meta.data),"pANN")),which(startsWith(colnames(scrna.unintegrated@meta.data),"MP")))]

saveRDS(scrna.unintegrated, paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))



################################

### extract tumor cells only

scrna.tumor <- subset(scrna.unintegrated,subset=seurat_clusters %in% c("1","2","3","4","5","7","9","11","12","16","19"))

## remove unnecessary columns in meta.data

scrna.tumor@meta.data <- scrna.tumor@meta.data[, -c(which(startsWith(colnames(scrna.tumor@meta.data),"DF")),which(startsWith(colnames(scrna.tumor@meta.data),"pANN")),which(startsWith(colnames(scrna.tumor@meta.data),"MP")))]


epi_sub <- scrna.tumor

epi_sub <- NormalizeData(epi_sub)
epi_sub <- FindVariableFeatures(epi_sub)
epi_sub <- ScaleData(epi_sub,vars.to.regress = c("percent.mito"))
epi_sub <- RunPCA(epi_sub, npcs = 30, verbose = F)

epi_sub <- FindNeighbors(epi_sub, reduction = "pca", dims = 1:30)
epi_sub <- FindClusters(epi_sub, resolution = 0.1)

epi_sub <- RunUMAP(epi_sub, reduction = "pca",dims = 1:30, reduction.name = "umap")


DimPlot(
  epi_sub,
  reduction = "umap",
  group.by = c("sample"),
  combine = FALSE, label.size = 2
)

ggsave(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_by_sample_epi.pdf"))

DimPlot(
  epi_sub,
  reduction = "umap",
  label.size = 6,
  label=TRUE
) + theme(legend.position = "none")


ggsave(here(paste0("../../plots/",Sys.Date(),"_unintegration_UMAP_by_clusters_epi.pdf")))

### SCLC subtype annotation


## add known marker for SCLC subtype


gene_sets <- list(
  pASCL1 = c("ASCL1", "SOX4", "STMN2", "DOC2A"),
  pNEUROD1 = c("NEUROD1", "ADCYAP1","NRXN1", "SSTR2", "ID1", "ID3", "SST", "DLK1"),
  pPOU2F3 = c("POU2F3", "ASCL2", "CD44", "MYC", "KIT", "YBX1"),
  pYAP1=c("YAP1"),
  pInflamed= c("CD274","PDCD1","CD80","CD86","CTLA4","CD38","IDO1","TIGIT","ICOS","LAG3","CCL5","CXCL10",
              "HLA-A","HLA-B","HLA-C","HLA-E","HLA-F","HLA-G","HLA-DRA","HLA-DRB1","HLA-DQA1","HLA-DMA","HLA-DPA1",
              "HLA-DMB","HLA-DPB1","HLA-DQB1","HLA-DOA","HLA-DOB")
)

epi_sub <- AddModuleScore_UCell(epi_sub, features = gene_sets, ncores=4, name = "")

VlnPlot(epi_sub, features=names(gene_sets), pt.size = 0, ncol=2)

ggsave((paste0("../../plots/",Sys.Date(),"_SCLC_MP_score_epi.pdf")))


FeaturePlot(epi_sub,features=c("ASCL1","NEUROD1","POU2F3","YAP1","HLA-A","HLA-B","HLA-DRA","HLA-DRB1","CTLA4"))

ggsave((paste0("../../plots/",Sys.Date(),"_SCLC_marker_UMAP_epi.pdf")))



### classify individual cells for max

# Get the score matrix (remove the "1" suffix added by AddModuleScore)
score_matrix <- FetchData(epi_sub, vars = names(gene_sets))
colnames(score_matrix) <- c("SCLC-A","SCLC-N","SCLC-P","SCLC-Y","SCLC-I")

# Assign cell type based on highest score
epi_sub$SCLC_subtype <- colnames(score_matrix)[max.col(score_matrix)]

# Visualize results
DimPlot(epi_sub, group.by = "SCLC_subtype")

ggsave(paste0("../../plots/",Sys.Date(),"_UMAP_SCLC_sybtype_per_cell_using_addmodulescore.pdf"))


###plot stackedbar for subtype per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SCLC_subtype))
colnames(samType) <- c("Sample","cellType","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(cellType == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = cellType)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_SCLC_per_sample.pdf")))
```

## Step 16: check three target genes
```{r}
## calculate DLL3+ fraction per sample


## DLL3 pos vs neg
epi_sub$DLL3_group <- "neg"
 
epi_sub$DLL3_group[which(epi_sub[['RNA']]$data["DLL3",] > 0)] <- "pos" 
epi_sub$DLL3_group <- factor(epi_sub$DLL3_group)

## show UMAP
UMAPPlot(epi_sub, group.by="DLL3_group",cols=c("blue","red")) 
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_DLL3_pos_neg.pdf")))

### plot fraction of DLL3+ DLL3-

###plot stackedbar for subtype per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample,FUN = sum) * 100


## sort sample

reordered_data <- samType %>%
  filter(DLL3 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(text = element_text(size = 8),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_by_sample.pdf")))

## add treatment

epi_sub$treatment <- "naive"

epi_sub$treatment[which(epi_sub$sample %in% c("LP21","LP28","LP33","LP85"))] <- "treated"

## add stage

epi_sub$stage <- "metastatic"

epi_sub$stage[which(epi_sub$sample %in% c("LP19","LP21","LP91"))] <- "primary"


## separate by stage and treatment


## DLL3+ vs DLL3- by treatment

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$treatment))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by Treatment", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_by_treatment.pdf")))

### boxplot

df <- reordered_data %>%
   filter(DLL3 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%DLL3+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_per_sample_by_treatment.pdf")))


### violin plot

VlnPlot(epi_sub,features="DLL3",group.by="treatment",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_treatment.pdf")))


## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=DLL3_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "DLL3",
  cells=cells,
  group.by = "treatment",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_treatment_withBox.pdf")))


### violin plot with enhanced vlnplots for subtype

cells <- WhichCells(epi_sub, expression=(DLL3_group=="pos") & (SCLC_subtype %in% c("SCLC-A","SCLC-N","SCLC-P")))
  
VlnPlot2(
  epi_sub,
  features = "DLL3",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("#F8766D", "#00BF7D", "#00B0F6"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox.pdf")))


## DLL3+ vs DLL3- by stage

## DLL3+ vs DLL3- by treatment

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$stage))


colnames(samTreat) <- c("Sample","Stage","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Stage,desc(Value)) %>%
   select(Stage,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Stage","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by Stage", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_by_stage.pdf")))

### boxplot

df <- reordered_data %>%
   filter(DLL3 == "pos") %>%
   select(Stage, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Stage", y = "Value.y",
          color = "Stage", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%DLL3+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_per_sample_by_stage.pdf")))


##violin plot for DLL3 expression

VlnPlot(epi_sub,features="DLL3",group.by="stage",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_stage.pdf")))

## enhanced violin

cells <- WhichCells(epi_sub, expression=DLL3_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "DLL3",
  cells=cells,
  group.by = "stage",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_DLL3_by_stage_withBox.pdf")))


# Save integrated object
saveRDS(epi_sub, paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))



### SEZ6+ vs SEZ6-


## SEZ6 pos vs neg
epi_sub$SEZ6_group <- "neg"
 
epi_sub$SEZ6_group[which(epi_sub[['RNA']]$data["SEZ6",] > 0)] <- "pos" 
epi_sub$SEZ6_group <- factor(epi_sub$SEZ6_group)

## show UMAP
UMAPPlot(epi_sub, group.by="SEZ6_group",cols=c("blue","red")) + coord_fixed()
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_SEZ6_pos_neg.pdf")))

### plot fraction of SEZ6+ SEZ6-

###plot stackedbar per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample,FUN = sum) * 100


## sort sample

reordered_data <- samType %>%
  filter(SEZ6 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(text = element_text(size = 8),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_by_sample.pdf")))



### violin plot with enhanced vlnplots for subtype

cells <- WhichCells(epi_sub, expression=(SEZ6_group=="pos") & (SCLC_subtype %in% c("SCLC-A","SCLC-N","SCLC-P")))
  
VlnPlot2(
  epi_sub,
  features = "SEZ6",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("#F8766D", "#00BF7D", "#00B0F6"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_SCLC_subtype_withBox.pdf")))



## SEZ6+ vs SEZ6- by treatment

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$treatment))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(SEZ6 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample by Treatment", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_per_sample_by_treatment.pdf")))

### boxplot

df <- reordered_data %>%
   filter(SEZ6 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%SEZ6+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_per_sample_by_treatment.pdf")))

## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "SEZ6",
  cells=cells,
  group.by = "treatment",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_treatment_withBox.pdf")))




## SEZ6+ vs SEZ6- by stage

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$stage))


colnames(samTreat) <- c("Sample","Stage","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(SEZ6 == "pos") %>%
   arrange(Stage,desc(Value)) %>%
   select(Stage,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Stage","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample by Stage", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_per_sample_by_stage.pdf")))

### boxplot

df <- reordered_data %>%
   filter(SEZ6 == "pos") %>%
   select(Stage, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Stage", y = "Value.y",
          color = "Stage", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%SEZ6+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_per_sample_by_stage.pdf")))



## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "SEZ6",
  cells=cells,
  group.by = "stage",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_stage_withBox.pdf")))


### CD276+ vs CD276-


## CD276 pos vs neg
epi_sub$CD276_group <- "neg"
 
epi_sub$CD276_group[which(epi_sub[['RNA']]$data["CD276",] > 0)] <- "pos" 
epi_sub$CD276_group <- factor(epi_sub$CD276_group)

## show UMAP
UMAPPlot(epi_sub, group.by="CD276_group",cols=c("blue","red")) + coord_fixed()
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_CD276_pos_neg.pdf")))

### plot fraction of CD276+ CD276-

###plot stackedbar per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample,FUN = sum) * 100


## sort sample

reordered_data <- samType %>%
  filter(CD276 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(text = element_text(size = 8),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "CD276 expression per sample", y = "Proportion", x = "Sample")+ coord_fixed()

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_by_sample.pdf")))



### violin plot with enhanced vlnplots for subtype

cells <- WhichCells(epi_sub, expression=(CD276_group=="pos") & (SCLC_subtype %in% c("SCLC-A","SCLC-N","SCLC-P")))
  
VlnPlot2(
  epi_sub,
  features = "CD276",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("#F8766D", "#00BF7D", "#00B0F6"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_SCLC_subtype_withBox.pdf")))



## CD276+ vs CD276- by treatment

samType <- as.data.frame(table(epi_sub$sample,epi_sub$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$treatment))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(CD276 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "B7H3 expression per sample by Treatment", y = "Proportion", x = "",fill="B7H3")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_per_sample_by_treatment.pdf")))

### boxplot

df <- reordered_data %>%
   filter(CD276 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%CD276+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_CD276_per_sample_by_treatment.pdf")))

## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=CD276_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "CD276",
  cells=cells,
  group.by = "treatment",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_treatment_withBox.pdf")))




## CD276+ vs CD276- by stage

samType <- as.data.frame(table(epi_sub$sample,epi_sub$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(epi_sub$sample,epi_sub$stage))


colnames(samTreat) <- c("Sample","Stage","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(CD276 == "pos") %>%
   arrange(Stage,desc(Value)) %>%
   select(Stage,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Stage","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "CD276 expression per sample by Stage", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_per_sample_by_stage.pdf")))

### boxplot

df <- reordered_data %>%
   filter(CD276 == "pos") %>%
   select(Stage, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Stage", y = "Value.y",
          color = "Stage", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%CD276+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_CD276_per_sample_by_stage.pdf")))



## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(epi_sub, expression=CD276_group=="pos")
  
VlnPlot2(
  epi_sub,
  features = "CD276",
  cells=cells,
  group.by = "stage",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_stage_withBox.pdf")))


##############################

### overlap

# Save integrated object
saveRDS(epi_sub, paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))


## only keep top 3 SCLC-subtypes

cells <- WhichCells(epi_sub, expression=SCLC_subtype %in% c("SCLC-A", "SCLC-N", "SCLC-P"))

epi_sub1 <- subset(epi_sub,cells=cells) 

## Q1: What fraction of all SCLC are DLL3+ vs DLL3-, and what are the numbers across the classical different histological subsets?

data <- as.data.frame(epi_sub1@meta.data[,c("SCLC_subtype","DLL3_group")])

df <- data %>% group_by(SCLC_subtype,DLL3_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(DLL3=DLL3_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_DLL3_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =DLL3 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,labelpositionThreshold=0.01,showRatioThreshold = 0.005,
		ratioByGroup = FALSE)

dev.off()


#Same question for SEZ6, plus how much overlap is there for individual cells expressing DLL3 and SEZ6?

data <- as.data.frame(epi_sub1@meta.data[,c("SCLC_subtype","SEZ6_group")])

df <- data %>% group_by(SCLC_subtype,SEZ6_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(SEZ6=SEZ6_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_SEZ6_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =SEZ6 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,labelpositionThreshold=0.01,showRatioThreshold = 0.003,
		ratioByGroup = FALSE)

dev.off()

#Same question for B7H3, plus how much overlap is there for individual cells expressing DLL3, SEZ6 or B7H3?

data <- as.data.frame(epi_sub1@meta.data[,c("SCLC_subtype","CD276_group")])

df <- data %>% group_by(SCLC_subtype,CD276_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(CD276=CD276_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_CD276_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =CD276 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,labelpositionThreshold=0.005,showRatioThreshold = 0.0015,
		ratioByGroup = FALSE)

dev.off()


## cells with DLL3

DLL3_pos <- WhichCells(epi_sub1, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub1, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub1, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

## install.packages("eulerr")
pdf((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler.pdf")))
plot(euler(a, shape = "ellipse"), quantities = TRUE)
dev.off()


### split by subtype
subtypes <- unique(epi_sub1$SCLC_subtype)
genes <- c("DLL3","SEZ6","CD276")
### loop through three subtypes
for (st in subtypes){
  cells <- WhichCells(epi_sub1,expression=SCLC_subtype==st)
  
  epi <- subset(epi_sub1,cells=cells)
  
  ## loop through three genes
  for (gene in genes){
    group <- paste0(gene,"_group")
    data <- as.data.frame(epi@meta.data[,c(group,"SCLC_subtype")])
    
    df <- data %>% group_by(across(all_of(group))) %>% summarise(count=n(),.groups="drop") 
    
    fill_col <- sym(group)
    
    ggplot(df, aes(x = "", y = count, fill = !!fill_col)) +
      geom_col(color = "black") +
      geom_text(aes(label = count),
                position = position_stack(vjust = 0.5)) +
      coord_polar(theta = "y") +
      scale_fill_manual(values = c("blue","red")) + 
      labs(fill=gene) + ggtitle(st) + 
      theme_void() + theme(plot.title = element_text(hjust = 0.5,vjust = -6))
    
    ggsave(here(paste0("../../plots/",Sys.Date(),"_pie_",gene,"_",st,".pdf")))
  }
  ## venn diagram
  
  ## cells with DLL3
  
  DLL3_pos <- WhichCells(epi, expression=DLL3_group=="pos")
  SEZ6_pos <- WhichCells(epi, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(epi, expression=CD276_group=="pos")
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
  
  p <- plot(euler(a, shape = "ellipse",main=st), quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=st)
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,".pdf")),g)
  #pdf((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,".pdf")))
  #print(p)
  #dev.off()

}


### split by treatment and subtype

### split by subtype
subtypes <- unique(epi_sub1$SCLC_subtype)
treat <- unique(epi_sub1$treatment)
### loop through three subtypes
for (st in subtypes){
  for (tr in treat){ 
    cells <- WhichCells(epi_sub1,expression=SCLC_subtype==st & treatment==tr)
    
    epi <- subset(epi_sub1,cells=cells)
  
    ## venn diagram
    
    ## positive cells
    
    DLL3_pos <- WhichCells(epi, expression=DLL3_group=="pos")
    SEZ6_pos <- WhichCells(epi, expression=SEZ6_group=="pos")
    B7H3_pos <- WhichCells(epi, expression=CD276_group=="pos")
    
    a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
    
    p <- plot(euler(a, shape = "ellipse",main=paste0(st,":",tr)), quantities = TRUE)
    g <- grid.arrange(grobs=list(p),top=paste0(st,":",tr))
    ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,"_",tr,".pdf")),g)
   
  }
    
  
}

### fill in summary table
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))
#table(epi_sub1$SCLC_subtype,epi_sub1$SEZ6_group)/rowSums(table(epi_sub1$SCLC_subtype,epi_sub1$SEZ6_group))

cells <- WhichCells(epi_sub, expression=SCLC_subtype %in% c("SCLC-A", "SCLC-N", "SCLC-P"))

epi_sub1 <- subset(epi_sub,cells=cells) 

table(epi_sub1$SCLC_subtype,epi_sub1$SEZ6_group)

table(epi_sub1$SCLC_subtype,epi_sub1$SEZ6_group)/rowSums(table(epi_sub1$SCLC_subtype,epi_sub1$SEZ6_group))



sub <- subset(epi_sub1,cells=WhichCells(epi_sub1,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))


sub1 <- subset(sub,subset=SCLC_subtype=="SCLC-P")
count <- colSums(GetAssayData(object = sub1, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## triple negative

sub <- subset(epi_sub,subset=SCLC_subtype=="SCLC-P")

length(WhichCells(sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

### NAIVE/treated
epi_sub2 <- subset(epi_sub1,subset=treatment=="treated")

table(epi_sub2$SCLC_subtype,epi_sub2$CD276_group)/rowSums(table(epi_sub2$SCLC_subtype,epi_sub2$CD276_group))

sub <- subset(epi_sub2,cells=WhichCells(epi_sub2,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))


sub1 <- subset(sub,subset=SCLC_subtype=="SCLC-P")
count <- colSums(GetAssayData(object = sub1, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## triple negative

sub <- subset(epi_sub2,subset=SCLC_subtype=="SCLC-P")

length(WhichCells(sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

### total triple negative

count <- colSums(GetAssayData(object = epi_sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)

#    0    1    2    3 
# 9113 6693 3224  790 

table(count)/sum(table(count))
#          0          1          2          3 
# 0.45978809 0.33768920 0.16266398 0.03985873 


#### DLL3+ per sample

table(epi_sub$sample,epi_sub$DLL3_group)/rowSums(table(epi_sub$sample,epi_sub$DLL3_group))
dsum <- as.data.frame.matrix(table(epi_sub$sample,epi_sub$DLL3_group)/rowSums(table(epi_sub$sample,epi_sub$DLL3_group)))
# > dsum
#            neg       pos
# LP19 0.7569241 0.2430759
# LP20 0.7010652 0.2989348
# LP21 0.6396700 0.3603300
# LP28 0.4331683 0.5668317
# LP33 0.4352332 0.5647668
# LP34 0.7815235 0.2184765
# LP35 0.5721584 0.4278416
# LP71 0.6883762 0.3116238
# LP85 0.6671941 0.3328059
# LP91 0.5644531 0.4355469
# > median(dsum$pos)
# [1] 0.3465679
# > range(dsum$pos)
# [1] 0.2184765 0.5668317
## get CPM
epi_cpm <- NormalizeData(epi_sub, assay = "RNA", normalization.method = "RC", scale.factor = 1000000)

DLL3_CPM <- epi_cpm@assays$RNA$data["DLL3",which(epi_cpm@assays$RNA$data["DLL3",])>0]
#> range(DLL3_CPM)
#[1]   12.68649 1614.20500

## naive 

summary(dsum[c("LP19","LP20","LP34","LP35","LP71","LP91"),"pos"])

## treated

summary(dsum[c("LP21","LP28","LP33","LP85"),"pos"])

```

## Step 17: venn diagram at sample-level

```{r}
## try count at sample subtype level
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))


### pie for three genes
##DLL3 

data <- as.data.frame(epi_sub@meta.data[,c("DLL3_group")])
colnames(data) <- "DLL3_group"

df <- data %>% group_by(DLL3_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(DLL3=DLL3_group)


ggplot(df, aes(x = "", y = count, fill = DLL3)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar(theta = "y") +
  geom_text(aes(label = paste(count,"(", round(count/sum(count) * 100), "%)")), 
            position = position_stack(vjust = 0.5)) +
  scale_fill_manual(values = c( "blue", "red")) +
  theme_void()

ggsave((paste0("../../plots/",Sys.Date(),"_pie_DLL3.pdf")))

##SEZ6

data <- as.data.frame(epi_sub@meta.data[,c("SEZ6_group")])
colnames(data) <- "SEZ6_group"

df <- data %>% group_by(SEZ6_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(SEZ6=SEZ6_group)


ggplot(df, aes(x = "", y = count, fill = SEZ6)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar(theta = "y") +
  geom_text(aes(label = paste(count,"(", round(count/sum(count) * 100), "%)")), 
            position = position_stack(vjust = 0.5)) +
  scale_fill_manual(values = c( "blue", "red")) +
  theme_void()

ggsave((paste0("../../plots/",Sys.Date(),"_pie_SEZ6.pdf")))

##CD276 

data <- as.data.frame(epi_sub@meta.data[,c("CD276_group")])
colnames(data) <- "CD276_group"

df <- data %>% group_by(CD276_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(B7H3=CD276_group)


ggplot(df, aes(x = "", y = count, fill = B7H3)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar(theta = "y") +
  geom_text(aes(label = paste(count,"(", round(count/sum(count) * 100), "%)")), 
            position = position_stack(vjust = 0.5)) +
  scale_fill_manual(values = c( "blue", "red")) +
  theme_void()

ggsave((paste0("../../plots/",Sys.Date(),"_pie_CD276.pdf")))

### venn diagram for three genes in all samples


DLL3_pos <- WhichCells(epi_sub, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

## install.packages("eulerr")

p<- plot(euler(a, shape = "ellipse"), quantities = TRUE)
g <- grid.arrange(grobs=list(p),top="SCLC-A")

ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_all_sample.pdf")),g)


#### count for total


table(epi_sub$DLL3_group)

table(epi_sub$DLL3_group)/sum(table(epi_sub$DLL3_group))


sub <- subset(epi_sub,cells=WhichCells(epi_sub,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))


count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## triple negative

length(WhichCells(epi_sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = epi_sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

### NAIVE/treated
epi_sub2 <- subset(epi_sub,subset=treatment=="naive")
table(epi_sub2$CD276_group)

table(epi_sub2$CD276_group)/sum(table(epi_sub2$CD276_group))

sub <- subset(epi_sub2,cells=WhichCells(epi_sub2,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## triple negative


length(WhichCells(epi_sub2,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = epi_sub2, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

### total triple negative

count <- colSums(GetAssayData(object = epi_sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)

#    0    1    2    3 
# 9113 6693 3224  790 

table(count)/sum(table(count))
#          0          1          2          3 
# 0.45978809 0.33768920 0.16266398 0.03985873 


```

## step 18: a few figures

```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

FeaturePlot(epi_sub,features="DLL3")

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_by_DLL3_002_maligOnly.pdf")))


FeaturePlot(epi_sub,features="SEZ6") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_by_SEZ6_002_maligOnly.pdf")))


FeaturePlot(epi_sub,features="CD276")

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_by_CD276_002_maligOnly.pdf")))


FeaturePlot3(epi_sub, feature.1 = "DLL3", feature.2 = "SEZ6", feature.3 = "CD276", pt.size = 1)

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_by_3targets_002_maligOnly.pdf")))


### Supplementary Table X

#prop.test
xtab <- as.table(rbind(c(4278,14660-4278),c(1930,5160-1930)))

dimnames(xtab) <- list(
  group = c("MGH", "public"),
  target = c("pos", "neg")
)

library(rstatix)
prop_test(xtab, detailed = TRUE)


#### dentisty plot for redefining positive cells

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

genes_to_plot <- c("DLL3","SEZ6","CD276","sample")

# 1. Fetch data
data_to_plot <- FetchData(epi_sub, vars = c(genes_to_plot))

# 2. Prepare data for ggplot2
data_long <- pivot_longer(data_to_plot, cols = all_of(genes_to_plot[1:3]), names_to = "gene", values_to = "expression")

# 3. Create ggplot2 density plot
ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  facet_wrap(~ sample) +
  labs(title = "Gene Expression Density", x = "Expression (log-normalized)", y = "Density") +
  scale_fill_brewer(palette = "Set2") +
  scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_3targets_maligOnly_splitBy_Sample.pdf")))


### dentisy of raw UMI counts


# 1. Fetch data
data_to_plot <- FetchData(epi_sub, vars = c(genes_to_plot),layer="count")

# 2. Prepare data for ggplot2
data_long <- pivot_longer(data_to_plot, cols = all_of(genes_to_plot[1:3]), names_to = "gene", values_to = "expression")

# 3. Create ggplot2 density plot
ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  facet_wrap(~ sample) +
  labs(title = "Gene Expression Density", x = "Expression (raw #UMI)", y = "Density") +
  scale_fill_brewer(palette = "Set2") +
  scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_3targets_maligOnly_splitBy_Sample_rawUMI.pdf")))


```

## Step 19: attempts for redefining positive cells with raw UMI cutoffs

```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

ucuts <- c(1,2,3,5)

for (cut in ucuts){
  
  SEZ6_pos <- WhichCells(epi_sub, expression=SEZ6 >=cut,slot="count")
  B7H3_pos <- WhichCells(epi_sub, expression=CD276 >=cut,slot="count")
  DLL3_pos <- WhichCells(epi_sub, expression=DLL3 >=cut,slot="count") 
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
    
  p <- plot(euler(a, shape = "ellipse",main=st), quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=paste0("(>=",cut,"UMI)"))
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",cut,"UMI.pdf")),g)
    
}

### chose 1 as cutoff


```

## Step 20: finalize some figures on 2025-06-30

```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## figure 3A: 2025-04-13_stackBar_SCLC_per_sample.pdf

## figure 3B: DLL3 fraction sample in the same order as 3A

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samType <- samType %>% mutate(Sample = fct_relevel(Sample, 
            "LP33", "LP21", "LP35", 
            "LP91", "LP34", "LP28", 
            "LP85", "LP19","LP20","LP71"))


ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_reorder.pdf")))


## 3C: 2025-03-18_stackedbar_by_DLL3_per_sample_by_main_SCLC_subtype_postfilter.pdf (CancerCell)
## 3D: 2025-05-22_boxplot_by_DLL3_per_sample_by_treatment.pdf (MGH)
## 3E: 2025-03-18_boxplot_by_DLL3_per_sample_by_treatment_postfilter.pdf (CancerCell)
## 3F: bell shaped heterogeneity of DLL3 in tumor cells in MGH


genes_to_plot <- c("DLL3","sample")

## normalized data

# 1. Fetch data
data_to_plot <- FetchData(epi_sub, vars = c(genes_to_plot))

# 2. Prepare data for ggplot2
data_long <- pivot_longer(data_to_plot, cols = all_of(genes_to_plot[1]), names_to = "gene", values_to = "expression")

# 3. Create ggplot2 density plot
ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  facet_wrap(~ sample) +
  labs(title = "DLL3 Expression Density", x = "Expression (log-normalized)", y = "Density") #+
  #scale_fill_brewer(palette = "Set2") +
  #scale_color_brewer(palette = "Set2")

## figure S3A
ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_splitBy_Sample.pdf")))


## all samples together

ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  #facet_wrap(~ sample) +
  labs(title = "DLL3 Expression Density", x = "Expression (log-normalized)", y = "Density") #+
  #scale_fill_brewer(palette = "Set2") +
  #scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_pooledSample.pdf")))


## 3G: Cancer Cell


### Figure 7


## 7A:2025-05-22_unintegrated_UMAP_by_final_anno.pdf 

## update 7A showing tumor cells with binary expression

epi_sub$DLL3_ <- ifelse(epi_sub$DLL3_group=="neg",0,1)
epi_sub$SEZ6_ <- ifelse(epi_sub$SEZ6_group=="neg",0,1)
epi_sub$B7H3_ <- ifelse(epi_sub$CD276_group=="neg",0,1)

FeaturePlot3(epi_sub, feature.1 = "DLL3_", feature.2 = "SEZ6_", feature.3 = "B7H3_", pt.size = 1, color.range=c(0,1))

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_SEZ6_DLL3_B7H3_binary_tumorOnly.pdf")))

DimPlot(epi_sub,group.by = "sample")

### check original with continuous variable

FeaturePlot3(epi_sub, feature.1 = "DLL3", feature.2 = "SEZ6", feature.3 = "CD276", pt.size = 1, color.range=c(0,1))


## 7B: A wants the same size for circles

## cells with DLL3

DLL3_pos <- WhichCells(epi_sub, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

ggvenn(a)

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_sameSize.pdf")))


```

## Step 21: explore co-expression variation across individuals

```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

for (samp in unique(epi_sub$sample)){
  epi_tmp <- subset(epi_sub,subset=sample==samp)
  SEZ6_pos <- WhichCells(epi_tmp, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(epi_tmp, expression=CD276_group=="pos")
  DLL3_pos <- WhichCells(epi_tmp, expression=DLL3_group=="pos") 
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
    
  p <- plot(euler(a, shape = "circle"), quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=samp)
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",samp,".pdf")),g)
    
}



### 3A: re-annotate 

# Get the score matrix (remove the "1" suffix added by AddModuleScore)
score_matrix <- FetchData(epi_sub, vars = c("pASCL1","pNEUROD1","pPOU2F3","pInflamed" ))
colnames(score_matrix) <- c("SCLC-A","SCLC-N","SCLC-P","SCLC-I")

# Assign cell type based on highest score
epi_sub$SCLC_subtype2 <- colnames(score_matrix)[max.col(score_matrix)]

saveRDS(epi_sub,paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

# Visualize results
DimPlot(epi_sub, group.by = "SCLC_subtype2")

ggsave(paste0("../../plots/",Sys.Date(),"_UMAP_SCLC_sybtype2_per_cell_using_addmodulescore.pdf"))


###plot stackedbar for subtype per sample

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SCLC_subtype2))
colnames(samType) <- c("Sample","Subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(Subtype == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = Subtype)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_SCLC_per_sample_4subtype.pdf")))


## 3B

samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samType <- samType %>% mutate(Sample = fct_relevel(Sample, 
            "LP33", "LP21", "LP91", 
            "LP35", "LP34", "LP28", 
            "LP85", "LP20","LP71","LP19"))


ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_reorder.pdf")))

### calculate correlation between % subtypes and % DLL3+

## get fractions of subtypes per sample

df_st <- data.frame(unclass(table(epi_sub$sample, epi_sub$SCLC_subtype2)/rowSums(table(epi_sub$sample, epi_sub$SCLC_subtype2))))

colnames(df_st) <- c("SCLC_A","SCLC_I","SCLC_N","SCLC_P")

df_st$sample <- rownames(df_st)

# df_st$SCLC_A
#  [1] 0.9628543 0.9427430 0.7585302 0.8836634 0.6321244 0.8333874 0.8173691 0.9585280 0.9092827 0.8144531

# summary(df_st$SCLC_A)

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.6321  0.8152  0.8585  0.8513  0.9344  0.9629 

## get fractions of DLL3 positive cells per sample

df_dl <- data.frame(unclass(table(epi_sub$sample, epi_sub$DLL3_group)/rowSums(table(epi_sub$sample, epi_sub$DLL3_group))))

df_dl$sample <- rownames(df_dl)

df_dl <- df_dl[,c("sample","pos")]

colnames(df_dl) <- c("sample","DLL3")

# df_dl$DLL3
# 
#  [1] 0.2430759 0.2989348 0.3603300 0.5668317 0.5647668 0.2184765 0.4278416 0.3116238 0.3328059
# [10] 0.4355469

#summary(df_dl$DLL3)

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.2185  0.3021  0.3466  0.3760  0.4336  0.5668 

## join

df <- merge(df_st,df_dl)

# List of x variables
x_vars <- c("SCLC_A", "SCLC_N", "SCLC_P", "SCLC_I")

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  sp <- ggscatter(df, x = x, y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE
  ) + xlab(paste0("%", x)) + ylab("%DLL3+")
  
  # Add correlation
  p <- sp + stat_cor(method = "pearson")
  
  # Store plot in list with name
  plot_list[[x]] <- p
}
 
ggarrange(plotlist = plot_list, ncol = 2, nrow = 2)

ggsave(here(paste0("../../plots/",Sys.Date(),"_pearson_corr_DLL3_by_subtype.pdf")))

### use spearman 

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  sp <- ggscatter(df, x = x, y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE
  ) + xlab(paste0("%", x)) + ylab("%DLL3+")
  
  # Add correlation
  p <- sp + stat_cor(method = "spearman",cor.coef.name ="rho")
  
  # Store plot in list with name
  plot_list[[x]] <- p
}
 
ggarrange(plotlist = plot_list, ncol = 2, nrow = 2)

ggsave(here(paste0("../../plots/",Sys.Date(),"_spearman_corr_DLL3_by_subtype.pdf")))



### get correlation for only subtype-A

sp <- ggscatter(df, x = "SCLC_A", y = "DLL3",
                add = "reg.line",
                add.params = list(color = "blue", fill = "lightgray"),
                conf.int = TRUE
  ) + xlab("%SCLC-A") + ylab("%DLL3+")

# Add correlation
sp + stat_cor(method = "pearson")
ggsave(here(paste0("../../plots/",Sys.Date(),"_pearson_corr_DLL3_and_subtypeA.pdf")))

### get density plot of % DLL3+

ggplot(df, aes(x = DLL3)) +
  geom_density() +
  theme_bw() +
  #facet_wrap(~ sample) +
  labs(title = "%DLL3 + Density", x = "%DLL3+", y = "Density") #+
  #scale_fill_brewer(palette = "Set2") +
  #scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_pos_fraction.pdf")))


## histogram with curve

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

df <- data.frame(unclass(table(epi_sub$sample, epi_sub$DLL3_group)/rowSums(table(epi_sub$sample, epi_sub$DLL3_group))))[,2]

df <- data.frame(df)
colnames(df) <- "DLL3"

# Histogram 
ggplot(df, aes(x = DLL3)) + 
  geom_histogram() #+geom_density(aes(y=..count..))

ggsave(here(paste0("../../plots/",Sys.Date(),"_hist_by_DLL3_pos_fraction.pdf")))

### histogram with curve

ggplot(df, aes(x = DLL3)) + 
  geom_histogram() +geom_density(aes(y=..density..)) + theme_bw()


ggsave(here(paste0("../../plots/",Sys.Date(),"_histWithCurve_by_DLL3_pos_fraction.pdf")))


### 7A with stacked bar graph with same order

## DLL3
samType <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_alpha_order.pdf")))

## CD276

samType <- as.data.frame(table(epi_sub$sample,epi_sub$CD276_group))
colnames(samType) <- c("Sample","B7H3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

ggplot(samType, aes(x = Sample, y = Value, fill = B7H3)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "B7H3 expression per sample", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_B7H3_per_sample_alpha_order.pdf")))

## SEZ6
samType <- as.data.frame(table(epi_sub$sample,epi_sub$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

ggplot(samType, aes(x = Sample, y = Value, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_per_sample_alpha_order.pdf")))

#### count %DLL3+

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

df <- data.frame(unclass(table(epi_sub$sample,epi_sub$DLL3_group)/rowSums(table(epi_sub$sample,epi_sub$DLL3_group))))

summary(df$pos)

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.2185  0.3021  0.3466  0.3760  0.4336  0.5668 

dll3_pos <- WhichCells(epi_sub,expression=DLL3_group=="pos")

dll3_normalized <- epi_sub@assays$RNA$data["DLL3", dll3_pos]

range(dll3_normalized)
# [1] 0.1194394 2.8415345

## get UMI count per million

## get CPM
epi_cpm <- NormalizeData(epi_sub, assay = "RNA", normalization.method = "RC", scale.factor = 1000000)

DLL3_CPM <- epi_cpm@assays$RNA$data["DLL3",which(epi_cpm@assays$RNA$data["DLL3",]>0)]
#> range(DLL3_CPM)
#[1]   12.68649 1614.20500

## treatment
table(epi_sub$sample,epi_sub$treatment)

## 6 naive and 4 treated

## %DLL3 by treatment
## naive
summary(df$pos[which(rownames(df) %in% c("LP19","LP20","LP34","LP35","LP71","LP91"))])

#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.2185  0.2570  0.3053  0.3226  0.3988  0.4355 

## treated
summary(df$pos[which(!rownames(df) %in% c("LP19","LP20","LP34","LP35","LP71","LP91"))])

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.3328  0.3534  0.4625  0.4562  0.5653  0.5668 

###
#####
## 7C: circles but different sizes

## cells with DLL3

DLL3_pos <- WhichCells(epi_sub, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

p <- plot(euler(a, shape = "circle"), fills = list(fill = c("red","orange", "steelblue4")),quantities = TRUE)
g <- grid.arrange(grobs=list(p))

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize.pdf")),g)

## include all cells

DLL3_pos <- WhichCells(epi_sub, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub, expression=CD276_group=="pos")
total <- colnames(epi_sub)

a <- list(`negative`=total,`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

p <- plot(euler(a, shape = "circle"), fills = list(fill = c("gray","red","orange", "steelblue4")),quantities = TRUE)
g <- grid.arrange(grobs=list(p))

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize_allIn.pdf")),g)



### breakdown of positive cells by treatment

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## get naive cells

epi_sub_naive <- subset(epi_sub,subset=treatment=="naive")


DLL3_pos <- WhichCells(epi_sub_naive, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub_naive, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub_naive, expression=CD276_group=="pos")
total <- colnames(epi_sub_naive)

a <- list(`negative`=total,`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

p <- plot(euler(a, shape = "circle"), fills = list(fill = c("gray","red","orange", "steelblue4")),quantities = TRUE)
g <- grid.arrange(grobs=list(p))

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize_allIn_naive.pdf")),g)


## get treated cells

epi_sub_treated <- subset(epi_sub,subset=treatment=="treated")


DLL3_pos <- WhichCells(epi_sub_treated, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub_treated, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub_treated, expression=CD276_group=="pos")
total <- colnames(epi_sub_treated)

a <- list(`negative`=total,`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

p <- plot(euler(a, shape = "circle"), fills = list(fill = c("gray","red","orange", "steelblue4")),quantities = TRUE)
g <- grid.arrange(grobs=list(p))

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize_allIn_treated.pdf")),g)

### 2025-06-13-SCLC_DLL3_Supplementary_Table_X_ML: sheet Supp_Table_Y_cell_level
## prop.test between naive and treated
## DLL3+ using all cells in each group as total
res <- prop.test(x=c(4278,1930),n=c(14660,5160))
res$p.value

## SEZ6+
res <- prop.test(x=c(3921,1221),n=c(14660,5160))
res$p.value

## B7H3+
res <- prop.test(x=c(3277,884),n=c(14660,5160))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(1520,657),n=c(4278,1930))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(1274,478),n=c(4278,1930))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(606,184),n=c(4278,1930))
res$p.value




#############
## at sample level

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

### create a dataframe to store these following metrics:

# sample identity
# cohort
# subtype info
# treatment
# %DLL3+
# %SEZ6+
# %B7H3+
# %SEZ6+ in DLL3+
# %B7H3+ in DLL3+
# %SEZ6+&B7H3+ in DLL3+
# any of three target +

df_sample <-  data.frame(
  Sample = character(),
  Cohort = character(),
  Subtype = character(),
  Treatment = character(),
  Pdll3 = numeric(),
  Psez6 = numeric(),
  Pb7h3 = numeric(),
  Psind = numeric(),
  Pbind = numeric(),
  P2ind = numeric(),
  Pany3 = numeric(),
  stringsAsFactors = FALSE
)

for (samp in unique(epi_sub$sample)){
  epi_sub1 <- subset(epi_sub, subset=sample==samp)
  treat <- ifelse(unique(epi_sub1$treatment)=="naive","Naive","Treated")
  
  
  ## total cell
  tcell <- ncol(epi_sub1)
  
  ## DLL3+
  
  cdll3 <- length(which(epi_sub1$DLL3_group=="pos"))
  
  tmp_Pdll3 <- cdll3/tcell
  
  ## SEZ6+
  csez6 <- length(which(epi_sub1$SEZ6_group=="pos"))
  
  tmp_Psez6 <- csez6/tcell  
  
  ## B7H3+
  cb7h3<- length(which(epi_sub1$CD276_group=="pos"))
  
  tmp_Pb7h3 <- cb7h3/tcell  
  
  
  ## SEZ6+ in DLL3+
  
  csind <- length(which(epi_sub1$DLL3_group=="pos" & epi_sub1$SEZ6_group=="pos"))
  
  tmp_Psind <- csind/cdll3
  
  ## B7H3+ in DLL3+
  
  cbind <- length(which(epi_sub1$DLL3_group=="pos" & epi_sub1$CD276_group=="pos"))
  
  tmp_Pbind <- cbind/cdll3
  
  ## SEZ6+&B7H3+ in DLL3+
  
  c2ind <- length(which(epi_sub1$DLL3_group=="pos" & epi_sub1$CD276_group=="pos" & epi_sub1$SEZ6_group=="pos"))
  tmp_P2ind <- c2ind/cdll3
  
  ## positive for any of the three targets
  cany3 <- length(which(epi_sub1$DLL3_group=="pos" | epi_sub1$SEZ6_group=="pos" | epi_sub1$CD276_group=="pos" ))
  tmp_Pany3 <- cany3/tcell
  
  
   # Create a data frame for the new row
  new_row <- data.frame(Sample = samp, 
                        Cohort = "MGH", 
                        Subtype = "SCLC-A", 
                        Treatment = treat, 
                        Pdll3 =tmp_Pdll3,
                        Psez6 =tmp_Psez6,
                        Pb7h3 = tmp_Pb7h3,
                        Psind = tmp_Psind,
                        Pbind = tmp_Pbind,
                        P2ind = tmp_P2ind,
                        Pany3 = tmp_Pany3,
                        stringsAsFactors = FALSE)

  # Add the new row to the data frame using rbind()
  df_sample <- rbind(df_sample, new_row)
}

write.xlsx(df_sample,paste0("../../tables/",Sys.Date(),"_coexpression_summary_at_sample_level.xlsx"))

### load summary table from Cancer Cell data

df_cc <- read.xlsx("../../../SCLC/tables/2025-07-07_coexpression_summary_at_sample_level.xlsx")

## compare SCLC-A within MGH cohort

## between treatment

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value


## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value



## compare SCLC-A between cohort

## naive

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_cc$Pdll3[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Naive")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_cc$Psez6[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Naive")])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_cc$Pb7h3[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Naive")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_cc$Psind[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Naive")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_cc$Pbind[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Naive")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_cc$P2ind[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Naive")])
res$p.value

## treated

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_cc$Pdll3[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Treated")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_cc$Psez6[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Treated")])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_cc$Pb7h3[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Treated")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_cc$Psind[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Treated")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_cc$Pbind[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Treated")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_cc$P2ind[which(df_cc$Subtype=="SCLC-A" & df_cc$Treatment=="Treated")])
res$p.value

## total
## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" )], df_cc$Pdll3[which(df_cc$Subtype=="SCLC-A")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" )], df_cc$Psez6[which(df_cc$Subtype=="SCLC-A" )])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" )], df_cc$Pb7h3[which(df_cc$Subtype=="SCLC-A" )])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" )], df_cc$Psind[which(df_cc$Subtype=="SCLC-A" )])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" )], df_cc$Pbind[which(df_cc$Subtype=="SCLC-A" )])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" )], df_cc$P2ind[which(df_cc$Subtype=="SCLC-A" )])
res$p.value

## sequencing depth

   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 1001    4700    8748   11138   14196  170785 


### plot density plot for imaging-based assessment of % DLL3+ CTCs in 16 patients based on swimmer plot

df <- data.frame(DLL3=c(0.1,0.08,0.07,0,0.2,0.43,0.13,0.79,0.88,0.8,0.5,0.07,1,0.03,0.96,0.75))

ggplot(df, aes(x = DLL3)) +
  geom_histogram() +geom_density(aes(y=..density..)) + theme_bw() +
  labs(title = "Imaging-based", x = "%DLL3+", y = "Density") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_pos_fraction_Imaging_CTC.pdf")))


### pool all primary tumors together and plot

## load MGH data
df_MGH <- read.xlsx(paste0("../../tables/2025-07-07_coexpression_summary_at_sample_level.xlsx"))

### load summary table from Cancer Cell data

df_cc <- read.xlsx("../../../SCLC/tables/2025-07-07_coexpression_summary_at_sample_level.xlsx")

df <- rbind(df_MGH,df_cc)

ggplot(df, aes(x = Pdll3)) +
  geom_histogram() +geom_density(aes(y=..density..)) + theme_bw() +
  labs(title = "Pooled primary tumors", x = "%DLL3+", y = "Density") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_pos_fraction_pooled_primary_tumors.pdf")))


### SCLC-A only

ggplot(df[which(df$Subtype=="SCLC-A"),], aes(x = Pdll3)) +
  geom_histogram() +geom_density(aes(y=..density..)) + theme_bw() +
  labs(title = "Pooled primary tumors: SCLC-A only", x = "%DLL3+", y = "Density") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_pos_fraction_pooled_primary_tumors_SCLC-A_only.pdf")))


### 2025-07-09 update figures for Figure 7 accoording to 2025-07-09-figure 7 outline-dh.pptx

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))
## A: Show UMAP for the 10 MGH tumors, Label each tumor and show within Each tumor the color for DLL3, SEZ6, B7H3 (color consistent throughout the figure) With each color in small dot, so it looks Like mosaic, rather than brown aggregate)

## 2025-07-16 switch the color of DLL3 and SEZ6 to make orange for DLL3
epi_sub$DLL3_ <- ifelse(epi_sub$DLL3_group=="neg",0,1)
epi_sub$SEZ6_ <- ifelse(epi_sub$SEZ6_group=="neg",0,1)
epi_sub$B7H3_ <- ifelse(epi_sub$CD276_group=="neg",0,1)

plot <- FeaturePlot3(epi_sub, feature.2 = "DLL3_", feature.1 = "SEZ6_", feature.3 = "B7H3_", pt.size = 0.1, color.range=c(0,1)) 

## add text labels for samples/clusters

umap_coords <- Embeddings(epi_sub, "umap") %>% as.data.frame()

umap_coords$sample <- epi_sub$sample

min_max_normalize <- function(x) {
  return ((x - min(x)) / (max(x) - min(x)))
}
cluster_centers <- umap_coords %>%
  dplyr::group_by(sample) %>%
  dplyr::summarise(
    umap_1 = mean(min_max_normalize(umap_1)),
    umap_2 = mean(min_max_normalize(umap_2))
  )


#plot + annotate("text",x=cluster_centers$umap_1,y=cluster_centers$umap_2,label=cluster_centers$sample)

## tried median/mean combined with min_max normalization but could not figure out an automatic way
## manually adjust for each sample for now
### test out and manually adjust coordinates for samples to match that from Dimplot (group.by sample)

as.data.frame(cluster_centers)

#    sample    umap_1    umap_2
# 1    LP19 0.2639542 0.8656770
# 2    LP20 0.1425793 0.5340136
# 3    LP21 0.4665456 0.2856084
# 4    LP28 0.8791862 0.5508492
# 5    LP33 0.0519351 0.1098735
# 6    LP34 0.7851148 0.8139349
# 7    LP35 0.3744394 0.7030061
# 8    LP71 0.5409432 0.8615526
# 9    LP85 0.1663942 0.7740266
# 10   LP91 0.2674083 0.1738285

cluster_centers2 <- as.data.frame(cluster_centers)

cluster_centers2[cluster_centers2$sample=="LP19",2:3] <- c(0.43,0.94)
cluster_centers2[cluster_centers2$sample=="LP20",2:3] <- c(0.2,0.54)
cluster_centers2[cluster_centers2$sample=="LP21",2:3] <- c(0.46,0.20)
cluster_centers2[cluster_centers2$sample=="LP28",2:3] <- c(0.67,0.32)
cluster_centers2[cluster_centers2$sample=="LP33",2:3] <- c(0.32,0.36)
cluster_centers2[cluster_centers2$sample=="LP34",2:3] <- c(0.77,0.54)
cluster_centers2[cluster_centers2$sample=="LP35",2:3] <- c(0.57,0.38)
cluster_centers2[cluster_centers2$sample=="LP71",2:3] <- c(0.57,0.76)
cluster_centers2[cluster_centers2$sample=="LP85",2:3] <- c(0.35,0.58)
cluster_centers2[cluster_centers2$sample=="LP91",2:3] <- c(0.44,0.39)

#plot + annotate("text",x=0.44,y=0.4,label="LP91")

plot + annotate("text",x=cluster_centers2$umap_1,y=cluster_centers2$umap_2,label=cluster_centers2$sample) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_SEZ6_DLL3_B7H3_binary_tumorOnly_withLabel.pdf")))

DimPlot(epi_sub,group.by = "sample",label=TRUE) + NoLegend()


## 7B: single cell expression of DLL3, SEZ6, B7H3 in 10 tumors

DLL3_pos <- WhichCells(epi_sub, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

## ggVennDiagram could not change the size of circles
#ggVennDiagram(a, shape_id = "301f") + scale_fill_distiller(palette = "RdBu")

## use euler 
p <- plot(euler(a, shape = "circle"), quantities = list(type = c("counts", "percent")),fill = list(fill = c("orangered", "gold","royalblue")))
g <- grid.arrange(grobs=list(p))
ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_bothCP_differentSize.pdf")),g)


## 7C: single cell expression Of EpCAM vs “any of DLL3, SEZ6, B7H3” Out of denominator of CNV+ cells  in 10 tumors

EPCAM_pos <- WhichCells(epi_sub, expression=EPCAM > 0)
Any1_pos <- WhichCells(epi_sub, expression=CD276_group=="pos" | DLL3_group=="pos" | SEZ6_group=="pos")
all <- colnames(epi_sub)

a <- list(`negative`=all,`EpCAM+`=EPCAM_pos,`DLL3/SEZ6/B7H3+`=Any1_pos)

p <- plot(euler(a, shape = "circle"), quantities = list(type = c("counts", "percent")),fill = list(fill = c("gray", "lightblue","orangered")))
g <- grid.arrange(grobs=list(p))
ggsave((paste0("../../plots/",Sys.Date(),"_venn_allIn_euler.pdf")),g)

## let's use ggvenn and add rectangle to the venn diagram

a2 <- list(`EpCAM+`=EPCAM_pos,`Any3+`=Any1_pos)

ggvenn(a2,fill_color=c("blue","orangered"), auto_scale=TRUE,show_percentage=FALSE,set_name_size=0,text_size=3.5) +
  annotate("rect",xmin = -1.4, xmax = 1.05, ymin = -1.05, ymax = 1.05, fill = "gray",alpha=0.3) +
  annotate("text",x=-0.83,y=0.1,label="EpCAM+") +
  annotate("text",x=-0.82,y=-0.1,label="(33%)") +
  annotate("text",x=0.18,y=-0.1,label="(48%)") +
  annotate("text",x=0.82,y=0.1,label="Any3+") +
  annotate("text",x=0.82,y=-0.1,label="(6%)") +
  annotate("text",x=0.75,y=-0.77,label="Negative") +
  annotate("text",x=0.75,y=-0.87,label="2541") +
  annotate("text",x=0.75,y=-0.97,label="(13%)")

ggsave((paste0("../../plots/",Sys.Date(),"_venn_all_ggvenn.pdf")))


## plot Figure 2A with two stacked bar (one for SCLC subtyes and the other for DLL3 pos vs neg)

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## subtype

samType <- as.data.frame(table(epi_sub$sample,epi_sub$SCLC_subtype2))
colnames(samType) <- c("Sample","Subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100
samType$fraction_type <- "subtype"
colnames(samType)[2] <- "group"

## DLL3
samType1 <- as.data.frame(table(epi_sub$sample,epi_sub$DLL3_group))
colnames(samType1) <- c("Sample","DLL3","Freq")
samType1$Value <- samType1$Freq / ave(samType1$Freq, samType1$Sample, FUN = sum) * 100

samType1$fraction_type <- "DLL3"
colnames(samType1)[2] <- "group"


df <- rbind(samType,samType1)

reordered_data <- df %>%
  filter(group == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

df$Sample <- factor(df$Sample,levels=reordered_data$Sample)

df$fraction_type <- factor(df$fraction_type,levels=c("subtype","DLL3"))

df$group <- factor(df$group,levels=c("SCLC-A","SCLC-N","SCLC-P","SCLC-I","neg","pos"))

### merge two data.frame

ggplot(df, aes(x = fraction_type, y = Value, fill = group)) + geom_bar(stat = "identity", position = "stack") + facet_grid(~ factor(Sample), switch = "x") + theme(strip.placement = "outside", strip.background = element_rect(fill = NA, color = "white"), panel.spacing = unit(-.01,"cm"), axis.text.x = element_text(angle = 45, hjust = 1)) + labs(x="") +  
  scale_fill_manual(values = c("pos"= "red",
                                "neg" = "blue",
                                "SCLC-A" = "sandybrown", 
                               "SCLC-N" = "#56B4E9", 
                               "SCLC-P" = "forestgreen", 
                               "SCLC-I" = "gray")) + 
  scale_y_continuous(expand = expansion(mult = 0, add = 0)) 


ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_and_subtype_per_sample_orderBySCLCA.pdf")))


### 2E plot violin plots for two primary cohorts


df_mgh <- read.xlsx(paste0("../../tables/2025-07-07_coexpression_summary_at_sample_level.xlsx"))

### load summary table from Cancer Cell data

df_cc <- read.xlsx("../../../SCLC/tables/2025-07-07_coexpression_summary_at_sample_level.xlsx")


df <- rbind(df_mgh[,c("Cohort","Pdll3")],df_cc[,c("Cohort","Pdll3")])

df$Cohort <- ifelse(df$Cohort=="MGH","B","C")

ggplot(df, aes(x = Cohort, y = Pdll3)) +
  geom_violin(position = "dodge", fill = "orange",width = 0.8) + geom_jitter(position = position_jitter(seed = 1, width = 0.1)) +
  theme_bw() +
  theme(panel.border = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.line = element_line(color = "black")) + labs(y="DLL3%")

ggsave(here(paste0("../../plots/",Sys.Date(),"_Violin_DLL3_percent_in_two_cohorts.pdf")))



### 7A for orange-DLL3, blue-SEZ6, and yellow-B7H3

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## 2025-07-16  change colors for orange-DLL3, blue-SEZ6, and yellow-B7H3
## first plot dots with integrated color
epi_sub$DLL3_ <- ifelse(epi_sub$DLL3_group=="neg",0,1)
epi_sub$SEZ6_ <- ifelse(epi_sub$SEZ6_group=="neg",0,1)
epi_sub$B7H3_ <- ifelse(epi_sub$CD276_group=="neg",0,1)

df <- as.data.frame(FetchData(epi_sub,c("umap_1","umap_2","DLL3_","SEZ6_","B7H3_","sample")))

df_grouped <- df %>%
  group_by(DLL3_, SEZ6_, B7H3_) %>%
  mutate(group = cur_group_id()) %>%
  ungroup() %>% as.data.frame()

## unique(df_grouped[,c(3,4,5,7)])
#     DLL3_ SEZ6_ B7H3_ group
# 1       0     0     0     1
# 3       1     0     0     5
# 6       0     0     1     2
# 8       0     1     1     4
# 19      1     0     1     6
# 23      0     1     0     3
# 47      1     1     0     7
# 100     1     1     1     8

## colors

# 
# rgb(1,0.647,0) "#FFA500"  ## DLL3
# rgb(1,1,0)  "#FFFF00" ## B7H3
# rgb(0.39,0.72,1) "#63B8FF" ## SEZ6
# rgb(1,1,1) "#FFFFFF"
# rgb(0,0,0) "#000000"
# rgb(0.5,0.5,0.5) "#808080"
# 
# r_d <- c(1,0.647,0)
# r_b <- c(1,1,0)
# r_s <- c(0.39,0.72,1)

ref_col <- data.frame(group=1:8,col_val=c(rgb(1,1,1), # group 1
                                         rgb(1,1,0), # 2
                                         rgb(0.39,0.72,1), # 3
                                         rgb(0.695, 0.860, 0.500), #4: colMeans(rbind(r_s,r_b))
                                         rgb(1,0.647,0), # 5
                                         rgb(1,0.8235,0), #6
                                         rgb(0.6950, 0.6835, 0.5), #7
                                         rgb(0.7966667, 0.7890000, 0.3333333) #8
                                         ))

ref_col$col_val
#[1] "#808080" "#FFFF00" "#63B8FF" "#B1DB80" "#FFA500" "#FFD200" "#B1AE80" "#CBC955"

df_col <- merge(df_grouped,ref_col)

data_col <- c("1"="#FFFFFF","2"="#FFFF00","3"="#63B8FF","4"="#B1DB80","5"="#FFA500", "6"="#FFD200","7"="#B1AE80","8"= "#CBC955")

## plot


## custom legend with three genes

legend_df <- data.frame(
  x = 13,
  y = c(2, 0,-2),
  label = c('DLL3', 'SEZ6', 'B7H3'),
  color=c("#FFA500", "#63B8FF","#FFFF00") # The desired colors for the legend
)

### get label for samples

## add text labels for samples/clusters

umap_coords <- Embeddings(epi_sub, "umap") %>% as.data.frame()

umap_coords$sample <- epi_sub$sample

## use mean

cluster_centers <- umap_coords %>%
  dplyr::group_by(sample) %>%
  dplyr::summarise(
    umap_1 = mean(umap_1),
    umap_2 = mean(umap_2)
  )

as.data.frame(cluster_centers)
#    sample      umap_1      umap_2
# 1    LP19  -1.6817195  10.3005750
# 2    LP20 -11.0616706  -0.3713637
# 3    LP21  -0.8203757 -11.9230559
# 4    LP28   5.6626382  -7.8014069
# 5    LP33  -5.5918049  -6.6694991
# 6    LP34   9.6479422  -0.7355175
# 7    LP35   2.8017295  -4.5473336
# 8    LP71   2.9563915   4.8297233
# 9    LP85  -2.2676736   0.5394287
# 10   LP91  -1.1810069  -4.7434572

## mean performs similar to DimPlot but still need to adjust locations


cluster_centers2 <- as.data.frame(cluster_centers)


cluster_centers2[cluster_centers2$sample=="LP19",2:3] <- c(-1.68,15)
cluster_centers2[cluster_centers2$sample=="LP20",2:3] <- c(-11,4)
cluster_centers2[cluster_centers2$sample=="LP21",2:3] <- c(-0.8,-15)
cluster_centers2[cluster_centers2$sample=="LP28",2:3] <- c(5.67,-10)
cluster_centers2[cluster_centers2$sample=="LP33",2:3] <- c(-5.6,-5)
cluster_centers2[cluster_centers2$sample=="LP34",2:3] <- c(8,2)
cluster_centers2[cluster_centers2$sample=="LP35",2:3] <- c(2.8,-7.5)
cluster_centers2[cluster_centers2$sample=="LP71",2:3] <- c(3,8)
cluster_centers2[cluster_centers2$sample=="LP85",2:3] <- c(-5.2,0.54)
cluster_centers2[cluster_centers2$sample=="LP91",2:3] <- c(-1.2,-7.4)


p <- ggplot(df_col,aes(x=umap_1,y=umap_2)) + geom_point(aes(color=factor(group)),alpha=0.8,size=0.3,stroke=0.1) + 
  scale_colour_manual(values=data_col,guide="none") + theme_minimal() +
  theme(panel.border = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.line = element_line(color = "black"))  

 
p+
  geom_point(data = legend_df,
             aes(x = x, y = y, color = label ),
             size = 2) +
  geom_text(data = legend_df,
            aes(x = x + 0.3, y = y, label = label),
            hjust = 0,size=2.5) +# get all elements but need to change the color for new legend at this layer
  scale_color_manual(name="",values = c(data_col,setNames(legend_df$color, legend_df$label)) ,guide="none")  +
   annotate("text",x=cluster_centers2$umap_1,y=cluster_centers2$umap_2,label=cluster_centers2$sample) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_SEZ6_DLL3_B7H3_binary_tumorOnly_withLabel_newColor.pdf")))

## add three legends
FeaturePlot(epi_sub,features=c("DLL3","SEZ6","CD276"))
DimPlot(epi_sub, group.by="sample", label=TRUE) + NoLegend()

## try different color II: 

ref_col <- data.frame(group=1:8,col_val=c(rgb(1,1,1), # group 1
                                         rgb(0,0.76,0.6), # 2
                                         rgb(0.39,0.72,1), # 3
                                         rgb(0.195, 0.74, 0.8), #4: colMeans(rbind(r_s,r_b))
                                         rgb(1,0.647,0), # 5
                                         rgb(0.5,0.7035,0.3), #6
                                         rgb(0.6950, 0.6835, 0.5), #7
                                         rgb(0.4633, 0.709, 0.533) #8
                                         ))

df_col <- merge(df_grouped,ref_col)

data_col <- c("1"="#FFFFFF","2"="#00C299","3"="#63B8FF","4"="#32BDCC","5"="#FFA500", "6"="#80B34D","7"="#B1AE80","8"= "#76B588")


legend_df <- data.frame(
  x = 11,
  y = c(10, 9,8),
  label = c('DLL3', 'SEZ6', 'B7H3'),
  color=c("#FFA500", "#63B8FF","#00C299") # The desired colors for the legend
)

p <- ggplot(df_col,aes(x=umap_1,y=umap_2)) + geom_point(aes(color=factor(group)),alpha=0.8,size=0.3,stroke=0.1) + 
  scale_colour_manual(values=data_col,guide="none") + theme_minimal() +
  theme(panel.border = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.line = element_line(color = "black"))  

 
p+
  geom_point(data = legend_df,
             aes(x = x, y = y, color = label ),
             size = 2) +
  geom_text(data = legend_df,
            aes(x = x + 0.3, y = y, label = label),
            hjust = 0,size=2.5) +# get all elements but need to change the color for new legend at this layer
  scale_color_manual(name="",values = c(data_col,setNames(legend_df$color, legend_df$label)) ,guide="none")  +
   annotate("text",x=cluster_centers2$umap_1,y=cluster_centers2$umap_2,label=cluster_centers2$sample) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_SEZ6_DLL3_B7H3_binary_tumorOnly_withLabel_newColor_2.pdf")))


### try different color V: 

ref_col <- data.frame(group=1:8,col_val=c(rgb(1,1,1), # group 1
                                         rgb(0.62,0.35,0.68), # 2
                                         rgb(0.39,0.72,1), # 3
                                         rgb(0.505, 0.535, 0.84), #4: colMeans(rbind(r_s,r_b))
                                         rgb(1,0.647,0), # 5
                                         rgb(0.81, 0.4985, 0.34), #6
                                         rgb(0.6950, 0.6835, 0.5), #7
                                         rgb(0.67, 0.57, 0.56) #8
                                         ))

df_col <- merge(df_grouped,ref_col)

data_col <- c("1"="#FFFFFF","2"="#9E59AD","3"="#63B8FF","4"="#8188D6","5"="#FFA500", "6"="#CF7F57","7"="#B1AE80","8"= "#AB918F")



legend_df <- data.frame(
  x = 11,
  y = c(10, 8,6),
  label = c('DLL3', 'SEZ6', 'B7H3'),
  color=c("#FFA500", "#63B8FF","#9E59AD") # The desired colors for the legend
)

p <- ggplot(df_col,aes(x=umap_1,y=umap_2)) + geom_point(aes(color=factor(group)),size=0.5,stroke=0.1) + 
  scale_colour_manual(values=data_col,guide="none") + theme_minimal() +
  theme(panel.border = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.line = element_line(color = "black"))  

 
p+
  geom_point(data = legend_df,
             aes(x = x, y = y, color = label ),
             size = 3) +
  geom_text(data = legend_df,
            aes(x = x + 0.3, y = y+0.1, label = label),
            hjust = 0,size=3) +# get all elements but need to change the color for new legend at this layer
  scale_color_manual(name="",values = c(data_col,setNames(legend_df$color, legend_df$label)) ,guide="none")  +
   annotate("text",x=cluster_centers2$umap_1,y=cluster_centers2$umap_2,label=cluster_centers2$sample) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_SEZ6_DLL3_B7H3_binary_tumorOnly_withLabel_newColor_5.pdf")))

### 2C specify y scale and increase dot size

## load MGH data
df_MGH <- read.xlsx(paste0("../../tables/2025-07-07_coexpression_summary_at_sample_level.xlsx"))

#df_cc <- read.xlsx("../../../SCLC/tables/2025-07-07_coexpression_summary_at_sample_level.xlsx")
df_MGH$Pdll3 <- df_MGH$Pdll3*100
ggboxplot(df_MGH, x = "Treatment", y = "Pdll3", color = "Treatment", palette=c("blue","red"),
          add = "jitter",add.params = list(size = 2.5,jitter=0.3)) + 
          labs(x="",y="%DLL3+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90)  +
          scale_y_continuous(breaks=c(0,25,50,75,100), limits=c(0,100),expand = expansion(mult = c(0, 0)))

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_per_sample_by_stage.pdf")))


```

## Step 22: pseudobulk comparison

```{r}
## load all tumors
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(epi_sub, assays = "RNA", return.seurat=T,group.by = c("sample"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("sample","DLL3","SEZ6","CD276"))

## add treatment info

dtreat <- as.data.frame(unclass(table(epi_sub$sample,epi_sub$treatment)))

dtreat$treatment <- ifelse(dtreat$naive >0, "naive","treated")

dtreat$sample <- rownames(dtreat)

dtreat <- dtreat[,c("sample","treatment")]

## merge

df <- merge(df,dtreat)

### compare DLL3

ggboxplot(df, x = "treatment", y = "DLL3",
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average DLL3") + theme(legend.position = "right") +
          stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_pseudobulk_by_treatment_allCell.pdf")))

### compare SEZ6

ggboxplot(df, x = "treatment", y = "SEZ6",
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average SEZ6") + theme(legend.position = "right") +
          stat_compare_means(label.x=1.4, label.y= 0.6) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_pseudobulk_by_treatment_allCell.pdf")))

### compare B7H3

ggboxplot(df, x = "treatment", y = "CD276",
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average CD276") + theme(legend.position = "right") +
          stat_compare_means(label.x=1.4) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_B7H3_pseudobulk_by_treatment_allCell.pdf")))


#### positive cells only ###

## DLL3
  
tmp <- subset(epi_sub,subset=DLL3_group=="pos")

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(tmp, assays = "RNA", return.seurat=T,group.by = c("sample"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("sample","DLL3"))

## merge

df <- merge(df,dtreat)

### compare target genes

ggboxplot(df, x = "treatment", y = "DLL3",
        color = "treatment", palette = c("blue","red"),
        add = "jitter") + labs(x="",y=paste0("log-normalized average ","DLL3 (positive only)")) + theme(legend.position = "right") +
        stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_","DLL3","_pseudobulk_by_treatment_posCellOnly.pdf")))



## SEZ6
  
tmp <- subset(epi_sub,subset=SEZ6_group=="pos")

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(tmp, assays = "RNA", return.seurat=T,group.by = c("sample"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("sample","SEZ6"))

## merge

df <- merge(df,dtreat)

### compare target genes

ggboxplot(df, x = "treatment", y = "SEZ6",
        color = "treatment", palette = c("blue","red"),
        add = "jitter") + labs(x="",y=paste0("log-normalized average ","SEZ6 (positive only)")) + theme(legend.position = "right") +
        stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_","SEZ6","_pseudobulk_by_treatment_posCellOnly.pdf")))

###

## CD276
  
tmp <- subset(epi_sub,subset=CD276_group=="pos")

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(tmp, assays = "RNA", return.seurat=T,group.by = c("sample"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("sample","CD276"))

## merge

df <- merge(df,dtreat)

### compare target genes

ggboxplot(df, x = "treatment", y = "CD276",
        color = "treatment", palette = c("blue","red"),
        add = "jitter") + labs(x="",y=paste0("log-normalized average ","CD276 (positive only)")) + theme(legend.position = "right") +
        stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_","CD276","_pseudobulk_by_treatment_posCellOnly.pdf")))

## no significant in any condition


```

## Step 23: figures for coexpression

```{r}

df_mgh <- read.xlsx(paste0("../../tables/2025-07-08_coexpression_summary_at_sample_level.xlsx"))

## plot two stacked bar side by side
df_mgh$`SEZ6+` <- df_mgh$Psez6
df_mgh$`SEZ6-` <- 1-df_mgh$Psez6

df_mgh$`B7H3+` <- df_mgh$Pb7h3
df_mgh$`B7H3-` <- 1-df_mgh$Pb7h3


df <- df_mgh[,c(1,(ncol(df_mgh)-3):ncol(df_mgh))]

## convert to long

long_df <- melt(df, id.vars = c("Sample"), 
                variable.name = "group", 
                value.name = "fraction")

#epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))


# reordered_data <- long_df %>%
#   filter(group == "SEZ6+") %>%
#   arrange(fraction) %>%
#   mutate(Sample = factor(Sample, levels = Sample))
# 
# df_long$Sample <- factor(df$Sample,levels=reordered_data$Sample)
# 
# df$fraction_type <- factor(df$fraction_type,levels=c("subtype","DLL3"))

long_df$group <- factor(long_df$group,levels=c("SEZ6-","SEZ6+","B7H3-","B7H3+"))
long_df$gene <- ifelse(startsWith(as.character(long_df$group),"SEZ6"),"SEZ6","B7H3")
long_df$gene <- factor(long_df$gene,levels=c("SEZ6","B7H3"))
### merge two data.frame

ggplot(long_df, aes(x = gene, y = fraction, fill = group)) + geom_bar(stat = "identity", position = "stack") + facet_grid(~ Sample) + theme(strip.placement = "outside", strip.background = element_rect(fill = NA, color = "white"), panel.spacing = unit(-.01,"cm"), axis.text.x = element_text(angle = 45, hjust = 1)) + labs(x="") +  
  scale_fill_manual(values = c("SEZ6-"= "springgreen3",
                                "SEZ6+" = "#63B8FF",
                               "B7H3-" = "forestgreen", 
                               "B7H3+" = "#9E59AD")) + 
  scale_y_continuous(expand = expansion(mult = 0, add = 0)) 



ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_B7H3_per_sample_orderBySCLCA.pdf")))



### Venn diagram for pooled data

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

DLL3_pos <- WhichCells(epi_sub, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(epi_sub, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(epi_sub, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

## use euler 
p <- plot(euler(a, shape = "circle"), quantities = list(type = c("counts", "percent")),fill = list(fill = c("#FFA500", "#63B8FF","#9E59AD"))) 
g <- grid.arrange(grobs=list(p),top=textGrob("Cohort B", gp = gpar(fontface = "bold")))
ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_bothCP_differentSize.pdf")),g)


### for each individual
plot.list <- list()
for (samp in unique(epi_sub$sample)){
  epi_tmp <- subset(epi_sub,subset=sample==samp)
  SEZ6_pos <- WhichCells(epi_tmp, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(epi_tmp, expression=CD276_group=="pos")
  DLL3_pos <- WhichCells(epi_tmp, expression=DLL3_group=="pos") 
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
  p <- plot(euler(a, shape = "circle"), quantities = list(type = c("counts", "percent"), cex=0.5),labels = list(cex = 0.6),fill = list(fill = c("#FFA500", "#63B8FF","#9E59AD"))) 
  g <- grid.arrange(grobs=list(p),top=textGrob(samp, gp = gpar(fontsize = 8,fontface = "bold")))

  plot.list[[samp]] <- g
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",samp,".pdf")),g)
    
}

## plot all in one 
ggarrange(plotlist = plot.list, ncol = 4, nrow = 3)

ggsave(here(paste0("../../plots/",Sys.Date(),"_three_target_by_individual_allIn1.pdf")))


```

## Step 24: GSEA using known NE signatures

```{r}
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

data1 <- FetchData(epi_sub, vars = c("DLL3","SEZ6","CD276"))


##set default as triple negative
epi_sub$group_gsea <- "triple_negative"

### for DLL3+ only

dll3.posOnly <- WhichCells(epi_sub, expression=DLL3_group=="pos" & SEZ6_group=="neg" & CD276_group=="neg")
epi_sub$group_gsea[dll3.posOnly] <- "DLL3+Only"

### for SEZ6+ only

sez6.posOnly <- WhichCells(epi_sub, expression=DLL3_group=="neg" & SEZ6_group=="pos" & CD276_group=="neg")
epi_sub$group_gsea[sez6.posOnly] <- "SEZ6+Only"

### for CD276+ only
cd276.posOnly <- WhichCells(epi_sub, expression=DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="pos")
epi_sub$group_gsea[cd276.posOnly] <- "B7H3+Only"

## for coexpression of two or three

epi_sub$group_gsea[which(rowSums(data1>0)>1)] <-"2or3+"

## get NE signatures

### NE Alshalafa

NE_Alshalalfa <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Alshalalfa",colNames = FALSE)

NE_Alshalalfa_All <- NE_Alshalalfa$X1

NE_Alshalalfa_UP <- NE_Alshalalfa$X1[which(NE_Alshalalfa$X2=="UP")]

NE_Alshalalfa_DOWN <- NE_Alshalalfa$X1[which(NE_Alshalalfa$X2=="DOWN")]

NE_Alshalalfa <- list(Alshalalfa_All=NE_Alshalalfa_All,Alshalalfa_UP=NE_Alshalalfa_UP,Alshalalfa_DOWN=NE_Alshalalfa_DOWN)


## NE Balanis

NE_Balanis <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Balanis",colNames = FALSE)[,1]


## NE Beltran

NE_Beltran <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Beltran",colNames = FALSE)

NE_Beltran_All <- NE_Beltran$X1

NE_Beltran_UP <- NE_Beltran$X1[which(NE_Beltran$X2=="UP")]

NE_Beltran_DOWN <- NE_Beltran$X1[which(NE_Beltran$X2=="DOWN")]



## NE Kumar

NE_Kumar <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Kumar",colNames = FALSE)[,1]


## NE Tsai

NE_Tsai69 <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Tsai69",colNames = FALSE)

NE_Tsai306 <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Tsai306",colNames = FALSE)

NE_Tsai <- rbind(NE_Tsai69,NE_Tsai306)

NE_Tsai_All <- NE_Tsai$X1

NE_Tsai_UP <- NE_Tsai$X1[which(NE_Tsai$X2=="UP")]

NE_Tsai_DOWN <- NE_Tsai$X1[which(NE_Tsai$X2=="DOWN")]


### put all list together

NE <- list(Alshalalfa_All=NE_Alshalalfa_All,Alshalalfa_UP=NE_Alshalalfa_UP,Alshalalfa_DOWN=NE_Alshalalfa_DOWN,
           Balanis=NE_Balanis,
           Beltran_All=NE_Beltran_All,Beltran_UP=NE_Beltran_UP,Beltran_DOWN=NE_Beltran_DOWN,
           Kumar=NE_Kumar,
           Tsai_All=NE_Tsai_All,Tsai_UP=NE_Tsai_UP,Tsai_DOWN=NE_Tsai_DOWN)

### save NE signatures with the object name of NE
save(NE, file="/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/NE.pathway.list.RData")


epi_sub <- AddModuleScore_UCell(epi_sub, features = NE, ncores=4, name = "")

epi_sub$group_gsea <- factor(epi_sub$group_gsea,levels=c("DLL3+Only","SEZ6+Only","B7H3+Only","2or3+","triple_negative"))

VlnPlot(epi_sub, features=names(NE), group.by="group_gsea",pt.size = 0, ncol=4) &
  theme(
    axis.text.x = element_text(size = 6), 
    axis.text.y = element_text(size = 6), 
    axis.title.x = element_blank(),
    plot.title = element_text(size = 8, face = "bold", hjust = 0.5) 
  )

ggsave((paste0("../../plots/",Sys.Date(),"_NE_signature_GSEA1.pdf")))


## GSEA II

data1 <- FetchData(epi_sub, vars = c("DLL3","EPCAM"))

epi_sub$epcam_gsea <- "otherTumorCell"

epi_sub$epcam_gsea[which(data1$DLL3 >0 & data1$EPCAM==0)] <- "DLL3+Only"

epi_sub$epcam_gsea[which(data1$DLL3 ==0 & data1$EPCAM>0)] <- "EPCAM+Only"

epi_sub$epcam_gsea[which(data1$DLL3 >0 & data1$EPCAM>0)] <- "EPCAM&DLL3+"


epi_sub$epcam_gsea <- factor(epi_sub$epcam_gsea,levels=c("EPCAM+Only","DLL3+Only","EPCAM&DLL3+","otherTumorCell"))

VlnPlot(epi_sub, features=names(NE), group.by="epcam_gsea",pt.size = 0, ncol=4) &
  theme(
    axis.text.x = element_text(size = 6), 
    axis.text.y = element_text(size = 6), 
    axis.title.x = element_blank(),
    plot.title = element_text(size = 8, face = "bold", hjust = 0.5) 
  )

ggsave((paste0("../../plots/",Sys.Date(),"_NE_signature_GSEA2_EPCAM.pdf")))


```

##Step 25: revision

```{r}
### create a folder for revision

dir.create("../../plots/revision")

## load data
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

### plot DLL3 expression across four groups with p values

## all cells 

mlevels <- c("SCLC-A","SCLC-N","SCLC-P","SCLC-I")
epi_sub$SCLC_subtype2 <- factor(epi_sub$SCLC_subtype2,levels=mlevels)

VlnPlot2(
  epi_sub,
  features = "DLL3",
  group.by = "SCLC_subtype2",
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue", "gray"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_allCells.pdf")))

### DLL3 positive cells only

cells <- WhichCells(epi_sub, expression=DLL3_group=="pos")

VlnPlot2(
  epi_sub,
  features = "DLL3",
  group.by = "SCLC_subtype2",
  cells=cells,
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue", "gray"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_posCells.pdf")))


### split DLL3+ cells into low-medium-high categories

## get gene expression of DLL3+ cells

expr <- FetchData(epi_sub, vars = "DLL3")[,1]
expr_pos <- expr[expr > 0]
q25_pos <- quantile(expr_pos, 0.25)
q75_pos <- quantile(expr_pos, 0.75)

epi_sub$DLL3_pos_subgroup <- "neg"

epi_sub$DLL3_pos_subgroup[expr>0] <- cut(expr[expr>0], breaks=c(min(expr_pos), q25_pos, q75_pos, max(expr_pos)),include.lowest=TRUE,labels=c("Low","Medium","High"))

epi_sub$DLL3_pos_subgroup <- factor(epi_sub$DLL3_pos_subgroup,
                        levels = c("1", "2", "3","neg"),
                        labels = c("Low", "Medium", "High","neg"))

### DLL3 positive cells only

## Low
cells <- WhichCells(epi_sub, expression=(DLL3_pos_subgroup=="Low"))

VlnPlot2(
  epi_sub,
  features = "DLL3",
  group.by = "SCLC_subtype2",
  cells=cells,
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue", "gray"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_posCellsLow.pdf")))


## Medium
cells <- WhichCells(epi_sub, expression=(DLL3_pos_subgroup=="Medium"))

VlnPlot2(
  epi_sub,
  features = "DLL3",
  group.by = "SCLC_subtype2",
  cells=cells,
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue", "gray"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_posCellsMed.pdf")))


## High

cells <- WhichCells(epi_sub, expression=(DLL3_pos_subgroup=="High"))

VlnPlot2(
  epi_sub,
  features = "DLL3",
  group.by = "SCLC_subtype2",
  cells=cells,
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue", "gray"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_posCellsHigh.pdf")))


### DLL3&ASCL1 correlation

FeatureScatter(epi_sub,feature1="ASCL1",feature2="DLL3",group.by="DLL3_pos_subgroup") + labs(color = "DLL3")

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3_and_ASCL1.pdf")))


### only positive

cells <- WhichCells(epi_sub, expression=(DLL3>0 & ASCL1>0))

FeatureScatter(epi_sub,feature1="ASCL1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3_and_ASCL1_bothPos.pdf")))

### only positive Low DLL3

cells <- WhichCells(epi_sub, expression=(DLL3>0 & ASCL1>0 & DLL3_pos_subgroup=="Low"))

FeatureScatter(epi_sub,feature1="ASCL1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3Low_and_ASCL1_bothPos.pdf")))


### only positive Medium DLL3

cells <- WhichCells(epi_sub, expression=(DLL3>0 & ASCL1>0 & DLL3_pos_subgroup=="Medium"))

FeatureScatter(epi_sub,feature1="ASCL1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3Med_and_ASCL1_bothPos.pdf")))

### only positive High DLL3

cells <- WhichCells(epi_sub, expression=(DLL3>0 & ASCL1>0 & DLL3_pos_subgroup=="High"))

FeatureScatter(epi_sub,feature1="ASCL1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3High_and_ASCL1_bothPos.pdf")))


### DLL3&NEUROD1 correlation

FeatureScatter(epi_sub,feature1="NEUROD1",feature2="DLL3",group.by="DLL3_pos_subgroup") + labs(color = "DLL3")

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3_and_NEUROD1.pdf")))


### only positive

cells <- WhichCells(epi_sub, expression=(DLL3>0 & NEUROD1>0))

FeatureScatter(epi_sub,feature1="NEUROD1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3_and_NEUROD1_bothPos.pdf")))

### only positive Low DLL3

cells <- WhichCells(epi_sub, expression=(DLL3>0 & NEUROD1>0 & DLL3_pos_subgroup=="Low"))

FeatureScatter(epi_sub,feature1="NEUROD1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3Low_and_NEUROD1_bothPos.pdf")))


### only positive Medium DLL3

cells <- WhichCells(epi_sub, expression=(DLL3>0 & NEUROD1>0 & DLL3_pos_subgroup=="Medium"))

FeatureScatter(epi_sub,feature1="NEUROD1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3Med_and_NEUROD1_bothPos.pdf")))

### only positive High DLL3

cells <- WhichCells(epi_sub, expression=(DLL3>0 & NEUROD1>0 & DLL3_pos_subgroup=="High"))

FeatureScatter(epi_sub,feature1="NEUROD1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_scatterplot_DLL3High_and_NEUROD1_bothPos.pdf")))


### get correlation for different categories


df_st <- data.frame(unclass(table(epi_sub$sample, epi_sub$SCLC_subtype2)/rowSums(table(epi_sub$sample, epi_sub$SCLC_subtype2))))
## order 
colnames(df_st) <- gsub("\\.","_",colnames(df_st))

df_st$sample <- rownames(df_st)


## get fractions of DLL3 positive cell per sample by three levels: low-med-high

df_dl <- data.frame(unclass(table(epi_sub$sample, epi_sub$DLL3_pos_subgroup)/rowSums(table(epi_sub$sample, epi_sub$DLL3_pos_subgroup))))

df_dl$sample <- rownames(df_dl)


## join

df <- merge(df_st,df_dl)

df$MedandHigh <- df$Medium+df$High

# List of x variables
x_vars <- c("SCLC_A", "SCLC_N", "SCLC_P", "SCLC_I")
y_vars <- c("Low","Medium","High","MedandHigh")

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  for (y in y_vars){
    sp <- ggscatter(df, x = x, y = y,
                    add = "reg.line",
                    add.params = list(color = "blue", fill = "lightgray"),
                    conf.int = TRUE
    ) + xlab(paste0("%", x)) + ylab(paste0("%DLL3+(",y,")")) + theme(text = element_text(size = 7))
    
    # Add correlation
    p <- sp + stat_cor(method = "pearson",size=3)
    
    # Store plot in list with name
    plot_list[[paste0(x,"-",y)]] <- p
  }
}
 
ggarrange(plotlist = plot_list, ncol = 4, nrow = 4)

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_pearson_corr_DLL3_by_subtype.pdf")))

### use spearman 

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  for (y in y_vars){
    sp <- ggscatter(df, x = x, y = y,
                    add = "reg.line",
                    add.params = list(color = "blue", fill = "lightgray"),
                    conf.int = TRUE
    ) + xlab(paste0("%", x)) + ylab(paste0("%DLL3+(",y,")"))+ theme(text = element_text(size = 7))
    
    # Add correlation
    p <- sp + stat_cor(method = "spearman",size=3)
    
    # Store plot in list with name
    plot_list[[paste0(x,"-",y)]] <- p
  }
}
 
ggarrange(plotlist = plot_list, ncol = 4, nrow = 4)

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_spearman_corr_DLL3_by_subtype.pdf")))


### check ASCL1, NEUROD1 and POU2F3 expression in subtypes to validate subtype annotation based on signatures
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))
## violin plot enhanced vlnplot2 in SeuratExtend

mlevels <- c("SCLC-A","SCLC-N","SCLC-P","SCLC-I")
epi_sub$SCLC_subtype2 <- factor(epi_sub$SCLC_subtype2,levels=mlevels)
## SCLC-A  
VlnPlot2(
  epi_sub,
  features = "ASCL1",
  group.by = "SCLC_subtype2",
  pt=F,
  width=0.5,
  cols=c("royalblue2", "#00B0F6","skyblue", "gray"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violinplot_by_ASCL1_by_subtype_withBox.pdf")))

## SCLC-N
VlnPlot2(
  epi_sub,
  features = "NEUROD1",
  group.by = "SCLC_subtype2",
  pt=F,
  width=0.5,
  cols=c("royalblue2", "#00B0F6","skyblue", "gray"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violinplot_by_NEUROD1_by_subtype_withBox.pdf")))

## SCLC-P

VlnPlot2(
  epi_sub,
  features = "POU2F3",
  group.by = "SCLC_subtype2",
  pt=F,
  width=0.5,
  cols=c("royalblue2", "#00B0F6","skyblue", "gray"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violinplot_by_POU2F3_by_subtype_withBox.pdf")))



### let's check low-med-high vs treatment

dtreat <- as.data.frame(unclass(table(epi_sub$sample,epi_sub$treatment)))

dtreat$treatment <- ifelse(dtreat$naive >0, "naive","treated")

dtreat$sample <- rownames(dtreat)

dtreat <- dtreat[,c("sample","treatment")]

dfinal <- merge(df,dtreat)

write.xlsx(dfinal,paste0("../../plots/revision/",Sys.Date(),"_DLL3_3tier_subtype_treatment_summary.xlsx"))


### compare naive vs treated for each tier

# List of x variables
y_vars <- c("Low","Medium","High","MedandHigh")

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (y in y_vars){
  p<- ggboxplot(dfinal, x = "treatment", y = y,
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y=paste0("%DLL3+(",y,")")) + theme(legend.position = "none") +
          stat_compare_means(label.x.npc="middle") 
  
  # Store plot in list with name
  plot_list[[y]] <- p
}
 
ggarrange(plotlist = plot_list, ncol = 2, nrow = 2)

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_boxplot_DLL3_3tier_by_treatment.pdf")))

saveRDS(epi_sub,paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))


###heterogeneity among the different keratin genes in the 10X RNA seq of primary SCLC tumors?


epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

FeaturePlot(epi_sub,features=c("KRT8","KRT18","KRT19","EPCAM"))

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_EPCAM_CK.pdf")))



UMAPPlot(epi_sub,group.by="sample", label=TRUE)
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_by_sampleID.pdf")))

## EPCAM/CK pos vs neg

expr_matrix <- FetchData(epi_sub, vars = c("KRT8","KRT18","KRT19","EPCAM"))
epi_sub$EPCAMCK_positive <- rowSums(expr_matrix > 0) > 0

UMAPPlot(epi_sub, group.by="EPCAMCK_positive",cols=c("blue","red"))
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_any_EPCAM_CK_positive.pdf")))


### plot CK, EpCAM and DLL3 at the same plot

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## 2025-10-27  plot orange-DLL3, blue-CK, and yellow-EpCAM on the same plot

## assign EPCAM pos/neg group
epi_sub$EPCAM_group <- "neg"
 
epi_sub$EPCAM_group[which(epi_sub[['RNA']]$data["EPCAM",] > 0)] <- "pos" 
epi_sub$EPCAM_group <- factor(epi_sub$EPCAM_group)


## define CK +/-
epi_sub$CK_group <- "neg"
expr_matrix <- FetchData(epi_sub, vars = c("KRT8","KRT18","KRT19"))
epi_sub$CK_group[which(rowSums(expr_matrix > 0) > 0)] <- "pos"
epi_sub$CK_group <- factor(epi_sub$CK_group)

## first plot dots with integrated color
epi_sub$DLL3_ <- ifelse(epi_sub$DLL3_group=="neg",0,1)
epi_sub$EPCAM_ <- ifelse(epi_sub$EPCAM_group=="neg",0,1)
epi_sub$CK_ <- ifelse(epi_sub$CK_group=="neg",0,1)

df <- as.data.frame(FetchData(epi_sub,c("umap_1","umap_2","DLL3_","EPCAM_","CK_","sample")))

df_grouped <- df %>%
  group_by(DLL3_, EPCAM_, CK_) %>%
  mutate(group = cur_group_id()) %>%
  ungroup() %>% as.data.frame()

## unique(df_grouped[,c(3,4,5,7)])
#     DLL3_ EPCAM_ CK_ group
# 1       0     0     0     1
# 3       1     0     0     5
# 6       0     0     1     2
# 8       0     1     1     4
# 19      1     0     1     6
# 23      0     1     0     3
# 47      1     1     0     7
# 100     1     1     1     8

## colors


#rgb(1,0.647,0) "#FFA500"  ## DLL3
#rgb(1,1,0)  "#FFFF00" ## CK
#rgb(0.39,0.72,1) "#63B8FF" ## EPCAM
#rgb(1,1,1) "#FFFFFF"
#rgb(0,0,0) "#000000"
#rgb(0.5,0.5,0.5) "#808080"

#r_d <- c(1,0.647,0)
#r_b <- c(1,1,0)
#r_s <- c(0.39,0.72,1)


## add text labels for samples/clusters

umap_coords <- Embeddings(epi_sub, "umap") %>% as.data.frame()

umap_coords$sample <- epi_sub$sample

## use mean

cluster_centers <- umap_coords %>%
  dplyr::group_by(sample) %>%
  dplyr::summarise(
    umap_1 = mean(umap_1),
    umap_2 = mean(umap_2)
  )

as.data.frame(cluster_centers)
#    sample      umap_1      umap_2
# 1    LP19  -1.6817195  10.3005750
# 2    LP20 -11.0616706  -0.3713637
# 3    LP21  -0.8203757 -11.9230559
# 4    LP28   5.6626382  -7.8014069
# 5    LP33  -5.5918049  -6.6694991
# 6    LP34   9.6479422  -0.7355175
# 7    LP35   2.8017295  -4.5473336
# 8    LP71   2.9563915   4.8297233
# 9    LP85  -2.2676736   0.5394287
# 10   LP91  -1.1810069  -4.7434572

## mean performs similar to DimPlot but still need to adjust locations


cluster_centers2 <- as.data.frame(cluster_centers)


cluster_centers2[cluster_centers2$sample=="LP19",2:3] <- c(-1.68,15)
cluster_centers2[cluster_centers2$sample=="LP20",2:3] <- c(-11,4)
cluster_centers2[cluster_centers2$sample=="LP21",2:3] <- c(-0.8,-15)
cluster_centers2[cluster_centers2$sample=="LP28",2:3] <- c(5.67,-10)
cluster_centers2[cluster_centers2$sample=="LP33",2:3] <- c(-5.6,-5)
cluster_centers2[cluster_centers2$sample=="LP34",2:3] <- c(8,2)
cluster_centers2[cluster_centers2$sample=="LP35",2:3] <- c(2.8,-7.5)
cluster_centers2[cluster_centers2$sample=="LP71",2:3] <- c(3,8)
cluster_centers2[cluster_centers2$sample=="LP85",2:3] <- c(-5.2,0.54)
cluster_centers2[cluster_centers2$sample=="LP91",2:3] <- c(-1.2,-7.4)


### try different color V: 

r_d <- c(1,0.647,0) #"#FFA500"
r_c <- c(0.62,0.35,0.68) # "#9E59AD"
r_e <- c(0.39,0.72,1) #"#63B8FF"


ref_col <- data.frame(group=1:8,col_val=c(rgb(1,1,1), # group 1
                                         rgb(0.62,0.35,0.68), # 2
                                         rgb(0.39,0.72,1), # 3
                                         rgb(0.505, 0.535, 0.84), #4: colMeans(rbind(r_s,r_b))
                                         rgb(1,0.647,0), # 5
                                         rgb(0.81, 0.4985, 0.34), #6
                                         rgb(0.6950, 0.6835, 0.5), #7
                                         rgb(0.67, 0.57, 0.56) #8
                                         ))

df_col <- merge(df_grouped,ref_col)

data_col <- c("1"="#FFFFFF","2"="#9E59AD","3"="#63B8FF","4"="#8188D6","5"="#FFA500", "6"="#CF7F57","7"="#B1AE80","8"= "#AB918F")



legend_df <- data.frame(
  x = 11,
  y = c(10, 8,6),
  label = c('DLL3', 'EPCAM', 'CK'),
  color=c("#FFA500", "#63B8FF","#9E59AD") # The desired colors for the legend
)

p <- ggplot(df_col,aes(x=umap_1,y=umap_2)) + geom_point(aes(color=factor(group)),size=0.5,stroke=0.1) + 
  scale_colour_manual(values=data_col,guide="none") + theme_minimal() +
  theme(panel.border = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.line = element_line(color = "black"))  

 
p+
  geom_point(data = legend_df,
             aes(x = x, y = y, color = label ),
             size = 3) +
  geom_text(data = legend_df,
            aes(x = x + 0.3, y = y+0.1, label = label),
            hjust = 0,size=3) +# get all elements but need to change the color for new legend at this layer
  scale_color_manual(name="",values = c(data_col,setNames(legend_df$color, legend_df$label)) ,guide="none")  +
   annotate("text",x=cluster_centers2$umap_1,y=cluster_centers2$umap_2,label=cluster_centers2$sample) 

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_EPCAM_DLL3_CK_binary_tumorOnly_withLabel_newColor_5.pdf")))

### draw Venn diagram for these three gene or sets as well

DLL3_pos <- WhichCells(epi_sub, expression=DLL3_group=="pos")
EPCAM_pos <- WhichCells(epi_sub, expression=EPCAM_group=="pos")
CK_pos <- WhichCells(epi_sub, expression=CK_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`EPCAM+`=EPCAM_pos,`CK+`=CK_pos)

## use euler 
p <- plot(euler(a, shape = "ellipse"), quantities = list(type = c("counts", "percent")),fill = list(fill = c("#FFA500", "#63B8FF","#9E59AD"))) 
g <- grid.arrange(grobs=list(p),top=textGrob("Cohort B", gp = gpar(fontface = "bold")))
ggsave((paste0("../../plots/revision/",Sys.Date(),"_venn_EPCAM_DLL3_CK_euler_bothCP_differentSize.pdf")),g)

## fix the 99% issue instead 100% as the sum (might be easier to fix it in illustrator)

library(VennDiagram)
venn.diagram(
x = list(`DLL3+`=DLL3_pos,`EPCAM+`=EPCAM_pos,`CK+`=CK_pos),
category.names = c("DLL3+", "EPCAM+", "CK+"),
col = "transparent",
fill = c("#FFA500", "#63B8FF","#9E59AD"),
alpha = 0.30,
print.mode=c("raw","percent"),
filename = paste0("../../plots/revision/",Sys.Date(),"_venn_EPCAM_DLL3_CK_euler_bothCP_sameSize.tiff"),
imagetype="tiff",
output=TRUE
)


#### get DLL3 expression in all cell types

scrna.unintegrated <-  readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))

UMAPPlot(scrna.unintegrated,group.by="final_anno")

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_all_cell_types.pdf")))

FeaturePlot(scrna.unintegrated,features="DLL3", label=TRUE)
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_DLL3_in_all_cell_types.pdf")))


VlnPlot(scrna.unintegrated,features="DLL3",group.by="final_anno")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violin_DLL3_in_all_cell_types.pdf")))

## EpCAM

FeaturePlot(scrna.unintegrated,features="EPCAM", label=TRUE)
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_EPCAM_in_all_cell_types.pdf")))


VlnPlot(scrna.unintegrated,features="EPCAM",group.by="final_anno")
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violin_EPCAM_in_all_cell_types.pdf")))


## CK

for (ck in c("KRT8","KRT18","KRT19")){
  FeaturePlot(scrna.unintegrated,features=ck, label=TRUE)
  ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_",ck,"_in_all_cell_types.pdf")))
  
  VlnPlot(scrna.unintegrated,features=ck,group.by="final_anno")
  ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_violin_",ck,"_in_all_cell_types.pdf")))
}


## EpCAM and SCLC markers

FeaturePlot(scrna.unintegrated,features=c("EPCAM","ASCL1","NEUROD1","POU2F3"), label=TRUE)
ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_UMAP_EPCAM_SCLCMarkers_in_all_cell_types.pdf")))

### correlation of DLL3 (pseudobulk) expression and %-A type within each tumor

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(epi_sub, assays = "RNA", return.seurat=T,group.by = c("sample"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("sample","DLL3","ASCL1","NEUROD1","POU2F3"))

## spearman
ggscatter(df, x = "ASCL1", y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab("ASCL1 expression") + ylab("DLL3 expression")+ stat_cor(method = "spearman")

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_pseudobulk_DLL3_ASCL1_cor_spearman.pdf")))

smarker <- c("NEUROD1","POU2F3")

for(gene in smarker){
  ggscatter(df, x = gene, y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab(paste0(gene, " expression")) + ylab("DLL3 expression")+ stat_cor(method = "spearman")

  ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_pseudobulk_DLL3_",gene,"_cor_spearman.pdf")))
}


## pearson
ggscatter(df, x = "ASCL1", y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Pearson correlation (pseudobulk)"
  ) + xlab("ASCL1 expression (pseudobulk)") + ylab("DLL3 expression (pseudobulk)")+ stat_cor(method = "pearson")

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_pseudobulk_DLL3_ASCL1_cor_pearson.pdf")))

### correlation between %A and DLL3

colnames(df) <- gsub("-","_",colnames(df))

ggscatter(df, x = "SCLC_A", y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab("%SCLC-A") + ylab("DLL3 expression")+ stat_cor(method = "spearman")

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_pseudobulk_DLL3_Atype_cor_spearman.pdf")))

### correlation between %A and ASCL1
ggscatter(df, x = "SCLC_A", y = "ASCL1",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab("%SCLC-A") + ylab("ASCL1 expression")+ stat_cor(method = "spearman")

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_pseudobulk_ASCL1_Atype_cor_spearman.pdf")))

## add subtype info

dst <- as.data.frame(unclass(table(epi_sub$sample,epi_sub$SCLC_subtype2)/rowSums(table(epi_sub$sample,epi_sub$SCLC_subtype2))))

dst$sample <- rownames(dst)


## merge

df <- merge(df,dst)

## get treatment info

dtreat <- as.data.frame(unclass(table(epi_sub$sample,epi_sub$treatment)))

dtreat$treatment <- ifelse(dtreat$naive >0, "naive","treated")

dtreat$sample <- rownames(dtreat)

dtreat <- dtreat[,c("sample","treatment")]

dfinal <- merge(df,dtreat)


write.xlsx(dfinal,paste0("../../plots/revision/",Sys.Date(),"_DLL3_ASCL1_pseudobulk_subtype_treatment_summary.xlsx"))

### compare DLL3

ggboxplot(dfinal, x = "treatment", y = "DLL3",
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="pseudobulk log-normalized DLL3") + theme(legend.position = "none") + theme(text = element_text(size = 7)) +
          stat_compare_means(label.x.npc="middle", size=5) 

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_boxplot_by_DLL3_pseudobulk_by_treatment.pdf")))

### compare ASCL1

ggboxplot(dfinal, x = "treatment", y = "ASCL1",
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="pseudobulk log-normalized ASCL1") + theme(legend.position = "none") + theme(text = element_text(size = 7)) +
          stat_compare_means(label.x.npc="middle", size=5) 

ggsave(here(paste0("../../plots/revision/",Sys.Date(),"_boxplot_by_ASCL1_pseudobulk_by_treatment.pdf")))


### 4. GSEA DLL3+ vs DLL3-

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## load tumor aggressiveness pathway with name of gene_sets
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/tumor.aggressiveness.list.RData")

epi_sub <- AddModuleScore_UCell(epi_sub, features = gene_sets, ncores=4, name = "")

VlnPlot2(epi_sub, features=names(gene_sets), group.by = "DLL3_group", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("blue","red"), label="p.format",label.size = 2)

ggsave((paste0("../../plots/revision/",Sys.Date(),"_SCLC_SMpathway_score_epi.pdf")))


### load NE signatures with the object name of NE
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/NE.pathway.list.RData")

epi_sub <- AddModuleScore_UCell(epi_sub, features = NE, ncores=4, name = "")

VlnPlot2(epi_sub, features=names(NE), group.by = "DLL3_group", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("blue","red"), label="p.format",label.size=1.8)

ggsave((paste0("../../plots/revision/",Sys.Date(),"_SCLC_NE_pathway_score_epi.pdf")))

#### additional GSEA

epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

## add DLL3 at the beginning of DLL3_pos

epi_sub$DLL3_group <- factor(paste0("DLL3_",epi_sub$DLL3_group))

##### assign EPCAM high/low group
epi_sub$EPCAM_group <- "EPCAM_low"
epcam_median <- median(epi_sub[['RNA']]$data["EPCAM",])
epi_sub$EPCAM_group[which(epi_sub[['RNA']]$data["EPCAM",] > epcam_median)] <- "EPCAM_high" 
epi_sub$EPCAM_group <- factor(epi_sub$EPCAM_group)

### assign combine group

epi_sub$DLL3_EPCAM <- paste0(epi_sub$DLL3_group,"_",epi_sub$EPCAM_group)

epi_sub$DLL3_EPCAM  <- factor(epi_sub$DLL3_EPCAM, levels=c("DLL3_pos_EPCAM_low","DLL3_pos_EPCAM_high","DLL3_neg_EPCAM_high","DLL3_neg_EPCAM_low"))


## load tumor aggressiveness pathway with name of gene_sets
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/tumor.aggressiveness.list.RData")

epi_sub <- AddModuleScore_UCell(epi_sub, features = gene_sets, ncores=4, name = "")

VlnPlot2(epi_sub, features=names(gene_sets), group.by = "DLL3_EPCAM", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2.2)+ theme(axis.text.x = element_blank())

ggsave((paste0("../../plots/revision/",Sys.Date(),"_SCLC_SMpathway_score_epi_by_DLL3_EPCAM.pdf")))
### NE

### load NE signatures with the object name of NE
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/NE.pathway.list.RData")

epi_sub <- AddModuleScore_UCell(epi_sub, features = NE, ncores=4, name = "")

VlnPlot2(epi_sub, features=names(NE), group.by = "DLL3_EPCAM", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2) + theme(axis.text.x = element_blank())

ggsave((paste0("../../plots/revision/",Sys.Date(),"_SCLC_NE_pathway_score_epi_by_DLL3_EPCAM.pdf")))


### get DLL3+ cell distribution

scrna.unintegrated <-  readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.withanno.rds"))

scrna.unintegrated$DLL3_group <- "neg"

scrna.unintegrated$DLL3_group[which(scrna.unintegrated[['RNA']]$data["DLL3",] > 0)] <- "pos" 
scrna.unintegrated$DLL3_group <- factor(scrna.unintegrated$DLL3_group)

data <- data.frame(unclass(table(scrna.unintegrated$final_anno,scrna.unintegrated$DLL3_group)/colSums(table(scrna.unintegrated$final_anno,scrna.unintegrated$DLL3_group))))

data$cellType <- rownames(data)

colnames(data) <- c("DLL3_neg","DLL3_pos","cellType")

data$lbls <- ifelse(data$DLL3_pos >0.1,paste0(round(data$DLL3_pos*100),"%"),"")

data$cellType <- factor(data$cellType,levels=c("epithelial",data$cellType[-which(data$cellType=="epithelial")]))
# Basic piechart
ggplot(data, aes(x="", y=DLL3_pos, fill=cellType)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0) +
  geom_text(aes(label = lbls),position = position_stack(vjust = 0.5)) +theme_void() +
  ggtitle("%DLL3-positive cells by cell type")

ggsave((paste0("../../plots/revision/",Sys.Date(),"_DLL3_pos_cell_distribution_by_cellType.pdf")))

### get EPCAM and CK expression range
epi_sub <- readRDS(paste0("../../rdata/","scrna.unintegrated.seurat.final.rds"))

exp(range(epi_sub[['RNA']]$data["EPCAM",which(epi_sub[['RNA']]$data["EPCAM",] > 0)]))-1
#[1]  0.2302238 57.2333171
## define CK +/-
expr_matrix <- FetchData(epi_sub, vars = c("KRT8","KRT18","KRT19"))

exp(range(expr_matrix[expr_matrix != 0]))-1

## [1]  0.1286273 97.5609756
```

