---
title: "01_SCLC_scRNA"
output: html_document
date: "2025-01-10"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Dataset I: Chan, 2021, Cancer Cell

This dataset was downloaded from this paper: https://pmc.ncbi.nlm.nih.gov/articles/PMC8628860/#SM1. The dataset is stored on the Human Tumor Atlas Network (HTAN) data portal at https://data.humantumoratlas.org/. URL: url: https://cellxgene.cziscience.com/collections/62e8f058-9c37-48bc-9200-e767f318a8ec. Download using: wget https://datasets.cellxgene.cziscience.com/a87fde70-4b3e-4982-afea-ae6f4cd1b48b.rds. reference: hg38.

## step 0: load 

```{r}
library(Seurat)
library(here)
library(ggplot2)
library(forcats)

library(dplyr)

library(presto)
library(ggrepel)
library(msigdbr)
library(clusterProfiler)
library(stringr)

library(openxlsx)
library(DESeq2)
library(ggpubr)

library(infercnv)

library(SeuratExtend)

library(webr)
library(ggvenn)
library(eulerr)
library(gridExtra)

library(tidyverse)

library(GeneNMF)
library(fgsea)
library(UCell)
```



## step 1: load data
```{r}
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"
sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells.rds"))

# change rownames from Ensembl ID to gene names
RNA <- sclc_epi@assays$RNA

newnames <- as.character(sclc_epi@assays$RNA@meta.features$feature_name)


if (nrow(RNA) == length(newnames)) {
    if (length(RNA@counts)) RNA@counts@Dimnames[[1]]            <- newnames
    if (length(RNA@data)) RNA@data@Dimnames[[1]]                <- newnames
    if (length(RNA@scale.data)) RNA@scale.data@Dimnames[[1]]    <- newnames
  } else {"Unequal gene sets: nrow(RNA) != nrow(newnames)"}

sclc_epi@assays$RNA <- RNA

saveRDS(sclc_epi,paste0(inDir,"/sclc_epithelial_cells_with_gene_as_rowname.rds"))

```

## Step 2: visualize

```{r}
#library(Seurat)
#library(here)
#library(ggplot2)
sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_with_gene_as_rowname.rds"))

## umap by SCLC_subtype

UMAPPlot(sclc_epi, group.by="SCLC_subtype")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_SCLC_subtype.pdf")))

## check subtype markers

FeaturePlot(sclc_epi, features = c("ASCL1","NEUROD1","POU2F3","YAP1"))
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_SCLC_markers.pdf")))
FeaturePlot(sclc_epi, features = c( "NRXN1", "SSTR2", "ID1", "ID3"))
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_SCLC_N_additional_markers.pdf")))

## umap by donor.id

UMAPPlot(sclc_epi, group.by="donor_id")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_donor_id.pdf")))

##check DLL3 expression
FeaturePlot(sclc_epi, features = "DLL3")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_DLL3.pdf")))

## assign new groups

## DLL3 pos vs neg
sclc_epi$DLL3_group <- "neg"
 
sclc_epi$DLL3_group[which(sclc_epi[['RNA']]$data["DLL3",] > 0)] <- "pos" 
sclc_epi$DLL3_group <- factor(sclc_epi$DLL3_group)

## show UMAP
UMAPPlot(sclc_epi, group.by="DLL3_group")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_DLL3_pos_neg.pdf")))

Idents(sclc_epi) <- "SCLC_subtype"

VlnPlot(object=sclc_epi, features = 'DLL3', split.by = 'DLL3_group',pt.size = 0)
ggsave(here(paste0("../../plots/",Sys.Date(),"_violin_DLL3_by_SCLC_pos_neg.pdf")))

## naive vs treatment: 7 naive, 12 treated
sclc_epi$treatment2 <- ifelse(sclc_epi$treatment == "Naive","Naive","Treated")
sclc_epi$newCellType <- paste(sclc_epi$SCLC_subtype,sclc_epi$DLL3_group,sep="_")
sclc_epi$newCellType_treatment <- ifelse(sclc_epi$treatment == "Naive",paste0(sclc_epi$newCellType,"_Naive"),paste0(sclc_epi$newCellType,"_Treated"))

saveRDS(sclc_epi, file = paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))


## 

sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))


## distribution of SCLC subtype by donor

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$SCLC_subtype))
colnames(samType) <- c("Sample","cellType","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(cellType == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = cellType)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "Sample")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_SCLC_per_sample.pdf")))



## get treatment information for each sample

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$SCLC_subtype))
colnames(samType) <- c("Sample","cellType","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(cellType == "SCLC-A") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = cellType)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "Cell Subtype per sample by treatment", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SCLC_subtype_by_treatment.pdf")))


## get DLL3 pos and neg for each sample

## distribution of DLLE+ vs DLL3- by donor

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(DLL3 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "DLL3 expression per sample", y = "Proportion", x = "Sample")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_DLL3_per_sample.pdf")))

## get DLL3 by treatment information for each sample

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by treatment", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_by_treatment.pdf")))


## get DLL3 by treatment and subtype for each sample

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_group,sclc_epi$SCLC_subtype))
colnames(samType) <- c("Sample","DLL3","SCLC_subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(SCLC_subtype,Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment+SCLC_subtype, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per subtype per sample by treatment", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_by_subtype_by_treatment.pdf")))


## add primary and metastasis info

## get DLL3 by treatment and subtype for each sample

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_group,sclc_epi$stage))
colnames(samType) <- c("Sample","DLL3","stage","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$stage,FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(stage,Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment+stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by treatment by stage", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_by_stage_by_treatment.pdf")))

## compare SCLC-A and SCLC-N between naive and treated groups
#library(ggpubr)

## get DLL3 by treatment and subtype for each sample

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_group,sclc_epi$SCLC_subtype))
colnames(samType) <- c("Sample","DLL3","SCLC_subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

samSum <- inner_join(samType, samTreat, by="Sample")

## keep -A AND -N

samSum1 <- samSum[which(samSum$SCLC_subtype != "SCLC-P" & samSum$DLL3 == "neg"),]


p <- ggplot(samSum1, aes(x = SCLC_subtype, y = Value, fill = Treatment)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  labs(title = "Comparison of naive vs treated of % DLL3- by SCLC subtype",
       x = "SCLC_subtype",
       y = "Value") +
  theme_minimal()

# Add statistical comparisons
p <- p + stat_compare_means(aes(group = Treatment), method = "wilcox.test")

print(p)

ggsave(here(paste0("../../plots/",Sys.Date(),"_A_N_neg_percent_between_treatment.pdf")))


## remove those with a total of cells < 10


samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_group,sclc_epi$SCLC_subtype))
colnames(samType) <- c("Sample","DLL3","SCLC_subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100
samType$Sum <- ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum)

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

samSum <- inner_join(samType, samTreat, by="Sample")

## keep -A AND -N

samSum1 <- samSum[which(samSum$SCLC_subtype != "SCLC-P" & samSum$DLL3 == "neg" & samSum$Sum > 10),]


p <- ggplot(samSum1, aes(x = SCLC_subtype, y = Value, fill = Treatment)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  labs(title = "Comparison of % DLL3- by SCLC subtype between naive vs treated (total >10)",
       x = "SCLC_subtype",
       y = "Value") +
  theme_minimal()

# Add statistical comparisons
p <- p + stat_compare_means(aes(group = Treatment), method = "wilcox.test")

print(p)

ggsave(here(paste0("../../plots/",Sys.Date(),"_A_N_neg_percent_between_treatment_total_gt10.pdf")))

## how about total >30

## keep -A AND -N

samSum1 <- samSum[which(samSum$SCLC_subtype != "SCLC-P" & samSum$DLL3 == "neg" & samSum$Sum > 30),]


p <- ggplot(samSum1, aes(x = SCLC_subtype, y = Value, fill = Treatment)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  labs(title = "Comparison of % DLL3- by SCLC subtype between naive vs treated (total >30)",
       x = "SCLC_subtype",
       y = "Value") +
  theme_minimal()

# Add statistical comparisons
p <- p + stat_compare_means(aes(group = Treatment), method = "wilcox.test")

print(p)

ggsave(here(paste0("../../plots/",Sys.Date(),"_A_N_neg_percent_between_treatment_total_gt30.pdf")))



### by DLL3+

## get DLL3 by treatment and subtype for each sample

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_group,sclc_epi$SCLC_subtype))
colnames(samType) <- c("Sample","DLL3","SCLC_subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

samSum <- inner_join(samType, samTreat, by="Sample")

## keep -A AND -N

samSum1 <- samSum[which(samSum$SCLC_subtype != "SCLC-P" & samSum$DLL3 == "pos"),]


p <- ggplot(samSum1, aes(x = SCLC_subtype, y = Value, fill = Treatment)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  labs(title = "Comparison of naive vs treated of % DLL3+ by SCLC subtype",
       x = "SCLC_subtype",
       y = "Value") +
  theme_minimal()

# Add statistical comparisons
p <- p + stat_compare_means(aes(group = Treatment), method = "wilcox.test")

print(p)

ggsave(here(paste0("../../plots/",Sys.Date(),"_A_N_pos_percent_between_treatment.pdf")))


## remove those with a total of cells < 10


samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_group,sclc_epi$SCLC_subtype))
colnames(samType) <- c("Sample","DLL3","SCLC_subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100
samType$Sum <- ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum)

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

samSum <- inner_join(samType, samTreat, by="Sample")

## keep -A AND -N

samSum1 <- samSum[which(samSum$SCLC_subtype != "SCLC-P" & samSum$DLL3 == "pos" & samSum$Sum > 10),]


p <- ggplot(samSum1, aes(x = SCLC_subtype, y = Value, fill = Treatment)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  labs(title = "Comparison of % DLL3+ by SCLC subtype between naive vs treated (total >10)",
       x = "SCLC_subtype",
       y = "Value") +
  theme_minimal()

# Add statistical comparisons
p <- p + stat_compare_means(aes(group = Treatment), method = "wilcox.test")

print(p)

ggsave(here(paste0("../../plots/",Sys.Date(),"_A_N_pos_percent_between_treatment_total_gt10.pdf")))

## how about total >30

## keep -A AND -N

samSum1 <- samSum[which(samSum$SCLC_subtype != "SCLC-P" & samSum$DLL3 == "pos" & samSum$Sum > 30),]


p <- ggplot(samSum1, aes(x = SCLC_subtype, y = Value, fill = Treatment)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  labs(title = "Comparison of % DLL3+ by SCLC subtype between naive vs treated (total >30)",
       x = "SCLC_subtype",
       y = "Value") +
  theme_minimal()

# Add statistical comparisons
p <- p + stat_compare_means(aes(group = Treatment), method = "wilcox.test")

print(p)

ggsave(here(paste0("../../plots/",Sys.Date(),"_A_N_pos_percent_between_treatment_total_gt30.pdf")))


```
## step 3: scDGE

```{r}
#library(presto)

Idents(sclc_epi) <- 'DLL3_group'
dll3.deg.markers <- FindMarkers(sclc_epi, ident.1 = "pos", ident.2 = "neg")
head(dll3.deg.markers,n=10)

#library(ggrepel)
dll3.deg.markers$Gene <- rownames(dll3.deg.markers)
# The significantly differentially expressed genes are the ones found in the upper-left and upper-right corners.
# Add a column to the data frame to specify if they are UP- or DOWN- regulated (log2FoldChange respectively positive or negative)
# add a column of NAs
dll3.deg.markers$diffexpressed <- "NO"
fcthresh <- 2
pthresh <- 0.05
# if avg_log2FC > FCthreshold and p_val_adj < pthreshold, set as "UP"
dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC > log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "UP"
# if avg_log2FC < -FCthreshold and p_val_adj < pthreshold, set as "DOWN"
dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC < -log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "DOWN"
# Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
dll3.deg.markers$delabel <- NA
dll3.deg.markers$delabel[dll3.deg.markers$diffexpressed != "NO"] <- dll3.deg.markers$Gene[dll3.deg.markers$diffexpressed != "NO"]

pdf(here(paste0("../../plots/",Sys.Date(),"_volcano_DLL3_pos_vs_neg_fc",fcthresh,"_p_",pthresh,".pdf")))
# plot adding up all layers we have seen so far
p <-
  ggplot(data=dll3.deg.markers, aes(x=avg_log2FC, y=-log10(p_val_adj), col=diffexpressed, label=delabel)) +
  geom_point() +
  theme_minimal() +
  geom_text_repel(size=2,max.overlaps = 40) +
  scale_color_manual(values=c("blue", "black", "red")) +
  geom_vline(xintercept=c(-log2(fcthresh), log2(fcthresh)), col="red") +
  geom_hline(yintercept=-log10(pthresh), col="red") +
  ggtitle("DLL3 pos vs neg") + xlab("log2FoldChange")
print(p)
dev.off()

#library(msigdbr)
#library(clusterProfiler)
#library(stringr)
# only HALLMARK set
TERM2GENE <- msigdbr(species = "Homo sapiens", category = "H")%>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()
emsigdbup <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="UP")], TERM2GENE = TERM2GENE)
dotplot(emsigdbup, showCategory=25, font.size = 7, title=paste0("MSigDB (UP)"))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))

ggsave(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_pos_vs_neg_fc",fcthresh,"_p",pthresh,".pdf")))
emsigdbdn <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="DOWN")], TERM2GENE = TERM2GENE)
dotplot(emsigdbdn, showCategory=25, font.size = 7, title=paste0("MSigDB (DOWN)"))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))
ggsave(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_pos_vs_neg_DOWN_fc",fcthresh,"_p",pthresh,".pdf")))

### subtype level DLL3 pos vs neg
#library(openxlsx)
Idents(sclc_epi) <- 'newCellType'

for(group in unique(sclc_epi$SCLC_subtype)){
  case <- paste0(group,"_pos")
  ctrl <- paste0(group,"_neg")
  dll3.deg.markers <- FindMarkers(sclc_epi, ident.1 = case, ident.2 = ctrl)
  dll3.deg.markers$Gene <- rownames(dll3.deg.markers)
  # add a column of NAs
  dll3.deg.markers$diffexpressed <- "NO"
  fcthresh <- 2
  pthresh <- 0.05
  # if avg_log2FC > FCthreshold and p_val_adj < pthreshold, set as "UP"
  dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC > log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "UP"
  # if avg_log2FC < -FCthreshold and p_val_adj < pthreshold, set as "DOWN"
  dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC < -log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "DOWN"
  # Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
  dll3.deg.markers$delabel <- NA
  dll3.deg.markers$delabel[dll3.deg.markers$diffexpressed != "NO"] <- dll3.deg.markers$Gene[dll3.deg.markers$diffexpressed != "NO"]
  ##save into table
  write.table(dll3.deg.markers, file=here(paste0("../../tables/",Sys.Date(),"_volcano_DLL3_",group,"_pos_vs_neg_fc",fcthresh,"_p",pthresh,".txt")),quote=FALSE,col.names = NA,sep="\t")
  write.xlsx(dll3.deg.markers,here(paste0("../../tables/",Sys.Date(),"_volcano_DLL3_",group,"_pos_vs_neg_fc",fcthresh,"_p",pthresh,".xlsx")))
  # plot adding up all layers we have seen so far
  try({
    pdf(here(paste0("../../plots/",Sys.Date(),"_volcano_DLL3_",group,"_pos_vs_neg_fc",fcthresh,"_p",pthresh,".pdf")))
    p <- ggplot(data=dll3.deg.markers, aes(x=avg_log2FC, y=-log10(p_val_adj), col=diffexpressed, label=delabel)) +
      geom_point() +
      theme_minimal() +
      geom_text_repel(size=2,max.overlaps = 40) +
      scale_color_manual(values=c("blue", "black", "red")) +
      geom_vline(xintercept=c(-log2(fcthresh), log2(fcthresh)), col="red") +
      geom_hline(yintercept=-log10(pthresh), col="red") +
      ggtitle(paste0("DLL3 pos vs neg: ",group)) + xlab("log2FoldChange")
    print(p)
    dev.off()})
  ## enrichment
  # only HALLMARK set
  TERM2GENE <- msigdbr(species = "Homo sapiens", category = "H")%>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()
  # UP
  try({
    emsigdbup <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="UP")], TERM2GENE = TERM2GENE)
    pdf(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_",group,"_pos_vs_neg_UP_fc",fcthresh,"_p",pthresh,".pdf")))
    p <- dotplot(emsigdbup, showCategory=25, font.size = 7, title=paste0("MSigDB (UP): ",group))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))
    print(p)
    dev.off()})
  ##DOWN
  try({
    emsigdbdn <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="DOWN")], TERM2GENE = TERM2GENE)
    pdf(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_",group,"_pos_vs_neg_DOWN_fc",fcthresh,"_p",pthresh,".pdf")))
    p<- dotplot(emsigdbdn, showCategory=25, font.size = 7, title=paste0("MSigDB (DOWN): ",group))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))
    print(p)
    dev.off()})
}

## DGE for subtypeXtreatment: pos vs neg
Idents(sclc_epi) <- 'newCellType_treatment'

for(group in unique(sclc_epi$SCLC_subtype)){
  for(treat in unique(sclc_epi$treatment2)){
    case <- paste0(group,"_pos_",treat)
    ctrl <- paste0(group,"_neg_",treat)
    dll3.deg.markers <- FindMarkers(sclc_epi, ident.1 = case, ident.2 = ctrl)
    dll3.deg.markers$Gene <- rownames(dll3.deg.markers)
    # add a column of NAs
    dll3.deg.markers$diffexpressed <- "NO"
    fcthresh <- 2
    pthresh <- 0.05
    # if avg_log2FC > FCthreshold and p_val_adj < pthreshold, set as "UP"
    dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC > log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "UP"
    # if avg_log2FC < -FCthreshold and p_val_adj < pthreshold, set as "DOWN"
    dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC < -log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "DOWN"
    # Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
    dll3.deg.markers$delabel <- NA
    dll3.deg.markers$delabel[dll3.deg.markers$diffexpressed != "NO"] <- dll3.deg.markers$Gene[dll3.deg.markers$diffexpressed != "NO"]
    ##save into table
    write.table(dll3.deg.markers, file=here(paste0("../../tables/",Sys.Date(),"_volcano_DLL3_",group,"_",treat,"_fc",fcthresh,"_p",pthresh,".txt")),quote=FALSE,col.names = NA,sep="\t")
    write.xlsx(dll3.deg.markers,here(paste0("../../tables/",Sys.Date(),"_volcano_DLL3_",group,"_",treat,"_fc",fcthresh,"_p",pthresh,".xlsx")))
    # plot adding up all layers we have seen so far
    try({
      pdf(here(paste0("../../plots/",Sys.Date(),"_volcano_DLL3_",group,"_",treat,"_fc",fcthresh,"_p",pthresh,".pdf")))
      p <- ggplot(data=dll3.deg.markers, aes(x=avg_log2FC, y=-log10(p_val_adj), col=diffexpressed, label=delabel)) +
        geom_point() +
        theme_minimal() +
        geom_text_repel(size=4,max.overlaps = 40) +
        scale_color_manual(values=c("blue", "black", "red")) +
        geom_vline(xintercept=c(-log2(fcthresh), log2(fcthresh)), col="red") +
        geom_hline(yintercept=-log10(pthresh), col="red") +
        ggtitle(paste0("DLL3 pos vs neg: ",group, ":",treat)) + xlab("log2FoldChange")
      print(p)
      dev.off()})
    ## enrichment
    # only HALLMARK set
    TERM2GENE <- msigdbr(species = "Homo sapiens", category = "H")%>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()
    # UP
    try({
      emsigdbup <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="UP")], TERM2GENE = TERM2GENE)
      pdf(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_",group,"_",treat,"_UP_fc",fcthresh,"_p",pthresh,".pdf")))
      p <- dotplot(emsigdbup, showCategory=25, font.size = 7, title=paste0("MSigDB (UP): ",group,":",treat))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))
      print(p)
      dev.off()})
    ##DOWN
    try({
      emsigdbdn <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="DOWN")], TERM2GENE = TERM2GENE)
      pdf(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_",group,"_",treat,"_DOWN_fc",fcthresh,"_p",pthresh,".pdf")))
      p<- dotplot(emsigdbdn, showCategory=25, font.size = 7, title=paste0("MSigDB (DOWN): ",group,":",treat))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))
      print(p)
      dev.off()})
  }
}

## add primary or metastasis information
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"


sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

## could not figure out which sample is  PleuralEffusion: could be RU1215 or Ru325(maybe this one)
sclc_epi$stage <- ifelse(sclc_epi$donor_id %in% c("RU1066","RU1145","RU1229","RU1231","RU426"),"Primary","Metastasis")

sclc_epi$age <- sapply(strsplit(as.character(sclc_epi$development_stage), "-"), function(x) x[1])

saveRDS(sclc_epi,paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

## 

## DGE for subtypeXtreatmentXstage: pos vs neg
sclc_epi$newCellType_treatment_stage <- paste(sclc_epi$newCellType_treatment,sclc_epi$stage,sep="_")
Idents(sclc_epi) <- 'newCellType_treatment_stage'

for(group in unique(sclc_epi$SCLC_subtype)){
  for(treat in unique(sclc_epi$treatment2)){
    for(stage in unique(sclc_epi$stage)){
      case <- paste0(group,"_pos_",treat,"_",stage)
      ctrl <- paste0(group,"_neg_",treat,"_",stage)
      if((case %in% unique(sclc_epi$newCellType_treatment_stage)) & (ctrl %in% unique(sclc_epi$newCellType_treatment_stage))){
        dll3.deg.markers <- FindMarkers(sclc_epi, ident.1 = case, ident.2 = ctrl)
        dll3.deg.markers$Gene <- rownames(dll3.deg.markers)
        # add a column of NAs
        dll3.deg.markers$diffexpressed <- "NO"
        fcthresh <- 2
        pthresh <- 0.05
        # if avg_log2FC > FCthreshold and p_val_adj < pthreshold, set as "UP"
        dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC > log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "UP"
        # if avg_log2FC < -FCthreshold and p_val_adj < pthreshold, set as "DOWN"
        dll3.deg.markers$diffexpressed[dll3.deg.markers$avg_log2FC < -log2(fcthresh) & dll3.deg.markers$p_val_adj < pthresh] <- "DOWN"
        # Create a new column "delabel" to de, that will contain the name of genes differentially expressed (NA in case they are not)
        dll3.deg.markers$delabel <- NA
        dll3.deg.markers$delabel[dll3.deg.markers$diffexpressed != "NO"] <- dll3.deg.markers$Gene[dll3.deg.markers$diffexpressed != "NO"]
        ##save into table
        write.table(dll3.deg.markers, file=here(paste0("../../tables/",Sys.Date(),"_volcano_DLL3_",group,"_",stage,"_",treat,"_fc",fcthresh,"_p",pthresh,".txt")),quote=FALSE,col.names = NA,sep="\t")
        write.xlsx(dll3.deg.markers,here(paste0("../../tables/",Sys.Date(),"_volcano_DLL3_",group,"_",stage,"_",treat,"_fc",fcthresh,"_p",pthresh,".xlsx")))
        # plot adding up all layers we have seen so far
        try({
          pdf(here(paste0("../../plots/",Sys.Date(),"_volcano_DLL3_",group,"_",stage,"_",treat,"_fc",fcthresh,"_p",pthresh,".pdf")))
          p <- ggplot(data=dll3.deg.markers, aes(x=avg_log2FC, y=-log10(p_val_adj), col=diffexpressed, label=delabel)) +
            geom_point() +
            theme_minimal() +
            geom_text_repel(size=4,max.overlaps = 40) +
            scale_color_manual(values=c("blue", "black", "red")) +
            geom_vline(xintercept=c(-log2(fcthresh), log2(fcthresh)), col="red") +
            geom_hline(yintercept=-log10(pthresh), col="red") +
            ggtitle(paste0("DLL3 pos vs neg: ",group, ":",stage,":",treat)) + xlab("log2FoldChange")
          print(p)
          dev.off()})
        ## enrichment
        # only HALLMARK set
        TERM2GENE <- msigdbr(species = "Homo sapiens", category = "H")%>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()
        # UP
        try({
          emsigdbup <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="UP")], TERM2GENE = TERM2GENE)
          pdf(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_",group,"_",stage,"_",treat,"_UP_fc",fcthresh,"_p",pthresh,".pdf")))
          p <- dotplot(emsigdbup, showCategory=25, font.size = 7, title=paste0("MSigDB (UP): ",group,":",stage,":",treat))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))
          print(p)
          dev.off()})
        ##DOWN
        try({
          emsigdbdn <- enricher(gene =  dll3.deg.markers$Gene[which(dll3.deg.markers$diffexpressed=="DOWN")], TERM2GENE = TERM2GENE)
          pdf(here(paste0("../../plots/",Sys.Date(),"_dotplot_enrichment_DLL3_",group,"_",stage,"_",treat,"_DOWN_fc",fcthresh,"_p",pthresh,".pdf")))
          p<- dotplot(emsigdbdn, showCategory=25, font.size = 7, title=paste0("MSigDB (DOWN): ",group,":",stage,":",treat))+scale_y_discrete(labels=function(x) str_wrap(x, width=25))
          print(p)
          dev.off()})
      }
    }
  }
}


```

## step 4: pseudobulk DGE

```{r}
#library(DESeq2)
# pseudobulk the counts based on donor-condition-celltype
pseudo_sclc_epi <- AggregateExpression(sclc_epi, assays = "RNA", return.seurat = T, group.by = c("donor_id", "SCLC_subtype","DLL3_group"))

# each 'cell' is a donor-celltype-dll3level pseudobulk profile
tail(Cells(pseudo_sclc_epi))

Idents(pseudo_sclc_epi) <- "DLL3_group"

bulk.sclc.de <- FindMarkers(object = pseudo_sclc_epi, 
                         ident.1 = "pos", 
                         ident.2 = "neg",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)



pseudo_sclc_epi$celltype.dll3 <- paste(pseudo_sclc_epi$SCLC_subtype, pseudo_sclc_epi$DLL3_group, sep = "_")

Idents(pseudo_sclc_epi) <- "celltype.dll3"

bulk.sclc.de <- FindMarkers(object = pseudo_sclc_epi, 
                         ident.1 = "SCLC-A_pos", 
                         ident.2 = "SCLC-A_neg",
                         test.use = "DESeq2")

bulk.sclc.de <- FindMarkers(object = pseudo_sclc_epi, 
                         ident.1 = "SCLC-N_pos", 
                         ident.2 = "SCLC-N_neg",
                         test.use = "DESeq2")

bulk.sclc.de <- FindMarkers(object = pseudo_sclc_epi, 
                         ident.1 = "SCLC-P_pos", 
                         ident.2 = "SCLC-P_neg",
                         test.use = "DESeq2")

head(bulk.sclc.de, n = 15)


```

## Step 5: CK expression vs DLL3 expression

```{r}
## cytokeratin 8, 18 and 19 expression (together) in the SCLC cells. Just for starters - you could look at CK expression in the DLL3+ and DLL3- groups. 
## The reason is - when we score CTCs, we stain it with a cocktail of pan CK antibodies (green channel) DLL3 antibody (yellow channel) WBC markers (red channel)

## The WBC negative CK+ cells are considered to be the CTCs. We then evaluate this population for DLL3 expression.

## So it would be interesting to see if both DLL3+ and DLL3- tumor cell populations have CK expression. 
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

Idents(sclc_epi) <- 'DLL3_group'

ckgenes <- c("KRT8","KRT18","KRT19")

FeaturePlot(sclc_epi, features = ckgenes)

ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_panCK.pdf")))

Idents(sclc_epi) <- 'SCLC_subtype'

for (gene in ckgenes) {
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap_",gene,".pdf")))
  p <- VlnPlot(sclc_epi, features = gene,pt.size = 0, split.by = "DLL3_group") 

  print(p)
  dev.off()
} 

## relax logFC threshold

Idents(sclc_epi) <- 'newCellType'
dll3.deg.markers <- FindMarkers(sclc_epi, ident.1 = "SCLC-P_pos", ident.2 = "SCLC-P_neg",logfc.threshold =0.01)
dll3.deg.markers[which(rownames(dll3.deg.markers) %in% c("KRT8","KRT18","KRT19")),]
```

## Step 6: new markers for grouping?

```{r}
## after monthly meeting on 2025-01-15
## may need to find new ways of grouping SCLC

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

## observation from previous UMAP by M:
## each of the patient "islands" has an inward-facing side (closer to the center)
## if we saw that the DLL+ cells were always on the outward-facing side, it would be intriguing

## suggested by R
## Since your scRNA-Seq data is mostly clustered by patients, do you think it would be useful to look at a few of them individually (1 patient at a time) to see if the cells would cluster based on DLL3 or not?


## let's pick a sample and see if DLL3 expression will be able to separate cell within this sample

unique(sclc_epi$donor_id)
#  [1] RU1215          RU1152          RU1231          RU1322          RU1181          RU1108          RU1145          RU1066          RU1293         
# [10] RU1080          RU1144          RU1229          PleuralEffusion RU426           RU1124          RU1195          RU779           RU1311         
# [19] RU1065 

## RU1108: SCLC-A; RU1293: SCLC-N 

## subset data
for(sample in unique(sclc_epi$donor_id)){
  tmp <- subset(x = sclc_epi, subset = donor_id == sample)
  
  
  counts <- GetAssayData(tmp,assay = "RNA",layer = "counts") 
  ## replace underscores in feature names with dashes
  newRowname <-  gsub("_", "-", rownames(tmp)) 
  rownames(counts) <- newRowname
  obj <- CreateSeuratObject(counts = counts,meta.data = tmp@meta.data)
  
  ## process
  obj <- NormalizeData(obj)
  obj <- FindVariableFeatures(obj)
  obj <- ScaleData(obj)
  obj <- RunPCA(obj)
  
  obj <- FindNeighbors(obj, dims = 1:30, reduction = "pca")
  obj <- FindClusters(obj)
  
  obj <- RunUMAP(obj, dims = 1:30, reduction = "pca")
  # visualize by Dll3_group annotation and SCLC subtype
  mainCell <- unique(obj$SCLC_major_subtype_of_sample)
  p <- DimPlot(obj, group.by = "DLL3_group") 
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap_",sample,"_by_dll3.pdf")))
  print(p)
  dev.off()

  # plot by SCLC subtype
  
  p <- DimPlot(obj, group.by = "SCLC_subtype") + ggtitle(paste0(sample, " : ", mainCell))
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap_",sample,"_by_subtype.pdf")))
  print(p)
  dev.off()
  
  # plot by known markers
  
  p <- FeaturePlot(obj, features = c("ASCL1","NEUROD1","POU2F3","CD274"))
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap_",sample,"_SCLC_markers.pdf")))
  print(p)
  dev.off()

}


```

## Step 7: UMAP3
```{r}

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

## let's check all samples' together first
## rerun all preprocessing
## Compute UMAP with 3 Components

counts <- GetAssayData(sclc_epi,assay = "RNA",layer = "counts") 
## replace underscores in feature names with dashes
newRowname <-  gsub("_", "-", rownames(sclc_epi)) 
rownames(counts) <- newRowname
obj <- CreateSeuratObject(counts = counts,meta.data = sclc_epi@meta.data)
  
obj <- NormalizeData(obj)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- RunPCA(obj)

#obj <- FindNeighbors(obj, dims = 1:30, reduction = "pca")
#obj <- FindClusters(obj, resolution = 1)

obj <- RunUMAP(obj, dims = 1:30, n.components = 3)

#saveRDS(sclc_epi,paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

DimPlot(obj,group.by='donor_id',cols=c("PleuralEffusion"="#7CAE00","RU426"="#660000","RU779"="#C77CFF","RU1065"="#008FC4","RU1066"="#FF61CC","RU1080"="#00A9FF","RU1108"="#ED68ED","RU1124"="#8494FF","RU1144"="#00C19A","RU1145"="#ABA300","RU1152"="#CD9600","RU1181"="#E68613","RU1195"="#9999CC","RU1215"="#333333","RU1229"="#FF3300","RU1231"="#99CCCC","RU1293"="#0033FF","RU1311"="#FF3399","RU1322"="#00FF00")) 
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap3_by_donor.pdf")))

DimPlot(obj, group.by = c( "DLL3_group","SCLC_subtype"))

ggsave(here(paste0("../../plots/",Sys.Date(),"_umap3_by_dll3_subtype.pdf")))

##Extract UMAP Coordinates
umap_3d <- Embeddings(obj, "umap")


## Plot UMAP in 3D using plotly

library(plotly)

# Create a 3D scatter plot
plot_ly(x = umap_3d[,1], 
        y = umap_3d[,2], 
        z = umap_3d[,3], 
        color = obj$DLL3_group, 
        colors = RColorBrewer::brewer.pal(n = 9, "Set1"),  # Customize colors
        type = "scatter3d", 
        mode = "markers",
        marker = list(size = 3))  # Adjust marker size

DimPlot(obj,dims=c(1,2),group.by="DLL3_group")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap3_12.pdf")))
DimPlot(obj,dims=c(1,3),group.by="DLL3_group")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap3_13.pdf")))
DimPlot(obj,dims=c(2,3),group.by="DLL3_group")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap3_23.pdf")))

#ggsave(here(paste0("../../plots/2025-01-23_umap3_23.pdf")))

FeaturePlot(sclc_epi,features = "EPCAM")
ggsave(here(paste0("../../plots/",Sys.Date(),"_EpCAM.pdf")))


#### for individual
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"
sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

## subset data
for(sample in unique(sclc_epi$donor_id)){
  tmp <- subset(x = sclc_epi, subset = donor_id == sample)
  
  
  counts <- GetAssayData(tmp,assay = "RNA",layer = "counts") 
  ## replace underscores in feature names with dashes
  newRowname <-  gsub("_", "-", rownames(tmp)) 
  rownames(counts) <- newRowname
  obj <- CreateSeuratObject(counts = counts,meta.data = tmp@meta.data)
  
  ## process
  obj <- NormalizeData(obj)
  obj <- FindVariableFeatures(obj)
  obj <- ScaleData(obj)
  obj <- RunPCA(obj)
  
  obj <- RunUMAP(obj, dims = 1:30, reduction = "pca",n.components = 3)
  
  ## save individual as the shape of UMAP keep changing
  saveRDS(obj,paste0(inDir,"/../../../rdata/",Sys.Date(),"_",sample,".rds"))
  # visualize by Dll3_group annotation and SCLC subtype
  mainCell <- unique(obj$SCLC_major_subtype_of_sample)
  
  ## UMAP
  
  ## UMAP1 vs UMAP2
  p <- DimPlot(obj, dims=c(1,2),group.by = "DLL3_group")
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap3_",sample,"_by_dll3_12.pdf")))
  print(p)
  dev.off()
  
   ## UMAP1 vs UMAP3
  p <- DimPlot(obj, dims=c(1,3),group.by = "DLL3_group") 
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap3_",sample,"_by_dll3_13.pdf")))
  print(p)
  dev.off()
  
   ## UMAP1 vs UMAP2
  p <- DimPlot(obj, dims=c(2,3),group.by = "DLL3_group") 
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap3_",sample,"_by_dll3_23.pdf")))
  print(p)
  dev.off()
  
   ## PCA
  
  ## PC1 vs PC2
  p <- DimPlot(obj, dims=c(1,2),reduction="pca",group.by = "DLL3_group")
  pdf(here(paste0("../../plots/",Sys.Date(),"_pc3_",sample,"_by_dll3_12.pdf")))
  print(p)
  dev.off()
  
   ## PC1 vs PC3
  p <- DimPlot(obj, dims=c(1,3),reduction="pca",group.by = "DLL3_group") 
  pdf(here(paste0("../../plots/",Sys.Date(),"_pc3_",sample,"_by_dll3_13.pdf")))
  print(p)
  dev.off()
  
   ## PC1 vs PC2
  p <- DimPlot(obj, dims=c(2,3),reduction="pca",group.by = "DLL3_group") 
  pdf(here(paste0("../../plots/",Sys.Date(),"_pc3_",sample,"_by_dll3_23.pdf")))
  print(p)
  dev.off()
  
  
  # plot by SCLC subtype
  
  p <- DimPlot(obj, group.by = "SCLC_subtype") + ggtitle(paste0(sample, " : ", mainCell))
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap_",sample,"_by_subtype.pdf")))
  print(p)
  dev.off()
  
  # plot by known markers
  
  p <- FeaturePlot(obj, features = c("ASCL1","NEUROD1","POU2F3","CD274"))
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap_",sample,"_SCLC_markers.pdf")))
  print(p)
  dev.off()
  
}



FeaturePlot(sclc_epi,features=c("TP53","RB1","DLL3","CD274"))
VlnPlot(sclc_epi,features="DLL3")

## normalized 

gene_expression <- GetAssayData(sclc_epi, layer = "data")["DLL3", ]
hist(gene_expression, main = "Expression Distribution of DLL3", xlab = "Expression Level")
plot(density(gene_expression))


## raw counts

gene_expression <- GetAssayData(sclc_epi, layer = "counts")["DLL3", ]
pdf(here(paste0("../../plots/",Sys.Date(),"_umap_dll3_read_counts.pdf")),width=7,height=7)
hist(gene_expression, main = "reads of DLL3", xlab = "#reads")
dev.off()

plot(density(gene_expression))

## get DLL3 percent

sclc_epi$percent.dll3 <- PercentageFeatureSet(sclc_epi,features="DLL3")

FeaturePlot(sclc_epi, features="percent.dll3") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_dll3_percent.pdf")),width=7,height=7)

## let's try minimum 1-3 counts for DLL3

UMAPPlot(sclc_epi, group.by="DLL3_group")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_dll3_group0.pdf")),width=7,height=7)

## >1
sclc_epi$DLL3_group1 <- "neg"
 
sclc_epi$DLL3_group1[which(sclc_epi[['RNA']]$counts["DLL3",] > 1)] <- "pos" 
sclc_epi$DLL3_group1 <- factor(sclc_epi$DLL3_group1)

## show UMAP
UMAPPlot(sclc_epi, group.by="DLL3_group1")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_dll3_group1.pdf")))

## >2
sclc_epi$DLL3_group2 <- "neg"
 
sclc_epi$DLL3_group2[which(sclc_epi[['RNA']]$counts["DLL3",] > 2)] <- "pos" 
sclc_epi$DLL3_group2 <- factor(sclc_epi$DLL3_group2)

## show UMAP
UMAPPlot(sclc_epi, group.by="DLL3_group2")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_dll3_group2.pdf")))


## >3
sclc_epi$DLL3_group3 <- "neg"
 
sclc_epi$DLL3_group3[which(sclc_epi[['RNA']]$counts["DLL3",] > 3)] <- "pos" 
sclc_epi$DLL3_group3 <- factor(sclc_epi$DLL3_group3)

## show UMAP
UMAPPlot(sclc_epi, group.by="DLL3_group3")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_dll3_group3.pdf")))

DimPlot(sclc_epi,group.by='donor_id',cols=c("PleuralEffusion"="#7CAE00","RU426"="#660000","RU779"="#C77CFF","RU1065"="#008FC4","RU1066"="#FF61CC","RU1080"="#00A9FF","RU1108"="#ED68ED","RU1124"="#8494FF","RU1144"="#00C19A","RU1145"="#ABA300","RU1152"="#CD9600","RU1181"="#E68613","RU1195"="#9999CC","RU1215"="#333333","RU1229"="#FF3300","RU1231"="#99CCCC","RU1293"="#0033FF","RU1311"="#FF3399","RU1322"="#00FF00")) 
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_donor_id.pdf")))


saveRDS(sclc_epi,paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))


### re-do individual for with group3

for(sample in unique(sclc_epi$donor_id)){
  
  obj <- readRDS(paste0(inDir,"/../../../rdata/2025-01-23_",sample,".rds"))
  
  obj$DLL3_group3 <- "low"
 
  obj$DLL3_group3[which(obj[['RNA']]$counts["DLL3",] > 3)] <- "high" 
  obj$DLL3_group3 <- factor(obj$DLL3_group3)

  # visualize by Dll3_group annotation and SCLC subtype
  mainCell <- unique(obj$SCLC_major_subtype_of_sample)
  
  ## UMAP
  
  ## UMAP1 vs UMAP2
  p <- DimPlot(obj, dims=c(1,2),group.by = "DLL3_group3")
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap3_",sample,"_by_dll3_group3_12.pdf")))
  print(p)
  dev.off()
  
   ## UMAP1 vs UMAP3
  p <- DimPlot(obj, dims=c(1,3),group.by = "DLL3_group3") 
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap3_",sample,"_by_dll3_group3_13.pdf")))
  print(p)
  dev.off()
  
   ## UMAP1 vs UMAP2
  p <- DimPlot(obj, dims=c(2,3),group.by = "DLL3_group3") 
  pdf(here(paste0("../../plots/",Sys.Date(),"_umap3_",sample,"_by_dll3_group3_23.pdf")))
  print(p)
  dev.off()
  

}

```

## Step 8: Notch pathway

```{r}
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"
sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))


## reported co-expressed gene

cogenes <- c("ASCL1", "NEUROD1","CHGA","SYP","NCAM1","NOTCH1", "NOTCH2", "NOTCH3", "NOTCH4", "HES1", "HEY1", "JAG1","JAG2","MYC", "BCL2", "SOX2", "POU2F3")

## count noon_zeros in this gene list

# Extract the count matrix (raw counts)
count_matrix <- GetAssayData(sclc_epi, layer = "counts")

# Subset the count matrix to include only the genes in your list
gene_counts <- count_matrix[rownames(count_matrix) %in% cogenes, ]

# Count how many cells have more than zero counts
cells_with_nonzero_counts <- colSums(gene_counts > 0)

pdf(here(paste0("../../plots/",Sys.Date(),"_hist_DLL3_coexpressed_gene_per_cell.pdf")))
hist(cells_with_nonzero_counts, breaks = seq(-0.5, 16.5, by = 1), main="Histogram of # of co-expressed genes of DLL3 per cell")

dev.off()

##create a group for DLL3

sclc_epi$DLL3_coNum <- cells_with_nonzero_counts

for (nco in 0:15){
sclc_epi$DLL3_co <- as.character(sclc_epi$DLL3_group)
 
sclc_epi$DLL3_co[which(sclc_epi$DLL3_group=="neg" & sclc_epi$DLL3_coNum <= nco)] <- "true_neg" 

sclc_epi$DLL3_co[which(sclc_epi$DLL3_group=="neg" & sclc_epi$DLL3_coNum >nco)] <- "false_neg" 

sclc_epi$DLL3_co <- factor(sclc_epi$DLL3_co, levels = c("true_neg","false_neg","pos"))


## plot

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$DLL3_co))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(DLL3 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = paste0("DLL3 expression per sample (false_neg: >= ",nco+1," co-exp genes)"), y = "Proportion", x = "Sample")


ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_using_",nco,"_coexp_by_subtype.pdf")))

}

#### plot by Notch Pathway

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"
sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))


## reported co-expressed gene

notch1 <- c("NOTCH1", "NOTCH2", "NOTCH3", "NOTCH4","JAG1","JAG2","DLL1","DLL3","DLL4")
notch2 <- c("NOTCH1", "NOTCH2", "NOTCH3", "NOTCH4","JAG1","JAG2","DLL1","DLL3","DLL4","RBPJ","HES1", "HES5","HEY1","HEY2")

notch <- list(notch1, notch2)
## add notch score

sclc_epi <- AddModuleScore(sclc_epi,features=notch,ctrl=20,name="NotchPathwayScore")

## plot by notch pathway score

FeaturePlot(sclc_epi, features="NotchPathwayScore1")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_notchpathway_short.pdf")))

FeaturePlot(sclc_epi, features="NotchPathwayScore2")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_notchpathway_long.pdf")))


## plot notch pathways scores for DLL3+ and DLL3-


df <- data.frame(sclc_epi$DLL3_group,sclc_epi$NotchPathwayScore2)

colnames(df) <- c("DLL3_group","NotchPathway")


ggboxplot(df, x = "DLL3_group", y = "NotchPathway",
          color = "DLL3_group", palette = "jco",
          add = "jitter") + stat_compare_means()

ggsave(paste0("../../plots/",Sys.Date(),"_boxplot_notch_score_by_dll3.pdf"))





### create notch high vs low group

sclc_epi$notch <- "low"

sclc_epi$notch[which(sclc_epi$NotchPathwayScore2 > 0)] <- "high" 

UMAPPlot(sclc_epi, group.by="notch")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_notch_low_high.pdf")))


### cell percentage by notch low and high

## plot

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$notch))
colnames(samType) <- c("Sample","NOTCH","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(NOTCH == "high") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = NOTCH)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "NOTCH signaling per sample", y = "Proportion", x = "Sample")


ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_NOTCH_signaling_per_sample.pdf")))

### let's split by SCLC_major_subtype_of_sample 


## get notch by SCLC_major_subtype_of_sample  information for each sample

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$notch))
colnames(samType) <- c("Sample","NOTCH","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$SCLC_major_subtype_of_sample))


colnames(samTreat) <- c("Sample","Main_SCLC_Subtype","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(NOTCH == "high") %>%
   arrange(Main_SCLC_Subtype,desc(Value)) %>%
   select(Main_SCLC_Subtype,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Main_SCLC_Subtype","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = NOTCH)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Main_SCLC_Subtype, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "NOTCH signaling per sample by Main_SCLC_Subtype", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_NOTCH_by_main_SCLC_subtype.pdf")))


### by SCLC_Subtype

## get notch by SCLC_major_subtype_of_sample  information for each sample

samType <- as.data.frame(table(sclc_epi$donor_id,sclc_epi$SCLC_subtype,sclc_epi$notch))
colnames(samType) <- c("Sample","SCLC_subtype","NOTCH","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100


ggplot(samType, aes(x = Sample, y = Value, fill = NOTCH)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ SCLC_subtype, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "NOTCH signaling per sample by SCLC_Subtype", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_NOTCH_by_SCLC_subtype.pdf")))



```


## Step 9: DLL3 in other cell types


```{r}
## load data
## downloaded from wget https://datasets.cellxgene.cziscience.com/44d1c1aa-7c79-4cca-9edf-f4e97edc047a.rds
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"
seurat_chan <- readRDS(paste0(inDir,"/chan_2021_all.rds"))
# 22397 features across 147137 samples
# change rownames from Ensembl ID to gene names
RNA <- seurat_chan@assays$RNA

newnames <- as.character(seurat_chan@assays$RNA@meta.features$feature_name)


if (nrow(RNA) == length(newnames)) {
    if (length(RNA@counts)) RNA@counts@Dimnames[[1]]            <- newnames
    if (length(RNA@data)) RNA@data@Dimnames[[1]]                <- newnames
    if (length(RNA@scale.data)) RNA@scale.data@Dimnames[[1]]    <- newnames
  } else {"Unequal gene sets: nrow(RNA) != nrow(newnames)"}

seurat_chan@assays$RNA <- RNA

saveRDS(seurat_chan,paste0(inDir,"/chan_2021_all_with_gene_as_rowname.rds"))


## 
seurat_chan <- readRDS(paste0(inDir,"/chan_2021_all_with_gene_as_rowname.rds"))

## umap by cell_type_general

UMAPPlot(seurat_chan, group.by="cell_type_general")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_cell_type_general.pdf")))

##check DLL3 expression
FeaturePlot(seurat_chan, features = "DLL3")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_DLL3.pdf")))


VlnPlot(seurat_chan,features="DLL3",pt.size = 0, group.by = "cell_type_general")

ggsave(here(paste0("../../plots/",Sys.Date(),"_violin_by_DLL3.pdf")))

## assign new groups

## DLL3 pos vs neg
seurat_chan$DLL3_group <- "neg"
 
seurat_chan$DLL3_group[which(seurat_chan[['RNA']]$data["DLL3",] > 0)] <- "pos" 
seurat_chan$DLL3_group <- factor(seurat_chan$DLL3_group)

## show UMAP
UMAPPlot(seurat_chan, group.by="DLL3_group")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_DLL3_pos_neg.pdf")))


## get DLL3 by celltype for each sample

samType <- as.data.frame(table(seurat_chan$donor_id,seurat_chan$DLL3_group,seurat_chan$cell_type_general))
colnames(samType) <- c("Sample","DLL3","CellType","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$CellType,FUN = sum) * 100

# reordered_data <- samType %>%
#    filter(DLL3 == "pos") %>%
#    arrange(CellType,desc(Value)) %>%
#    select(CellType,DLL3,Sample,Value) %>%
#    mutate(SortOrder = row_number()) %>%
#    mutate(Sample = reorder(Sample, SortOrder))

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(text = element_text(size = 6),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by general cell type", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_by_general_cell_type.pdf")))


### detailed cell types
## umap by cell_type

UMAPPlot(seurat_chan, group.by="cell_type")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_cell_type.pdf")))

samType <- as.data.frame(table(seurat_chan$donor_id,seurat_chan$DLL3_group,seurat_chan$cell_type))
colnames(samType) <- c("Sample","DLL3","CellType","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$CellType,FUN = sum) * 100

# reordered_data <- samType %>%
#    filter(DLL3 == "pos") %>%
#    arrange(CellType,desc(Value)) %>%
#    select(CellType,DLL3,Sample,Value) %>%
#    mutate(SortOrder = row_number()) %>%
#    mutate(Sample = reorder(Sample, SortOrder))

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(text = element_text(size = 4),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by general cell type", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_by_cell_type.pdf")))


### fine cell types
## umap by cell_type_fine

UMAPPlot(seurat_chan, group.by="cell_type_fine")
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_cell_type_fine.pdf")))

samType <- as.data.frame(table(seurat_chan$donor_id,seurat_chan$DLL3_group,seurat_chan$cell_type_fine))
colnames(samType) <- c("Sample","DLL3","CellType","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$CellType,FUN = sum) * 100

# reordered_data <- samType %>%
#    filter(DLL3 == "pos") %>%
#    arrange(CellType,desc(Value)) %>%
#    select(CellType,DLL3,Sample,Value) %>%
#    mutate(SortOrder = row_number()) %>%
#    mutate(Sample = reorder(Sample, SortOrder))

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(text = element_text(size = 4),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by fine cell type", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_by_cell_type.pdf")))



```

## Step 10: check gene capture rate per cell

```{r}
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

# Extract the count matrix (raw counts)
count_matrix <- GetAssayData(sclc_epi, layer = "counts")


# Count how many cells have more than zero counts
cells_with_nonzero_counts <- colSums(count_matrix > 0)

# get percentage

cells_with_nonzero_percents <- cells_with_nonzero_counts/nrow(count_matrix)

sclc_epi$capture_rate <- cells_with_nonzero_percents

pdf(here(paste0("../../plots/",Sys.Date(),"_hist_percent_none_zero_genes_per_cell.pdf")))
hist(cells_with_nonzero_percents, main="% of none_zero genes per cell")

dev.off()

## plot number of genes

# genes
as.data.frame(sclc_epi@meta.data) %>% 
  	ggplot(aes(color=donor_id, x=nFeaturess_RNA_by_counts, fill= donor_id)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 3000) + 
    ggtitle("Genes Per Cell")

ggsave(paste0("../../plots/",Sys.Date(),"_nGenes_per_sample.pdf"))


### plot DLL3+ vs DLL3- groups's capture rates

df <- data.frame(sclc_epi$DLL3_group,sclc_epi$nFeaturess_RNA_by_counts)

colnames(df) <- c("DLL3_group","nGenes")


ggboxplot(df, x = "DLL3_group", y = "nGenes",
          color = "DLL3_group", palette = "jco",
          add = "jitter") + stat_compare_means()

ggsave(paste0("../../plots/",Sys.Date(),"_boxplot_nGenes_by_dll3.pdf"))


# plot by capture rate

df <- data.frame(sclc_epi$DLL3_group,sclc_epi$capture_rate)

colnames(df) <- c("DLL3_group","capture_rate")


ggboxplot(df, x = "DLL3_group", y = "capture_rate",
          color = "DLL3_group", palette = "jco",
          add = "jitter") + stat_compare_means()

ggsave(paste0("../../plots/",Sys.Date(),"_boxplot_capture_rate_by_dll3.pdf"))

## plot density plot of capture rate

ggplot(df, aes(x=capture_rate)) + 
  geom_density() + geom_vline(aes(xintercept=mean(capture_rate)),
            color="blue", linetype="dashed", size=1)

ggsave(paste0("../../plots/",Sys.Date(),"_density_capture_rate.pdf"))


## filter I: 0.05 seem to be a good cutoff

sclc_sub <- subset(sclc_epi,subset= capture_rate>0.05)


### cell percentage by notch low and high

## plot

samType <- as.data.frame(table(sclc_sub$donor_id,sclc_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(DLL3 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "DLL3 expression per sample", y = "Proportion", x = "Sample")


ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_high_capture_rate005.pdf")))

### let's split by SCLC_major_subtype_of_sample 


## get dll3 by SCLC_major_subtype_of_sample  information for each sample

samType <- as.data.frame(table(sclc_sub$donor_id,sclc_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_sub$donor_id,sclc_sub$SCLC_major_subtype_of_sample))


colnames(samTreat) <- c("Sample","Main_SCLC_Subtype","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Main_SCLC_Subtype,desc(Value)) %>%
   select(Main_SCLC_Subtype,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Main_SCLC_Subtype","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Main_SCLC_Subtype, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by Main_SCLC_Subtype", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_high_capture_rate005_by_main_SCLC_subtype.pdf")))


### by SCLC_Subtype

## get dll3 by SCLC_subtype information for each sample
samType <- as.data.frame(table(sclc_sub$donor_id,sclc_sub$SCLC_subtype,sclc_sub$DLL3_group))
colnames(samType) <- c("Sample","SCLC_subtype","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100

samType <- samType %>%
  filter(!is.na(Value)) %>%
  mutate(Value = as.numeric(Value))


for (st in unique(samType$SCLC_subtype)){

df <- samType[which(samType$SCLC_subtype==st),]

sample_order <- df %>%
   filter(DLL3 == "pos") %>%
   arrange(SCLC_subtype, desc(Value)) %>%
   pull(Sample)


# Convert Sample to factor for proper ordering in the plot
df$Sample <- factor(df$Sample, levels = sample_order)

ggplot(df, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ SCLC_subtype, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(y = "Proportion", x = "Sample")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_high_capture_rate005_in_",st,".pdf")))

}


## filter II:  cutoff: top half

cutoff <- median(sclc_epi$capture_rate)

sclc_sub <- subset(sclc_epi,subset= capture_rate>cutoff)


### cell percentage by dll3 low and high

## plot

samType <- as.data.frame(table(sclc_sub$donor_id,sclc_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(DLL3 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "DLL3 expression per sample", y = "Proportion", x = "Sample")


ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_high_capture_rate_median.pdf")))

### let's split by SCLC_major_subtype_of_sample 


## get dll3 by SCLC_major_subtype_of_sample  information for each sample

samType <- as.data.frame(table(sclc_sub$donor_id,sclc_sub$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc_sub$donor_id,sclc_sub$SCLC_major_subtype_of_sample))


colnames(samTreat) <- c("Sample","Main_SCLC_Subtype","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Main_SCLC_Subtype,desc(Value)) %>%
   select(Main_SCLC_Subtype,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Main_SCLC_Subtype","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Main_SCLC_Subtype, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by Main_SCLC_Subtype", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_high_capture_median_by_main_SCLC_subtype.pdf")))


### by SCLC_Subtype

## get dll3 by SCLC_major_subtype_of_sample  information for each sample

samType <- as.data.frame(table(sclc_sub$donor_id,sclc_sub$SCLC_subtype,sclc_sub$DLL3_group))
colnames(samType) <- c("Sample","SCLC_subtype","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100

samType <- samType %>%
  filter(!is.na(Value)) %>%
  mutate(Value = as.numeric(Value))


for (st in unique(samType$SCLC_subtype)){

df <- samType[which(samType$SCLC_subtype==st),]

sample_order <- df %>%
   filter(DLL3 == "pos") %>%
   arrange(SCLC_subtype, desc(Value)) %>%
   pull(Sample)


# Convert Sample to factor for proper ordering in the plot
df$Sample <- factor(df$Sample, levels = sample_order)

ggplot(df, aes(x = Sample, y = Value, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ SCLC_subtype, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(y = "Proportion", x = "Sample")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_high_capture_median_in_",st,".pdf")))

}
##


### what are those cells with low capture rates

FeaturePlot(sclc_epi,features="capture_rate")

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_capture_rate.pdf")))

sclc_epi$capture <- "high"

sclc_epi$capture[which(sclc_epi$capture_rate <=0.05)] <- "low"

UMAPPlot(sclc_epi,group.by="capture")

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_capture_binary.pdf")))


```

## Step 12: prepare figures for manuscript

```{r}
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.0.rds"))

dir.create("../../manuscript/v1",recursive =TRUE)

### QC 1: malignancy validation using infercnv


## check all epithelial cells

epi <- readRDS(paste0(inDir,"/epithelial_cells.rds"))


# change rownames from Ensembl ID to gene names
RNA <- epi@assays$RNA

newnames <- as.character(sclc_epi@assays$RNA@meta.features$feature_name)


if (nrow(RNA) == length(newnames)) {
    if (length(RNA@counts)) RNA@counts@Dimnames[[1]]            <- newnames
    if (length(RNA@data)) RNA@data@Dimnames[[1]]                <- newnames
    if (length(RNA@scale.data)) RNA@scale.data@Dimnames[[1]]    <- newnames
  } else {"Unequal gene sets: nrow(RNA) != nrow(newnames)"}

epi@assays$RNA <- RNA

saveRDS(epi,paste0(inDir,"/epithelial_cells_with_gene_as_rowname.rds"))

## load data on ERIS

inDir <- "/data/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

epi <- readRDS(paste0(inDir,"/epithelial_cells_with_gene_as_rowname.rds"))

### verify cancer cells using inferCNV



library(infercnv)


dir.create("../../manuscript/v1/inferCNV/inputs",,recursive =TRUE)

## download gene order file from 


##### use Basal and Ciliated as control

##prepare input files

## input 1: gene order file

url <- "https://data.broadinstitute.org/Trinity/CTAT/cnv/hg38_gencode_v27.txt"

destfile <- "../../manuscript/v1/inferCNV/inputs/hg38_gencode_v27.txt"

download.file(url, destfile)

##Input 2: count matrix


## subset

epi_sub <- subset(epi,subset= author_cell_type %in% c("AE1","AE2","SCLC-A","SCLC-N","SCLC-P"))


## input 3: sample annotation file

annotation <- as.data.frame(epi_sub@meta.data[,c("donor_id","author_cell_type")])
annotation$cell_type <- as.character(annotation$author_cell_type)
annotation$cell <- rownames(annotation)
## downsample to 5000 per group to speed up
set.seed(123)
cells_keep <- annotation %>%
  group_by(cell_type) %>%
  sample_n(size = min(5000, n())) %>%
  pull(cell)

annotation <- annotation[which(annotation$cell %in% cells_keep),]

idx <- which(annotation$author_cell_type %in% c("SCLC-A","SCLC-N","SCLC-P"))
annotation$cell_type[idx] <- paste0(as.character(annotation$author_cell_type)[idx],"_",annotation$donor_id[idx])

idx <- which(annotation$author_cell_type %in% c("AE2","AE1"))

annotation$cell_type[idx]  <- "AE"

## save both cell and patient id
annotationBoth <- annotation[3]
write.table(annotationBoth, file="../../manuscript/v1/inferCNV/inputs/singleCell.annotation.txt",quote=FALSE,col.names = FALSE,row.names=TRUE,sep="\t")

## save only cell type

idx <- which(annotation$author_cell_type %in% c("AE2","AE1"))
annotation$cell_type2 <- as.character(annotation$author_cell_type)
annotation$cell_type2[idx]  <- "AE"
annotationCel <- annotation[5]

write.table(annotationCel, file="../../manuscript/v1/inferCNV/inputs/singleCell.annotation.celltype.txt",quote=FALSE,col.names = FALSE,row.names=TRUE,sep="\t")


epi_sub2 <- subset(epi_sub,cells=cells_keep)


counts <- epi_sub2@assays$RNA@counts

 
# create the infercnv object
infercnv_obj = CreateInfercnvObject(raw_counts_matrix=counts,
                                    annotations_file="../../manuscript/v1/inferCNV/inputs/singleCell.annotation.celltype.txt",
                                    delim="\t",
                                    gene_order_file="../../manuscript/v1/inferCNV/inputs/hg38_gencode_v27.txt",
                                    chr_exclude = c("chrX","chrY","chrM"),
                                    ref_group_names=c("AE"))

options(scipen = 100)
# perform infercnv operations to reveal cnv signal
infercnv_obj = infercnv::run(infercnv_obj,
                             cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
                             #out_dir="../../manuscript/v1/inferCNV/output_cluster_by_group" ## heatmap not clustered within each subtype
                             out_dir="../../manuscript/v1/inferCNV/output",  # dir is auto-created for storing outputs
                             cluster_by_groups=F,   # cluster
                             analysis_mode= "subclusters",
                             denoise=T,
                             HMM=F,
                             num_threads=4,
                             output_format = "pdf"
                             )


## failed due to memory; switch to ERIS


### QC 2: tumor annotation

## umap by SCLC_subtype

UMAPPlot(sclc_epi, group.by="SCLC_subtype")
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_SCLC_subtype.pdf")))

## check subtype markers

FeaturePlot(sclc_epi, features = c("ASCL1","NEUROD1","POU2F3","YAP1"))
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_SCLC_markers.pdf")))


### QC 3: capture rate

# Extract the count matrix (raw counts)
count_matrix <- GetAssayData(sclc_epi, layer = "counts")


# Count how many cells have more than zero counts
cells_with_nonzero_counts <- colSums(count_matrix > 0)

# get percentage

cells_with_nonzero_percents <- cells_with_nonzero_counts/nrow(count_matrix)

sclc_epi$capture_rate <- cells_with_nonzero_percents


# plot by capture rate

df <- data.frame(sclc_epi$DLL3_group,sclc_epi$capture_rate)

colnames(df) <- c("DLL3_group","capture_rate")


## plot density plot of capture rate

ggplot(df, aes(x=capture_rate)) + 
  geom_density() + geom_vline(aes(xintercept=0.05),
            color="blue", linetype="dashed", size=1)

ggsave(paste0("../../manuscript/v1/",Sys.Date(),"_density_capture_rate.pdf"))

## split into capture high and low
sclc_epi$capture <- "high"

sclc_epi$capture[which(sclc_epi$capture_rate <=0.05)] <- "low"

UMAPPlot(sclc_epi,group.by="capture")

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_UMAP_capture_binary.pdf")))


## check cell quality for capture low vs capture high

## get mitochondrial fraction

## get mitochonrial from all epi

epi <- readRDS(paste0(inDir,"/epithelial_cells_with_gene_as_rowname.rds"))

mito <- as.data.frame(epi$mito_frac)
colnames(mito) <- "mito_frac"
mito$cell <- rownames(mito)

## get indices of cells in sclc_epi in epi

idx <- match(colnames(sclc_epi),mito$cell)

sclc_epi$mito_frac <- mito$mito_frac[idx]

## plot all mito_frac by capture rate


VlnPlot(sclc_epi,features="mito_frac",group.by="capture",pt.size=0)
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_violin_mito_by_capture_binary.pdf")))

## plot nGene

VlnPlot(sclc_epi,features="nFeature_RNA",group.by="capture",pt.size=0)
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_violin_nGenes_by_capture_binary.pdf")))

## plot UMI

VlnPlot(sclc_epi,features="nCount_RNA",group.by="capture",pt.size=0)
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_violin_nUMI_by_capture_binary.pdf")))


## capture rates affected by treatment?

df <- as.data.frame(table(sclc_epi$treatment2,sclc_epi$capture))
colnames(df) <- c("treatment","capture","value")

ggplot(df, aes(fill=capture, y=value, x=treatment)) + 
    geom_bar(position="fill", stat="identity") + theme_classic()

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_stackedBar_treatment_capture_binary.pdf")))


## capture rates affected by stage?

df <- as.data.frame(table(sclc_epi$stage,sclc_epi$capture))
colnames(df) <- c("stage","capture","value")

ggplot(df, aes(fill=capture, y=value, x=stage)) + 
    geom_bar(position="fill", stat="identity") + theme_classic()

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_stackedBar_stage_capture_binary.pdf")))

## capture rates affected by SCLC subtype?

df <- as.data.frame(table(sclc_epi$SCLC_subtype,sclc_epi$capture))
colnames(df) <- c("SCLC","capture","value")

ggplot(df, aes(fill=capture, y=value, x=SCLC)) + 
    geom_bar(position="fill", stat="identity") + theme_classic()

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_stackedBar_sclc_subtype_capture_binary.pdf")))


### ## get dll3 by capture rate 

sclc_epi <- readRDS(paste0(inDir,"/sclc_epithelial_cells_v1.1.rds"))

df <- as.data.frame(table(sclc_epi$DLL3_group,sclc_epi$capture))
colnames(df) <- c("DLL3","capture","value")

ggplot(df, aes(fill=DLL3, y=value, x=capture)) + 
    geom_bar(position="fill", stat="identity") + 
    scale_fill_manual(values = c("pos" = "#F8766D", "neg" = "#00BFC4"))+
    theme_classic()

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_stackedBar_DLL3_pos_neg_capture_binary.pdf")))



### save
saveRDS(sclc_epi,(paste0(inDir,"/sclc_epithelial_cells_v1.1.rds")))


### exluce low capture group

sclc <- subset(sclc_epi,subset=capture=="high")

saveRDS(sclc,(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds")))

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

### figure 1: subtype

## umap by SCLC_subtype

UMAPPlot(sclc, group.by="SCLC_subtype")
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_SCLC_subtype_postfilter.pdf")))


## umap by SCLC_subtype_plus_TP53_RB1_wt

UMAPPlot(sclc, group.by="SCLC_subtype_plus_TP53_RB1_wt")
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_SCLC_subtype_plus_TP53_RB1_wt_postfilter.pdf")))


  ## umap by sample

UMAPPlot(sclc, group.by="donor_id")
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_donor_id_postfilter.pdf")))


## check subtype markers

FeaturePlot(sclc, features = c("ASCL1","NEUROD1","POU2F3","YAP1"))
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_SCLC_markers_postfilter.pdf")))

## additional SCLC-N markers

FeaturePlot(sclc, features = c("NRXN1","SSTR2","ID1","ID3"))
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_SCLC_N_additional_markers_postfilter.pdf")))


## immune checkpoints

ic <- c("CD274","PDCD1","CD80","CTLA4","CD38","IDO1","TIGIT","C10orf54","LAG3","ICOS")

FeaturePlot(sclc, features = ic)
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_immune_checkpoints_postfilter.pdf")))


## stacked bar for subtype in each group

## distribution of SCLC subtype by donor

samType <- as.data.frame(table(sclc$donor_id,sclc$SCLC_subtype))
colnames(samType) <- c("Sample","cellType","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(cellType == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = cellType)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "")
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_stackBar_SCLC_per_sample_postfilter.pdf")))



## cell counts

df <- as.data.frame(table(sclc$SCLC_subtype))
colnames(df) <- c("SCLC","count")


ggplot(df, aes(x=as.factor(SCLC), y=count, fill=as.factor(SCLC) )) +  
  geom_bar( stat="identity") + labs(x = "")+
  scale_fill_manual(values = c("red", "green", "blue") ) +
  geom_text(aes(label=count),position=position_dodge(0.5), vjust=-0.25) +
  theme_classic() + theme(legend.position="none") 

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_Bar_sclc_subtype_Ncell_postfilter.pdf")))


## HLA genes

hla <- rownames(sclc)[which(startsWith(rownames(sclc),"HLA"))]
hla1 <- c("HLA-A","HLA-B","HLA-C","HLA-E","HLA-E","HLA-F","HLA-G")

FeaturePlot(sclc, features = hla1)
ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_HLA1_postfilter.pdf")))

### figure 2: DLL3+ vs DLL3- histogram by subtype



## get dll3 by SCLC_major_subtype_of_sample  information for each sample

samType <- as.data.frame(table(sclc$donor_id,sclc$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc$donor_id,sclc$SCLC_major_subtype_of_sample))


colnames(samTreat) <- c("Sample","Main_SCLC_Subtype","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Main_SCLC_Subtype,desc(Value)) %>%
   select(Main_SCLC_Subtype,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Main_SCLC_Subtype","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Main_SCLC_Subtype, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by Main_SCLC_Subtype", y = "Proportion", x = "")

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_by_main_SCLC_subtype_postfilter.pdf")))

### figure 3: DLL3 violin plot

## violin plot

VlnPlot(sclc,features="DLL3",group.by="SCLC_subtype",split.by="DLL3_group",pt.size=0)+ labs(x="") + scale_fill_manual(values = c("pos" = "red", "neg" = "blue"))

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_violin_by_DLL3_per_sample_by_SCLC_subtype_postfilter.pdf")))


## violin plot with median with enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(sclc, expression=DLL3_group=="pos")
  
VlnPlot2(
  sclc,
  features = "DLL3",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  stat.method = "wilcox.test") +scale_fill_manual(values = c("red", "green", "blue") )


ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_violin_by_DLL3_per_sample_by_SCLC_subtype_with_box.pdf")))

### figure 4: DLL3+ vs DLL3- by treatment
## get dll3 by SCLC_major_subtype_of_sample  information for each sample

samType <- as.data.frame(table(sclc$donor_id,sclc$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc$donor_id,sclc$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(DLL3 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = DLL3)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "DLL3 expression per sample by Treatment", y = "Proportion", x = "")

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_stackedbar_by_DLL3_per_sample_by_treatment_postfilter.pdf")))

### boxplot

df <- reordered_data %>%
   filter(DLL3 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%DLL3+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_boxplot_by_DLL3_per_sample_by_treatment_postfilter.pdf")))

### figure 5: CK level: KRT8, KRT18, KRT19

## featureplot

ck8 <- FeaturePlot(sclc, features = c("KRT8"))

ck8v <- VlnPlot(sclc,features="KRT8",group.by="SCLC_subtype",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck8t <- VlnPlot(sclc,features="KRT8",group.by="treatment2",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck18 <- FeaturePlot(sclc, features = c("KRT18"))

ck18v <- VlnPlot(sclc,features="KRT18",group.by="SCLC_subtype",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck18t <- VlnPlot(sclc,features="KRT18",group.by="treatment2",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck19 <- FeaturePlot(sclc, features = c("KRT19"))

ck19v <- VlnPlot(sclc,features="KRT19",group.by="SCLC_subtype",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ck19t <- VlnPlot(sclc,features="KRT19",group.by="treatment2",split.by="DLL3_group",pt.size=0) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + labs(x="")

ggarrange(ck8,ck8v,ck8t,ck18,ck18v,ck18t,ck19,ck19v,ck19t,nrow=3,ncol=3,font.label = list(size = 5))

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_umap_by_CK_markers_postfilter.pdf")))

ggarrange(ck8v,ck8t,ck18v,ck18t,ck19v,ck19t,nrow=3,ncol=2,font.label = list(size = 5))

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_violin_CK_markers_postfilter.pdf")))

FeaturePlot(sclc, features = c("KRT8"))

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_UMAP_KRT8_postfilter.pdf")))

FeaturePlot(sclc, features = c("KRT18"))

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_UMAP_KRT18_postfilter.pdf")))

FeaturePlot(sclc, features = c("KRT19"))

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_UMAP_KRT19_postfilter.pdf")))

ggarrange(ck8,ck18,ck19,nrow=3,ncol=1,font.label = list(size = 5))

ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_UMAP_CK_markers_postfilter_3in1.pdf")))

## NOTCH pathway as baits

## reported co-expressed gene in Notch pathway

cogenes <- c("NOTCH1", "NOTCH2", "NOTCH3", "NOTCH4", "HES1", "HEY1", "JAG1","JAG2")

## count noon_zeros in this gene list

# Extract the count matrix (raw counts)
count_matrix <- GetAssayData(sclc, layer = "counts")

# Subset the count matrix to include only the genes in your list
gene_counts <- count_matrix[rownames(count_matrix) %in% cogenes, ]

# Count how many cells have more than zero counts
cells_with_nonzero_counts <- colSums(gene_counts > 0)

pdf(here(paste0("../../manuscript/v1/",Sys.Date(),"_hist_DLL3_coexpressed_gene_per_cell.pdf")))
hist(cells_with_nonzero_counts, breaks = seq(-0.5, 8.5, by = 1), main="Counts of DLL3 co-expressed genes per cell")

dev.off()

##create a group for DLL3

sclc$DLL3_coNum <- cells_with_nonzero_counts

for (nco in 0:8){
sclc$DLL3_co <- as.character(sclc$DLL3_group)
 
sclc$DLL3_co[which(sclc$DLL3_group=="neg" & sclc$DLL3_coNum <= nco)] <- "true_neg" 

sclc$DLL3_co[which(sclc$DLL3_group=="neg" & sclc$DLL3_coNum >nco)] <- "false_neg" 

sclc$DLL3_co <- factor(sclc$DLL3_co, levels = c("true_neg","false_neg","pos"))


## plot

samType <- as.data.frame(table(sclc$donor_id,sclc$DLL3_co))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(DLL3 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = DLL3)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = paste0("DLL3 expression per sample (false_neg: >= ",nco+1," co-exp genes)"), y = "Proportion", x = "")


ggsave(here(paste0("../../manuscript/v1/",Sys.Date(),"_stackedbar_by_DLL3_using_",nco,"_coexp_by_subtype.pdf")))

}



```

## Step 13: SEZ6, BH73 and pathways between DLL3+ and DLL3- cells

```{r}


inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))



## SEZ6 and B7H3(CD276)

## SCLC markers
FeaturePlot(sclc, features = c("SEZ6"))+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_UMAP_SEZ6.pdf"))

FeaturePlot(sclc, features = c("CD276"))+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_UMAP_CD276.pdf"))

FeaturePlot(sclc, features = c("DLL3"))+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_UMAP_DLL3.pdf"))


### correlation

FeatureScatter(sclc, feature1 = "DLL3", feature2 = "SEZ6",group.by = "donor_id")+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_scatter_SEZ6_vs_DLL3.pdf"))


FeatureScatter(sclc, feature1 = "DLL3", feature2 = "CD276",group.by = "donor_id")+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_scatter_CD276_vs_DLL3.pdf"))


FeatureScatter(sclc, feature1 = "SEZ6", feature2 = "CD276",group.by = "donor_id")+ coord_fixed()

ggsave(paste0("../../plots/",Sys.Date(),"_scatter_CD276_vs_SEZ6.pdf"))


## multiple on the same plot

FeaturePlot3(sclc, feature.1 = "SEZ6", feature.2 = "CD276", feature.3 = "DLL3", pt.size = 1)

ggsave(paste0("../../plots/",Sys.Date(),"_UMAP_SEZ6_CD276_DLL3.pdf"))


### SEZ6+ vs SEZ6-


## SEZ6 pos vs neg
sclc$SEZ6_group <- "neg"
 
sclc$SEZ6_group[which(sclc[['RNA']]$data["SEZ6",] > 0)] <- "pos" 
sclc$SEZ6_group <- factor(sclc$SEZ6_group)

## show UMAP
UMAPPlot(sclc, group.by="SEZ6_group",cols=c("blue","red")) + coord_fixed()
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_SEZ6_pos_neg.pdf")))

### plot fraction of SEZ6+ SEZ6-

###plot stackedbar per sample

samType <- as.data.frame(table(sclc$donor_id,sclc$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample,FUN = sum) * 100


## sort sample

reordered_data <- samType %>%
  filter(SEZ6 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(text = element_text(size = 8),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample", y = "Proportion", x = "Sample")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_by_sample.pdf")))



### violin plot with enhanced vlnplots for subtype

cells <- WhichCells(sclc, expression=(SEZ6_group=="pos") & (SCLC_subtype %in% c("SCLC-A","SCLC-N","SCLC-P")))
  
VlnPlot2(
  sclc,
  features = "SEZ6",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("#F8766D", "#00BF7D", "#00B0F6"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_SCLC_subtype_withBox.pdf")))



## SEZ6+ vs SEZ6- by treatment

samType <- as.data.frame(table(sclc$donor_id,sclc$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc$donor_id,sclc$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(SEZ6 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample by Treatment", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_per_sample_by_treatment.pdf")))

### boxplot

df <- reordered_data %>%
   filter(SEZ6 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%SEZ6+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_per_sample_by_treatment.pdf")))

## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(sclc, expression=SEZ6_group=="pos")
  
VlnPlot2(
  sclc,
  features = "SEZ6",
  cells=cells,
  group.by = "treatment2",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_treatment_withBox.pdf")))




## SEZ6+ vs SEZ6- by stage

samType <- as.data.frame(table(sclc$donor_id,sclc$SEZ6_group))
colnames(samType) <- c("Sample","SEZ6","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc$donor_id,sclc$stage))


colnames(samTreat) <- c("Sample","Stage","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(SEZ6 == "pos") %>%
   arrange(Stage,desc(Value)) %>%
   select(Stage,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Stage","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = SEZ6)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "SEZ6 expression per sample by Stage", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_per_sample_by_stage.pdf")))

### boxplot

df <- reordered_data %>%
   filter(SEZ6 == "pos") %>%
   select(Stage, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Stage", y = "Value.y",
          color = "Stage", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%SEZ6+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_per_sample_by_stage.pdf")))



## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(sclc, expression=SEZ6_group=="pos")
  
VlnPlot2(
  sclc,
  features = "SEZ6",
  cells=cells,
  group.by = "stage",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_stage_withBox.pdf")))


### CD276+ vs CD276-


## CD276 pos vs neg
sclc$CD276_group <- "neg"
 
sclc$CD276_group[which(sclc[['RNA']]$data["CD276",] > 0)] <- "pos" 
sclc$CD276_group <- factor(sclc$CD276_group)

## show UMAP
UMAPPlot(sclc, group.by="CD276_group",cols=c("blue","red")) + coord_fixed()
ggsave(here(paste0("../../plots/",Sys.Date(),"_umap_by_CD276_pos_neg.pdf")))

### plot fraction of CD276+ CD276-

###plot stackedbar per sample

samType <- as.data.frame(table(sclc$donor_id,sclc$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample,FUN = sum) * 100


## sort sample

reordered_data <- samType %>%
  filter(CD276 == "pos") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   #facet_wrap(~ CellType, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(text = element_text(size = 8),axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "CD276 expression per sample", y = "Proportion", x = "Sample") #+ coord_fixed()

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_by_sample.pdf")))



### violin plot with enhanced vlnplots for subtype

cells <- WhichCells(sclc, expression=(CD276_group=="pos") & (SCLC_subtype %in% c("SCLC-A","SCLC-N","SCLC-P")))
  
VlnPlot2(
  sclc,
  features = "CD276",
  cells=cells,
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("#F8766D", "#00BF7D", "#00B0F6"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_SCLC_subtype_withBox.pdf")))



## CD276+ vs CD276- by treatment

samType <- as.data.frame(table(sclc$donor_id,sclc$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc$donor_id,sclc$treatment2))


colnames(samTreat) <- c("Sample","Treatment","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(CD276 == "pos") %>%
   arrange(Treatment,desc(Value)) %>%
   select(Treatment,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Treatment","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Treatment, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "B7H3 expression per sample by Treatment", y = "Proportion", x = "",fill="B7H3")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_per_sample_by_treatment.pdf")))

### boxplot

df <- reordered_data %>%
   filter(CD276 == "pos") %>%
   select(Treatment, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Treatment", y = "Value.y",
          color = "Treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%CD276+") + theme(legend.position = "none") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_CD276_per_sample_by_treatment.pdf")))

## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(sclc, expression=CD276_group=="pos")
  
VlnPlot2(
  sclc,
  features = "CD276",
  cells=cells,
  group.by = "treatment2",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_treatment_withBox.pdf")))




## CD276+ vs CD276- by stage

samType <- as.data.frame(table(sclc$donor_id,sclc$CD276_group))
colnames(samType) <- c("Sample","CD276","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

samTreat <- as.data.frame(table(sclc$donor_id,sclc$stage))


colnames(samTreat) <- c("Sample","Stage","Freq")
samTreat <- samTreat[which(samTreat$Freq > 0),1:2]

#library(dplyr)

samSum <- inner_join(samType, samTreat, by="Sample")

# reorder samples

#library(forcats)
# Calculate the value of Subgroup "Y" for each Category

reordered_data <- samSum %>%
   filter(CD276 == "pos") %>%
   arrange(Stage,desc(Value)) %>%
   select(Stage,Sample,Value) %>%
   mutate(SortOrder = row_number()) %>%
   inner_join(samSum,by = c("Stage","Sample")) %>%
   mutate(Sample = reorder(Sample, SortOrder))

ggplot(reordered_data, aes(x = Sample, y = Value.y, fill = CD276)) +
   geom_bar(stat = "identity", position = "fill") +
   facet_wrap(~ Stage, scales = "free_x") +
   scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
   scale_x_discrete(expand = c(0, 0)) +
   scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) +
   theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) + 
   labs(title = "CD276 expression per sample by Stage", y = "Proportion", x = "")

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_CD276_per_sample_by_stage.pdf")))

### boxplot

df <- reordered_data %>%
   filter(CD276 == "pos") %>%
   select(Stage, Sample, Value.y) %>%
   as.data.frame()

ggboxplot(df, x = "Stage", y = "Value.y",
          color = "Stage", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%CD276+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_CD276_per_sample_by_stage.pdf")))



## violin plot enhanced vlnplot2 in SeuratExtend

cells <- WhichCells(sclc, expression=CD276_group=="pos")
  
VlnPlot2(
  sclc,
  features = "CD276",
  cells=cells,
  group.by = "stage",
  pt=F,
  width=0.5,
  cols=c("red","red"),
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_stage_withBox.pdf")))

## save seurat object
saveRDS(sclc,paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"
sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

## pathways

# proliferation: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/CELL_PROLIFERATION_GO_0008283

#Stemness : https://www.gsea-msigdb.org/gsea/msigdb/cards/MALTA_CURATED_STEMNESS_MARKERS.html

#Quiescence: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GRAHAM_NORMAL_QUIESCENT_VS_NORMAL_DIVIDING_UP.html

#Metastasis promoting pathways: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/RAMASWAMY_METASTASIS_UP.html

#Tumor aggressiveness: did not find, use MYC and E2F instead

##E2F: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/HALLMARK_E2F_TARGETS.html

#Invasion : https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/WANG_TUMOR_INVASIVENESS_UP.html?ex=1

## p53 pathway:https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/HALLMARK_P53_PATHWAY.html 

### load pathway signatures

proliferation <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="proliferation",colNames = FALSE)
stemness <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="stemness",colNames = FALSE)
quiescence <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="quiescence",colNames = FALSE)
metastasis <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="metastasis",colNames = FALSE)
MYC <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="MYC",colNames = FALSE)
E2F <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="E2F",colNames = FALSE)
invasion <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="invasion",colNames = FALSE)
p53 <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="p53",colNames = FALSE)
EMT <- read.xlsx("../../../pathways/pathway_genesets.xlsx",sheet="EMT",colNames = FALSE)

gene_sets <- list(
  proliferation = proliferation,
  stemness=stemness,
  quiescence=quiescence,
  metastasis=metastasis,
  pMYC=MYC,
  pE2F=E2F,
  invasion=invasion,
  pP53=p53,
  EMT=EMT
)

## save tumor aggressiveness pathway with name of gene_sets
save(gene_sets, file="/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/tumor.aggressiveness.list.RData")

sclc <- AddModuleScore_UCell(sclc, features = gene_sets, ncores=4, name = "")

VlnPlot2(sclc, features=names(gene_sets), group.by = "DLL3_group", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("blue","red"))

ggsave((paste0("../../plots/",Sys.Date(),"_SCLC_SMpathway_score_epi.pdf")))

## check SEZ6 and CD276 expression between DLL3+ and DLL3- cells

VlnPlot2(
  sclc,
  features = "CD276",
  group.by = "DLL3_group",
  pt=F,
  box=F,
  width=0.5,
  show.mean = TRUE,
  cols=c("blue","red"),
  lab_fill = "DLL3",
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_CD276_by_DLL3_withBox.pdf")))


VlnPlot2(
  sclc,
  features = "SEZ6",
  group.by = "DLL3_group",
  pt=F,
  box=F,
  width=0.5,
  show.mean = TRUE,
  cols=c("blue","red"),
  lab_fill = "DLL3",
  stat.method = "wilcox.test") 


ggsave(here(paste0("../../plots/",Sys.Date(),"_violinplot_by_SEZ6_by_DLL3_withBox.pdf")))


### three questions on 4/30/2025

#What fraction of all SCLC are DLL3+ vs DLL3-, and what are the numbers across the classical different histological subsets?
#Same question for SEZ6, plus how much overlap is there for individual cells expressing DLL3 and SEZ6?
#Same question for B7H3, plus how much overlap is there for individual cells expressing DLL3, SEZ6 or B7H3?

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"
sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))


## Q1: What fraction of all SCLC are DLL3+ vs DLL3-, and what are the numbers across the classical different histological subsets?

data <- as.data.frame(sclc@meta.data[,c("SCLC_subtype","DLL3_group")])

df <- data %>% group_by(SCLC_subtype,DLL3_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(DLL3=DLL3_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_DLL3_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =DLL3 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,
		ratioByGroup = FALSE)

dev.off()


#Same question for SEZ6, plus how much overlap is there for individual cells expressing DLL3 and SEZ6?

data <- as.data.frame(sclc@meta.data[,c("SCLC_subtype","SEZ6_group")])

df <- data %>% group_by(SCLC_subtype,SEZ6_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(SEZ6=SEZ6_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_SEZ6_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =SEZ6 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,showRatioThreshold = 0.0002,
		ratioByGroup = FALSE)

dev.off()

#Same question for B7H3, plus how much overlap is there for individual cells expressing DLL3, SEZ6 or B7H3?

data <- as.data.frame(sclc@meta.data[,c("SCLC_subtype","CD276_group")])

df <- data %>% group_by(SCLC_subtype,CD276_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(CD276=CD276_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_CD276_by_SCLCsubtye.pdf")))

PieDonut(df, aes(x =CD276 , y = SCLC_subtype, count = count),r0=0.3,r1=0.9,showRatioThreshold = 0.002,
		ratioByGroup = FALSE)

dev.off()


## cells with DLL3

DLL3_pos <- WhichCells(sclc, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(sclc, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(sclc, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

ggvenn(a)

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3.pdf")))


## venn diagram has the equal sizes and misleading

## try install.packages("eulerr") or upset plot
pdf((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler.pdf")))
plot(euler(a, shape = "ellipse"), quantities = TRUE)
dev.off()

library(gridExtra)
### split by subtype
subtypes <- unique(sclc$SCLC_subtype)
genes <- c("DLL3","SEZ6","CD276")
### loop through three subtypes
for (st in subtypes){
  cells <- WhichCells(sclc,expression=SCLC_subtype==st)
  
  epi <- subset(sclc,cells=cells)
  
  ## loop through three genes
  for (gene in genes){
    group <- paste0(gene,"_group")
    data <- as.data.frame(epi@meta.data[,c(group,"SCLC_subtype")])
    
    df <- data %>% group_by(across(all_of(group))) %>% summarise(count=n(),.groups="drop") 
    
    fill_col <- sym(group)
    
    ggplot(df, aes(x = "", y = count, fill = !!fill_col)) +
      geom_col(color = "black") +
      geom_text(aes(label = count),
                position = position_stack(vjust = 0.5)) +
      coord_polar(theta = "y") +
      scale_fill_manual(values = c("blue","red")) + 
      labs(fill=gene) + ggtitle(st) + 
      theme_void() + theme(plot.title = element_text(hjust = 0.5,vjust = -6))
    
    ggsave(here(paste0("../../plots/",Sys.Date(),"_pie_",gene,"_",st,".pdf")))
  }
  ## venn diagram
  
  ## cells with DLL3
  
  DLL3_pos <- WhichCells(epi, expression=DLL3_group=="pos")
  SEZ6_pos <- WhichCells(epi, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(epi, expression=CD276_group=="pos")
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
  
  p <- plot(euler(a, shape = "ellipse",main=st), quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=st)
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,".pdf")),g)
  #pdf((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,".pdf")))
  #print(p)
  #dev.off()

}

### split by treatment and subtype

### split by subtype
subtypes <- unique(sclc$SCLC_subtype)
treat <- unique(sclc$treatment2)
### loop through three subtypes
for (st in subtypes){
  for (tr in treat){ 
    cells <- WhichCells(sclc,expression=SCLC_subtype==st & treatment2==tr)
    
    epi <- subset(sclc,cells=cells)
  
    ## venn diagram
    
    ## positive cells
    
    DLL3_pos <- WhichCells(epi, expression=DLL3_group=="pos")
    SEZ6_pos <- WhichCells(epi, expression=SEZ6_group=="pos")
    B7H3_pos <- WhichCells(epi, expression=CD276_group=="pos")
    
    a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
    
    p <- plot(euler(a, shape = "ellipse",main=paste0(st,":",tr)), quantities = TRUE)
    g <- grid.arrange(grobs=list(p),top=paste0(st,":",tr))
    ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,"_",tr,".pdf")),g)
   
  }
    
  
}



### get count for summary table

#table(sclc$SCLC_subtype,sclc$SEZ6_group)/rowSums(table(sclc$SCLC_subtype,sclc$SEZ6_group))

sub <- subset(sclc,cells=WhichCells(sclc,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

sub1 <- subset(sub,subset=SCLC_subtype=="SCLC-P")
count <- colSums(GetAssayData(object = sub1, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)

## triple negative

sub <- subset(sclc,subset=SCLC_subtype=="SCLC-P")

length(WhichCells(sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))


### NAIVE/treated
sclc1 <- subset(sclc,subset=treatment2=="Treated")

table(sclc1$SCLC_subtype,sclc1$SEZ6_group)/rowSums(table(sclc1$SCLC_subtype,sclc1$SEZ6_group))

sub <- subset(sclc1,cells=WhichCells(sclc1,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

sub1 <- subset(sub,subset=SCLC_subtype=="SCLC-P")
count <- colSums(GetAssayData(object = sub1, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## triple negative


sub <- subset(sclc1,subset=SCLC_subtype=="SCLC-P")

length(WhichCells(sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

### triple negative for all

length(WhichCells(sclc,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sclc, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
#     0     1     2     3 
# 11907 22387 11948  3000 
table(count)/sum(table(count))
#         0         1         2         3 
# 0.2418058 0.4546322 0.2426384 0.0609236 


#### fill in manuscript

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"
sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

table(sclc$donor_id,sclc$DLL3_group)/rowSums(table(sclc$donor_id,sclc$DLL3_group))
dsum <- as.data.frame.matrix(table(sclc$donor_id,sclc$DLL3_group)/rowSums(table(sclc$donor_id,sclc$DLL3_group)))
# > summary(dsum$pos)
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#  0.0814  0.5438  0.6360  0.6521  0.8265  0.8951 

## naive

summary(dsum[c("RU426","RU1066","RU1145","RU1152","RU1215","RU1229","RU1231"),"pos"])

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.4872  0.5802  0.6360  0.6849  0.8076  0.8951 
 # 
 ## treated
summary(dsum[setdiff(rownames(dsum),c("RU426","RU1066","RU1145","RU1152","RU1215","RU1229","RU1231")),"pos"])

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.0814  0.5481  0.6421  0.6330  0.8133  0.8661 


### get boxplot comparing -A with -N type

samType <- as.data.frame(table(sclc$donor_id,sclc$DLL3_group,sclc$SCLC_major_subtype_of_sample))
colnames(samType) <- c("Sample","DLL3","SCLC_subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, samType$SCLC_subtype,FUN = sum) * 100


## keep -A AND -N

samType <- samType[which(samType$SCLC_subtype != "SCLC-P" & samType$DLL3 == "pos" & !is.na(samType$Value)),]


ggboxplot(samType, x = "SCLC_subtype", y = "Value",
          color = "SCLC_subtype", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="%DLL3+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_DLL3_pos_percent_between_subtype_A_and_N.pdf")))


#############
### pie for three targets by main subtype at sample level

data <- as.data.frame(sclc@meta.data[,c("SCLC_major_subtype_of_sample","DLL3_group")])

df <- data %>% group_by(SCLC_major_subtype_of_sample,DLL3_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(DLL3=DLL3_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_DLL3_by_MainSCLCsubtye.pdf")))

PieDonut(df, aes(x =DLL3 , y = SCLC_major_subtype_of_sample, count = count),r0=0.3,r1=0.9,
		ratioByGroup = FALSE)

dev.off()


#Same question for SEZ6, plus how much overlap is there for individual cells expressing DLL3 and SEZ6?

data <- as.data.frame(sclc@meta.data[,c("SCLC_major_subtype_of_sample","SEZ6_group")])

df <- data %>% group_by(SCLC_major_subtype_of_sample,SEZ6_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(SEZ6=SEZ6_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_SEZ6_by_MainSCLCsubtye.pdf")))

PieDonut(df, aes(x =SEZ6 , y = SCLC_major_subtype_of_sample, count = count),r0=0.3,r1=0.9,showRatioThreshold = 0.0002,
		ratioByGroup = FALSE)

dev.off()

#Same question for B7H3, plus how much overlap is there for individual cells expressing DLL3, SEZ6 or B7H3?

data <- as.data.frame(sclc@meta.data[,c("SCLC_major_subtype_of_sample","CD276_group")])

df <- data %>% group_by(SCLC_major_subtype_of_sample,CD276_group) %>% summarise(count=n(),.groups="drop") %>% dplyr::rename(B7H3=CD276_group)

pdf((paste0("../../plots/",Sys.Date(),"_piedonut_CD276_by_MainSCLCsubtye.pdf")))

PieDonut(df, aes(x =B7H3 , y = SCLC_major_subtype_of_sample, count = count),r0=0.3,r1=0.9,showRatioThreshold = 0.002,
		ratioByGroup = FALSE)

dev.off()

### split by main subtype per sample
subtypes <- unique(sclc$SCLC_major_subtype_of_sample)
genes <- c("DLL3","SEZ6","CD276")
### loop through three subtypes
for (st in subtypes){
  cells <- WhichCells(sclc,expression=SCLC_major_subtype_of_sample==st)
  
  epi <- subset(sclc,cells=cells)
  
  ## loop through three genes
  for (gene in genes){
    group <- paste0(gene,"_group")
    data <- as.data.frame(epi@meta.data[,c(group,"SCLC_major_subtype_of_sample")])
    
    df <- data %>% group_by(across(all_of(group))) %>% summarise(count=n(),.groups="drop") 
    
    fill_col <- sym(group)
    
    ggplot(df, aes(x = "", y = count, fill = !!fill_col)) +
      geom_col(color = "black") +
      geom_text(aes(label = count),
                position = position_stack(vjust = 0.5)) +
      coord_polar(theta = "y") +
      scale_fill_manual(values = c("blue","red")) + 
      labs(fill=gene) + ggtitle(st) + 
      theme_void() + theme(plot.title = element_text(hjust = 0.5,vjust = -6))
    
    ggsave(here(paste0("../../plots/",Sys.Date(),"_pie_",gene,"_",st,".pdf")))
  }
  ## venn diagram
  
  ## cells with DLL3
  
  DLL3_pos <- WhichCells(epi, expression=DLL3_group=="pos")
  SEZ6_pos <- WhichCells(epi, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(epi, expression=CD276_group=="pos")
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
  
  p <- plot(euler(a, shape = "ellipse",main=st), quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=st)
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,".pdf")),g)
  #pdf((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",st,".pdf")))
  #print(p)
  #dev.off()

}

### get count for summary table

#table(sclc$SCLC_major_subtype_of_sample,sclc$SEZ6_group)/rowSums(table(sclc$SCLC_major_subtype_of_sample,sclc$SEZ6_group))

## total
sub <- subset(sclc,cells=WhichCells(sclc,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## subtype

sub1 <- subset(sub,subset=SCLC_major_subtype_of_sample=="SCLC-P")
count <- colSums(GetAssayData(object = sub1, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)

table(count)
table(count)/sum(table(count))


## triple negative

sub <- subset(sclc,subset=SCLC_major_subtype_of_sample=="SCLC-A")

length(WhichCells(sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))


### NAIVE/treated
sclc1 <- subset(sclc,subset=treatment2=="Naive")

table(sclc1$SCLC_major_subtype_of_sample,sclc1$SEZ6_group)
table(sclc1$SCLC_major_subtype_of_sample,sclc1$SEZ6_group)/rowSums(table(sclc1$SCLC_major_subtype_of_sample,sclc1$SEZ6_group))

sub <- subset(sclc1,cells=WhichCells(sclc1,expression= DLL3_group=="pos" | SEZ6_group=="pos" | CD276_group=="pos"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

sub1 <- subset(sub,subset=SCLC_major_subtype_of_sample=="SCLC-N")
count <- colSums(GetAssayData(object = sub1, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))
## triple negative


sub <- subset(sclc1,subset=SCLC_major_subtype_of_sample=="SCLC-N")

length(WhichCells(sub,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sub, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
table(count)/sum(table(count))

### triple negative for all

length(WhichCells(sclc,expression= DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="neg"))

count <- colSums(GetAssayData(object = sclc, layer = "data")[c("DLL3","CD276","SEZ6"),]>0)
table(count)
#     0     1     2     3 
# 11907 22387 11948  3000 
table(count)/sum(table(count))
#         0         1         2         3 
# 0.2418058 0.4546322 0.2426384 0.0609236 




```




## step 14: additional figure for manuscript

```{r}
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))


FeaturePlot(sclc,features="DLL3")

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_by_DLL3_maligOnly.pdf")))


FeaturePlot(sclc,features="SEZ6") 

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_by_SEZ6_maligOnly.pdf")))


FeaturePlot(sclc,features="CD276")

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_by_CD276_maligOnly.pdf")))


FeaturePlot3(sclc, feature.1 = "DLL3", feature.2 = "SEZ6", feature.3 = "CD276", pt.size = 1)

ggsave(here(paste0("../../plots/",Sys.Date(),"_UMAP_by_3targets_maligOnly.pdf")))




### density plot for defining positive cells

genes_to_plot <- c("DLL3","SEZ6","CD276","donor_id")

## normalized data
# 1. Fetch data
data_to_plot <- FetchData(sclc, vars = c(genes_to_plot))

# 2. Prepare data for ggplot2
data_long <- pivot_longer(data_to_plot, cols = all_of(genes_to_plot[1:3]), names_to = "gene", values_to = "expression")

# 3. Create ggplot2 density plot
ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  facet_wrap(~ donor_id) +
  labs(title = "Gene Expression Density", x = "Expression (log-normalized)", y = "Density") +
  scale_fill_brewer(palette = "Set2") +
  scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_3targets_maligOnly_splitBy_Sample.pdf")))


### dentisy of raw UMI counts


# 1. Fetch data
data_to_plot <- FetchData(sclc, vars = c(genes_to_plot),layer="count")

# 2. Prepare data for ggplot2
data_long <- pivot_longer(data_to_plot, cols = all_of(genes_to_plot[1:3]), names_to = "gene", values_to = "expression")

# 3. Create ggplot2 density plot
ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  facet_wrap(~ donor_id) +
  labs(title = "Gene Expression Density", x = "Expression (raw #UMI)", y = "Density") +
  scale_fill_brewer(palette = "Set2") +
  scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_3targets_maligOnly_splitBy_Sample_rawUMI.pdf")))




```


## Step 15:  redefine positive cells with raw UMI cutoffs


```{r}

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

ucuts <- c(1,2,3,5)
subtypes <- unique(sclc$SCLC_major_subtype_of_sample)
for (subtype in subtypes){
  for (cut in ucuts){
    epi_tmp <- subset(sclc,subset=SCLC_major_subtype_of_sample==subtype)
    SEZ6_pos <- WhichCells(epi_tmp, expression=SEZ6 >=cut,slot="count")
    B7H3_pos <- WhichCells(epi_tmp, expression=CD276 >=cut,slot="count")
    DLL3_pos <- WhichCells(epi_tmp, expression=DLL3 >=cut,slot="count") 
    
    a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
      
    p <- plot(euler(a, shape = "ellipse"), quantities = TRUE)
    g <- grid.arrange(grobs=list(p),top=paste0(subtype,"(>=",cut,"UMI)"))
    ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",subtype,"_",cut,"UMI.pdf")),g)
      
  }
}


```



## Step 16: finalize figures on 2025-06-30
```{r}

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

## 3G: bell shaped heterogeneity of DLL3 in tumor cells in Cancer Cell



genes_to_plot <- c("DLL3","donor_id","SCLC_major_subtype_of_sample")

## normalized data

# 1. Fetch data
data_to_plot <- FetchData(sclc, vars = c(genes_to_plot))

# 2. Prepare data for ggplot2
data_long <- pivot_longer(data_to_plot, cols = all_of(genes_to_plot[1]), names_to = "gene", values_to = "expression")

# 3. Create ggplot2 density plot
ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  facet_wrap(~ donor_id) +
  labs(title = "DLL3 Expression Density", x = "Expression (log-normalized)", y = "Density") #+
  #scale_fill_brewer(palette = "Set2") +
  #scale_color_brewer(palette = "Set2")

## figure S3C
ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_splitBy_Sample.pdf")))


## S3B
## all samples together

ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  #facet_wrap(~ sample) +
  labs(title = "DLL3 Expression Density", x = "Expression (log-normalized)", y = "Density") #+
  #scale_fill_brewer(palette = "Set2") +
  #scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_pooledSample.pdf")))


## S3C: by subtype

ggplot(data_long, aes(x = expression, color = gene, fill = gene)) +
  geom_density(alpha = 0.3) +
  theme_bw() +
  facet_wrap(~ SCLC_major_subtype_of_sample) +
  labs(title = "DLL3 Expression Density", x = "Expression (log-normalized)", y = "Density") #+
  #scale_fill_brewer(palette = "Set2") +
  #scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_splitBy_subtype.pdf")))


## 7C: A wants the same size for circles

## cells with DLL3

DLL3_pos <- WhichCells(sclc, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(sclc, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(sclc, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

ggvenn(a)

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_sameSize.pdf")))


### SCLC-A

sclc1 <- subset(sclc,subset=SCLC_major_subtype_of_sample=="SCLC-A")
DLL3_pos <- WhichCells(sclc1, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(sclc1, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(sclc1, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

ggvenn(a)

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_sameSize_SCLC-A.pdf")))

### SCLC-N
sclc1 <- subset(sclc,subset=SCLC_major_subtype_of_sample=="SCLC-N")
DLL3_pos <- WhichCells(sclc1, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(sclc1, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(sclc1, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

ggvenn(a)

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_sameSize_SCLC-N.pdf")))


### SCLC-P

sclc1 <- subset(sclc,subset=SCLC_major_subtype_of_sample=="SCLC-P")
DLL3_pos <- WhichCells(sclc1, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(sclc1, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(sclc1, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

ggvenn(a)

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_sameSize_SCLC-P.pdf")))




```


## Step 17: check if large co-expression variation exist across individuals

```{r}
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

for (samp in unique(sclc$donor_id)){
  epi_tmp <- subset(sclc,subset=donor_id==samp)
  DLL3_pos <- WhichCells(epi_tmp, expression=DLL3_group=="pos")
  SEZ6_pos <- WhichCells(epi_tmp, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(epi_tmp, expression=CD276_group=="pos") 
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
    
  p <- plot(euler(a, shape = "circle"), quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=samp)
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",samp,".pdf")),g)
    
}


## plot subtype composition

samType <- as.data.frame(table(sclc$donor_id,sclc$SCLC_subtype))
colnames(samType) <- c("Sample","Subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(Subtype == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = Subtype)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_SCLC_subtype_per_sample.pdf")))


## plot for all subtypes
df_st <- data.frame(unclass(table(sclc$donor_id, sclc$SCLC_subtype)/rowSums(table(sclc$donor_id, sclc$SCLC_subtype))))

colnames(df_st) <- c("SCLC_A","SCLC_N","SCLC_P")

df_st$sample <- rownames(df_st)

## get fractions of DLL3 positive cells per sample

df_dl <- data.frame(unclass(table(sclc$donor_id, sclc$DLL3_group)/rowSums(table(sclc$donor_id, sclc$DLL3_group))))

df_dl$sample <- rownames(df_dl)

df_dl <- df_dl[,c("sample","pos")]

colnames(df_dl) <- c("sample","DLL3")

## join

df <- merge(df_st,df_dl)

# List of x variables
x_vars <- c("SCLC_A", "SCLC_N", "SCLC_P")

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  sp <- ggscatter(df, x = x, y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE
  ) + xlab(paste0("%", x)) + ylab("%DLL3+")
  
  # Add correlation
  p <- sp + stat_cor(method = "pearson")
  
  # Store plot in list with name
  plot_list[[x]] <- p
}
 
ggarrange(plotlist = plot_list, ncol = 2, nrow = 2)

ggsave(here(paste0("../../plots/",Sys.Date(),"_pearson_corr_DLL3_by_subtype.pdf")))




### only main sutype is -A

epi_sub <- subset(sclc,subset=SCLC_major_subtype_of_sample=="SCLC-A")

## plot subtype composition

samType <- as.data.frame(table(epi_sub$donor_id,epi_sub$SCLC_subtype))
colnames(samType) <- c("Sample","Subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

## remove NaN samples

samType <- na.omit(samType)

reordered_data <- samType %>%
  filter(Subtype == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = Subtype)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_SCLC_subtype_per_Asample.pdf")))


## calculate correlation of subtype-A cells and %DLL3+
df_st <- data.frame(unclass(table(epi_sub$donor_id, epi_sub$SCLC_subtype)/rowSums(table(epi_sub$donor_id, epi_sub$SCLC_subtype))))

colnames(df_st) <- c("SCLC_A","SCLC_N","SCLC_P")

df_st$sample <- rownames(df_st)

## get fractions of DLL3 positive cells per sample

df_dl <- data.frame(unclass(table(epi_sub$donor_id, epi_sub$DLL3_group)/rowSums(table(epi_sub$donor_id, epi_sub$DLL3_group))))

df_dl$sample <- rownames(df_dl)

df_dl <- df_dl[,c("sample","pos")]

colnames(df_dl) <- c("sample","DLL3")

## join

df <- merge(df_st,df_dl)


sp <- ggscatter(df, x = "SCLC_A", y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE
  ) + xlab("%SCLC-A") + ylab("%DLL3+")
  
  # Add correlation
sp + stat_cor(method = "pearson")

ggsave(here(paste0("../../plots/",Sys.Date(),"_pearson_corr_DLL3_and_SCLC_A.pdf")))

na.omit(df$SCLC_A)

#  [1] 0.9398340 0.8823529 0.9961876 0.9963913 0.9997109 0.9902507 0.9927087 0.9973536 0.9922041 0.8601399
# [11] 0.9961415 0.9887640 0.9404568

summary(na.omit(df$SCLC_A))
 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.8601  0.9405  0.9922  0.9671  0.9962  0.9997

# List of x variables
x_vars <- c("SCLC_A", "SCLC_N", "SCLC_P")

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  sp <- ggscatter(df, x = x, y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE
  ) + xlab(paste0("%", x)) + ylab("%DLL3+")
  
  # Add correlation
  p <- sp + stat_cor(method = "pearson")
  
  # Store plot in list with name
  plot_list[[x]] <- p
}
 
ggarrange(plotlist = plot_list, ncol = 2, nrow = 2)

ggsave(here(paste0("../../plots/",Sys.Date(),"_pearson_corr_DLL3_by_subtype_in_MainA.pdf")))

### spearman

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  sp <- ggscatter(df, x = x, y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE
  ) + xlab(paste0("%", x)) + ylab("%DLL3+")
  
  # Add correlation
  p <- sp + stat_cor(method = "spearman",cor.coef.name="rho")
  
  # Store plot in list with name
  plot_list[[x]] <- p
}
 
ggarrange(plotlist = plot_list, ncol = 2, nrow = 2)

ggsave(here(paste0("../../plots/",Sys.Date(),"_spearman_corr_DLL3_by_subtype_in_MainA.pdf")))




#### reannotate using the same gene sets as MGH

gene_sets <- list(
  pASCL1 = c("ASCL1", "SOX4", "STMN2", "DOC2A"),
  pNEUROD1 = c("NEUROD1", "ADCYAP1","NRXN1", "SSTR2", "ID1", "ID3", "SST", "DLK1"),
  pPOU2F3 = c("POU2F3", "ASCL2", "CD44", "MYC", "KIT", "YBX1"),
  pInflamed= c("CD274","PDCD1","CD80","CD86","CTLA4","CD38","IDO1","TIGIT","ICOS","LAG3","CCL5","CXCL10",
              "HLA-A","HLA-B","HLA-C","HLA-E","HLA-F","HLA-G","HLA-DRA","HLA-DRB1","HLA-DQA1","HLA-DMA","HLA-DPA1",
              "HLA-DMB","HLA-DPB1","HLA-DQB1","HLA-DOA","HLA-DOB")
)

sclc <- AddModuleScore_UCell(sclc, features = gene_sets, ncores=4, name = "")

### classify individual cells for max

# Get the score matrix (remove the "1" suffix added by AddModuleScore)
score_matrix <- FetchData(sclc, vars = names(gene_sets))
colnames(score_matrix) <- c("SCLC-A","SCLC-N","SCLC-P","SCLC-I")

# Assign cell type based on highest score
sclc$SCLC_subtype_UCell <- colnames(score_matrix)[max.col(score_matrix)]

saveRDS(sclc,paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

## plot subtype composition

samType <- as.data.frame(table(sclc$donor_id,sclc$SCLC_subtype_UCell))
colnames(samType) <- c("Sample","Subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

reordered_data <- samType %>%
  filter(Subtype == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = Subtype)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_SCLC_4subtype_per_sample.pdf")))

### only main sutype is -A

epi_sub <- subset(sclc,subset=SCLC_major_subtype_of_sample=="SCLC-A")

## plot subtype composition

samType <- as.data.frame(table(epi_sub$donor_id,epi_sub$SCLC_subtype_UCell))
colnames(samType) <- c("Sample","Subtype","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100

## remove NaN samples

samType <- na.omit(samType)

reordered_data <- samType %>%
  filter(Subtype == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

samType$Sample <- factor(samType$Sample,levels=reordered_data$Sample)

ggplot(samType, aes(x = Sample, y = Value, fill = Subtype)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_y_continuous(labels = scales::percent,expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1)) +
  labs(title = "Subtype per sample", y = "Proportion", x = "")
ggsave(here(paste0("../../plots/",Sys.Date(),"_stackBar_SCLC_subtype_per_Asample.pdf")))


## calculate correlation of subtype-A cells and %DLL3+
df_st <- data.frame(unclass(table(epi_sub$donor_id, epi_sub$SCLC_subtype_UCell)/rowSums(table(epi_sub$donor_id, epi_sub$SCLC_subtype_UCell))))

colnames(df_st) <- c("SCLC_A","SCLC_I","SCLC_N","SCLC_P")

df_st$sample <- rownames(df_st)

## get fractions of DLL3 positive cells per sample

df_dl <- data.frame(unclass(table(epi_sub$donor_id, epi_sub$DLL3_group)/rowSums(table(epi_sub$donor_id, epi_sub$DLL3_group))))

df_dl$sample <- rownames(df_dl)

df_dl <- df_dl[,c("sample","pos")]

colnames(df_dl) <- c("sample","DLL3")

## join

df <- merge(df_st,df_dl)


sp <- ggscatter(df, x = "SCLC_A", y = "DLL3",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE
  ) + xlab("%SCLC-A") + ylab("%DLL3+")
  
  # Add correlation
sp + stat_cor(method = "pearson")

ggsave(here(paste0("../../plots/",Sys.Date(),"_pearson_corr_DLL3_and_SCLC_A_UCell.pdf")))

na.omit(df$SCLC_A)

#  [1] 0.8278008 0.9764706 0.8680900 0.9855654 0.8975718 0.9958217 0.9919796 0.9910685 0.9943303
# [10] 0.8601399 0.9861736 0.8876404 0.9437194

summary(na.omit(df$SCLC_A))
 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.8278  0.8876  0.9765  0.9390  0.9911  0.9958 

###

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

df <- data.frame(unclass(table(sclc$donor_id, sclc$DLL3_group)/rowSums(table(sclc$donor_id, sclc$DLL3_group))))

df$sample <- rownames(df)

df <- df[,c("sample","pos")]

colnames(df) <- c("sample","DLL3")



ggplot(df, aes(x = DLL3)) +
  geom_density() +
  theme_bw() +
  #facet_wrap(~ sample) +
  labs(title = "%DLL3 + Density", x = "%DLL3+", y = "Density") #+
  #scale_fill_brewer(palette = "Set2") +
  #scale_color_brewer(palette = "Set2")

ggsave(here(paste0("../../plots/",Sys.Date(),"_densityPlot_by_DLL3_pos_fraction.pdf")))


# Histogram 
ggplot(df, aes(x = DLL3)) + 
  geom_histogram() #+geom_density(aes(y=..count..))

ggsave(here(paste0("../../plots/",Sys.Date(),"_hist_by_DLL3_pos_fraction.pdf")))

### histogram with curve

ggplot(df, aes(x = DLL3)) + 
  geom_histogram() +geom_density(aes(y=..density..)) + theme_bw()


ggsave(here(paste0("../../plots/",Sys.Date(),"_histWithCurve_by_DLL3_pos_fraction.pdf")))


### only subtype-A

sclc_sub <- subset(sclc, subset=SCLC_major_subtype_of_sample=="SCLC-A")

df <- data.frame(unclass(table(sclc_sub$donor_id, sclc_sub$DLL3_group)/rowSums(table(sclc_sub$donor_id, sclc_sub$DLL3_group))))

df$sample <- rownames(df)

df <- df[,c("sample","pos")]

colnames(df) <- c("sample","DLL3")

### histogram with curve

ggplot(df, aes(x = DLL3)) + 
  geom_histogram() +geom_density(aes(y=..density..)) + theme_bw()


ggsave(here(paste0("../../plots/",Sys.Date(),"_histWithCurve_by_DLL3_pos_fraction_subtypeA_only.pdf")))


### % DLL3


inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))


df <- data.frame(unclass(table(sclc$donor_id,sclc$DLL3_group)/rowSums(table(sclc$donor_id,sclc$DLL3_group))))

summary(df$pos)

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.0814  0.5438  0.6360  0.6521  0.8265  0.8951 

## -N
summary(df$pos[which(rownames(df) %in% c("RU779","RU1181","RU1215","RU1231","RU1293"))])

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.0814  0.4872  0.6193  0.5336  0.6270  0.8530 

## -P

df$pos[which(rownames(df) == "RU1322")]
# [1] 0.53524

##-A

summary(df$pos[which(!rownames(df) %in% c("RU779","RU1181","RU1215","RU1231","RU1293","RU1322"))])

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.4847  0.6212  0.6660  0.7067  0.8536  0.8951 

## get UMI count per million

## get CPM
epi_cpm <- NormalizeData(sclc, assay = "RNA", normalization.method = "RC", scale.factor = 1000000)

##Error: Attempting to add a different number of cells and/or features
## recreate figures
data <- CreateAssayObject(counts = sclc[["RNA"]]@counts)

epi_cpm <- NormalizeData(data, assay = "RNA", normalization.method = "RC", scale.factor = 1000000)

DLL3_CPM <- epi_cpm$data["DLL3",which(epi_cpm$data["DLL3",]>0)]
#> range(DLL3_CPM)
#[1]   14.44022 3000.07500

## treatment
table(sclc$sample,sclc$treatment2)

## 7 naive and 12 treated

## %DLL3 by treatment
## naive
summary(df$pos[which(rownames(df) %in% c("RU426","RU1066","RU1145","RU1152","RU1215","RU1229","RU1231"))])

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.4872  0.5802  0.6360  0.6849  0.8076  0.8951 

## treated
summary(df$pos[which(!rownames(df) %in% c("RU426","RU1066","RU1145","RU1152","RU1215","RU1229","RU1231"))])

 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.0814  0.5481  0.6421  0.6330  0.8133  0.8661 



#####
## 7C: circles but different sizes

## cells with DLL3

DLL3_pos <- WhichCells(sclc, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(sclc, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(sclc, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

p <- plot(euler(a, shape = "circle"), fills = list(fill = c("red","orange", "steelblue4")),quantities = TRUE)
g <- grid.arrange(grobs=list(p))

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize.pdf")),g)


### subtypes
for (st in unique(sclc$SCLC_major_subtype_of_sample)){
  sclc1 <- subset(sclc,subset=SCLC_major_subtype_of_sample==st)
  DLL3_pos <- WhichCells(sclc1, expression=DLL3_group=="pos")
  SEZ6_pos <- WhichCells(sclc1, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(sclc1, expression=CD276_group=="pos")
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
  
  p <- plot(euler(a, shape = "circle"), fills = list(fill = c("red","orange", "steelblue4")),quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=st)
  
  ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize_",st,".pdf")),g)
}

### include universe 

## cells with DLL3

DLL3_pos <- WhichCells(sclc, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(sclc, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(sclc, expression=CD276_group=="pos")
total <- colnames(sclc)

a <- list(`negative`=total,`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

p <- plot(euler(a, shape = "circle"), fills = list(fill = c("gray","red","orange", "steelblue4")),quantities = TRUE)
g <- grid.arrange(grobs=list(p))

ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize_allIn.pdf")),g)


### subtypes
for (st in unique(sclc$SCLC_major_subtype_of_sample)){
  sclc1 <- subset(sclc,subset=SCLC_major_subtype_of_sample==st)
  DLL3_pos <- WhichCells(sclc1, expression=DLL3_group=="pos")
  SEZ6_pos <- WhichCells(sclc1, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(sclc1, expression=CD276_group=="pos")
  total <- colnames(sclc1)
  
  a <- list(`negative`=total,`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
  
  p <- plot(euler(a, shape = "circle"), fills = list(fill = c("gray","red","orange", "steelblue4")),quantities = TRUE)
  g <- grid.arrange(grobs=list(p),top=st)
  
  ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize_",st,"_allIn.pdf")),g)
}


####### breakdown

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

### breakdown of positive cells by treatment
## at single cell level
### loop through subtypes at sample level and treatment
for (st in unique(sclc$SCLC_major_subtype_of_sample)){
  for (treat in unique(sclc$treatment2)){
    try({
    epi_sub <- subset(sclc,subset=SCLC_major_subtype_of_sample==st)
    epi_sub_treatment <- subset(epi_sub,subset=treatment2==treat)
    
    
    DLL3_pos <- WhichCells(epi_sub_treatment, expression=DLL3_group=="pos")
    SEZ6_pos <- WhichCells(epi_sub_treatment, expression=SEZ6_group=="pos")
    B7H3_pos <- WhichCells(epi_sub_treatment, expression=CD276_group=="pos")
    total <- colnames(epi_sub_treatment)
    
    a <- list(`negative`=total,`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
    
    p <- plot(euler(a, shape = "circle"), fills = list(fill = c("gray","red","orange", "steelblue4")),quantities = TRUE)
    g <- grid.arrange(grobs=list(p))
    
    ggsave(here(paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_circleDifferentSize_allIn_",st,"_",treat,".pdf")),g)

    })
  }
  
  
}

## prop.tests ###

## SCLC-A
## within cohort CC
## DLL3+ using all cells in each group as total
res <- prop.test(x=c(8198,13598),n=c(11393,17132))
res$p.value

## SEZ6+
res <- prop.test(x=c(3010,5015),n=c(11393,17132))
res$p.value

## B7H3+
res <- prop.test(x=c(3437,2803),n=c(11393,17132))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(2488,4224),n=c(8198,13598))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(2833,2250),n=c(8198,13598))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(991,868),n=c(8198,13598))
res$p.value

### between cohorts
## naive
#DLL3+ 
res <- prop.test(x=c(4278,8198),n=c(14660,11393))
res$p.value

## SEZ6+
res <- prop.test(x=c(3921,3010),n=c(14660,11393))
res$p.value

## B7H3+
res <- prop.test(x=c(3277,3437),n=c(14660,11393))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(1520,2488),n=c(4278,8198))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(1274,2833),n=c(4278,8198))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(606,184),n=c(4278,8198))
res$p.value

## treated

#DLL3+ 
res <- prop.test(x=c(1930,13598),n=c(5160,17132))
res$p.value

## SEZ6+
res <- prop.test(x=c(1221,5015),n=c(5160,17132))
res$p.value

## B7H3+
res <- prop.test(x=c(884,2803),n=c(5160,17132))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(657,4224),n=c(1930,13598))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(478,2250),n=c(1930,13598))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(184,868),n=c(1930,13598))
res$p.value

## total

#DLL3+ 
res <- prop.test(x=c(6208,21796),n=c(19820,28525))
res$p.value

## SEZ6+
res <- prop.test(x=c(5142,8025),n=c(19820,28525))
res$p.value

## B7H3+
res <- prop.test(x=c(4161,6240),n=c(19820,28525))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(2177,6712),n=c(6208,21796))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(1752,5083),n=c(6208,21796))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(790,1859),n=c(6208,21796))
res$p.value


### SCLC-N

## naive VS treated

#DLL3+ 
res <- prop.test(x=c(4646,4002),n=c(8473,9307))
res$p.value

## SEZ6+
res <- prop.test(x=c(1276,4392),n=c(8473,9307))
res$p.value

## B7H3+
res <- prop.test(x=c(1124,1824),n=c(8473,9307))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(734,2511),n=c(4646,4002))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(707,1366),n=c(4646,4002))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(129,1008),n=c(4646,4002))
res$p.value

## -A vs -N

## naive

#DLL3+ 
res <- prop.test(x=c(8198,4646),n=c(11393,8473))
res$p.value

## SEZ6+
res <- prop.test(x=c(3010,1276),n=c(11393,8473))
res$p.value

## B7H3+
res <- prop.test(x=c(3437,1124),n=c(11393,8473))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(2488,734),n=c(8198,4646))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(2833,707),n=c(8198,4646))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(991,129),n=c(8198,4646))
res$p.value

## treated

#DLL3+ 
res <- prop.test(x=c(13598,4002),n=c(17132,9307))
res$p.value

## SEZ6+
res <- prop.test(x=c(5015,4392),n=c(17132,9307))
res$p.value

## B7H3+
res <- prop.test(x=c(2803,1824),n=c(17132,9307))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(4224,2511),n=c(13598,4002))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(2250,1366),n=c(13598,4002))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(868,1008),n=c(13598,4002))
res$p.value

## total

#DLL3+ 
res <- prop.test(x=c(21796,8648),n=c(28525,17780))
res$p.value

## SEZ6+
res <- prop.test(x=c(8025,5668),n=c(28525,17780))
res$p.value

## B7H3+
res <- prop.test(x=c(6240,2948),n=c(28525,17780))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(6712,3245),n=c(21796,8648))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(5083,2073),n=c(21796,8648))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(1859,1137),n=c(21796,8648))
res$p.value


### -A -vs -P

## treated
#DLL3+ 
res <- prop.test(x=c(13598,1572),n=c(17132,2937))
res$p.value

## SEZ6+
res <- prop.test(x=c(5015,44),n=c(17132,2937))
res$p.value

## B7H3+
res <- prop.test(x=c(2803,342),n=c(17132,2937))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(4224,24),n=c(13598,1572))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(2250,199),n=c(13598,1572))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(868,4),n=c(13598,1572))
res$p.value

### -N vs -P
## treated
#DLL3+ 
res <- prop.test(x=c(4002,1572),n=c(9307,2937))
res$p.value

## SEZ6+
res <- prop.test(x=c(4392,44),n=c(9307,2937))
res$p.value

## B7H3+
res <- prop.test(x=c(1824,342),n=c(9307,2937))
res$p.value

## SEZ6+&DLL3+ using DLL3+ as total
res <- prop.test(x=c(2511,24),n=c(4002,1572))
res$p.value

## B7H3+&DLL3+
res <- prop.test(x=c(1366,199),n=c(4002,1572))
res$p.value

## SEZ6+&B7H3+&DLL3+
res <- prop.test(x=c(1008,4),n=c(4002,1572))
res$p.value



#############
## at sample level
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))
### create a dataframe to store these following metrics:

# sample identity
# cohort
# subtype info
# treatment
# %DLL3+
# %SEZ6+
# %B7H3+
# %SEZ6+ in DLL3+
# %B7H3+ in DLL3+
# %SEZ6+&B7H3+ in DLL3+
# any of three target +

df_sample <-  data.frame(
  Sample = character(),
  Cohort = character(),
  Subtype = character(),
  Treatment = character(),
  Pdll3 = numeric(),
  Psez6 = numeric(),
  Pb7h3 = numeric(),
  Psind = numeric(),
  Pbind = numeric(),
  P2ind = numeric(),
  Pany3 = numeric(),
  stringsAsFactors = FALSE
)

for (samp in unique(sclc$donor_id)){
  epi_sub <- subset(sclc, subset=donor_id==samp)
  subtype <- unique(epi_sub$SCLC_major_subtype_of_sample)
  treat <- unique(epi_sub$treatment2)
  
  ## total cell
  tcell <- ncol(epi_sub)
  
  ## DLL3+
  
  cdll3 <- length(which(epi_sub$DLL3_group=="pos"))
  
  tmp_Pdll3 <- cdll3/tcell
  
  ## SEZ6+
  csez6 <- length(which(epi_sub$SEZ6_group=="pos"))
  
  tmp_Psez6 <- csez6/tcell  
  
  ## B7H3+
  cb7h3<- length(which(epi_sub$CD276_group=="pos"))
  
  tmp_Pb7h3 <- cb7h3/tcell  
  
  
  ## SEZ6+ in DLL3+
  
  csind <- length(which(epi_sub$DLL3_group=="pos" & epi_sub$SEZ6_group=="pos"))
  
  tmp_Psind <- csind/cdll3
  
  ## B7H3+ in DLL3+
  
  cbind <- length(which(epi_sub$DLL3_group=="pos" & epi_sub$CD276_group=="pos"))
  
  tmp_Pbind <- cbind/cdll3
  
  ## SEZ6+&B7H3+ in DLL3+
  
  c2ind <- length(which(epi_sub$DLL3_group=="pos" & epi_sub$CD276_group=="pos" & epi_sub$SEZ6_group=="pos"))
  tmp_P2ind <- c2ind/cdll3
  
  ## positive for any of the three targets
  cany3 <- length(which(epi_sub$DLL3_group=="pos" | epi_sub$SEZ6_group=="pos" | epi_sub$CD276_group=="pos" ))
  tmp_Pany3 <- cany3/tcell
  
   # Create a data frame for the new row
  new_row <- data.frame(Sample = samp, 
                        Cohort = "CancerCell", 
                        Subtype = subtype, 
                        Treatment = treat, 
                        Pdll3 =tmp_Pdll3,
                        Psez6 =tmp_Psez6,
                        Pb7h3 = tmp_Pb7h3,
                        Psind = tmp_Psind,
                        Pbind = tmp_Pbind,
                        P2ind = tmp_P2ind,
                        Pany3 = tmp_Pany3,
                        stringsAsFactors = FALSE)

  # Add the new row to the data frame using rbind()
  df_sample <- rbind(df_sample, new_row)
}

write.xlsx(df_sample,paste0("../../tables/",Sys.Date(),"_coexpression_summary_at_sample_level.xlsx"))

##wilcoxon rank sum tests

df_mgh <- read.xlsx("../../../moshe/tables/2025-07-08_coexpression_summary_at_sample_level.xlsx")

## between treatment within cohort

## SCLC-A

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## same p value due to same sum of ranks for naive

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")])
res$p.value

### SCLC-N

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")], df_sample$Pdll3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")], df_sample$Psez6[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")], df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")], df_sample$Psind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")], df_sample$Pbind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")], df_sample$P2ind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value


### -A vs -N
## Naive 
## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pdll3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Psez6[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Psind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$Pbind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Naive")], df_sample$P2ind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Naive")])
res$p.value

## Treated

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Pdll3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Psez6[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Psind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Pbind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$P2ind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")])
res$p.value

## total

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" )], df_sample$Pdll3[which(df_sample$Subtype=="SCLC-N" )])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" )], df_sample$Psez6[which(df_sample$Subtype=="SCLC-N")])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" )], df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-N" )])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" )], df_sample$Psind[which(df_sample$Subtype=="SCLC-N" )])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" )], df_sample$Pbind[which(df_sample$Subtype=="SCLC-N" )])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A")], df_sample$P2ind[which(df_sample$Subtype=="SCLC-N" )])
res$p.value

## -A vs -P

## treated
## Treated

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Pdll3[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Psez6[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Psind[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$Pbind[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-A" & df_sample$Treatment=="Treated")], df_sample$P2ind[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## -N vs -P

## treated

## DLL3+
res <- wilcox.test(df_sample$Pdll3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")], df_sample$Pdll3[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+
res <- wilcox.test(df_sample$Psez6[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")], df_sample$Psez6[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+
res <- wilcox.test(df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")], df_sample$Pb7h3[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+ in DLL3+
res <- wilcox.test(df_sample$Psind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")], df_sample$Psind[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## B7H3+ in DLL3+
res <- wilcox.test(df_sample$Pbind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")], df_sample$Pbind[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## SEZ6+&B7H3+ in DLL3+
res <- wilcox.test(df_sample$P2ind[which(df_sample$Subtype=="SCLC-N" & df_sample$Treatment=="Treated")], df_sample$P2ind[which(df_sample$Subtype=="SCLC-P" & df_sample$Treatment=="Treated")])
res$p.value

## sequencing depth

## compare SCLC-A between cohort

## any3+

res <- wilcox.test(df_sample$Pany3[which(df_sample$Subtype=="SCLC-A")], df_mgh$Pany3[which(df_mgh$Subtype=="SCLC-A" )])
res$p.value
# [1] 0.0004702526




### replot Figure 2B to show two stacked bar for each sample split by subtype


inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

## DLL3
samType <- as.data.frame(table(sclc$donor_id,sclc$DLL3_group))
colnames(samType) <- c("Sample","DLL3","Freq")
samType$Value <- samType$Freq / ave(samType$Freq, samType$Sample, FUN = sum) * 100
samType$fraction_type <- "DLL3"
colnames(samType)[2] <- "group"


## subtype

samType1 <- as.data.frame(table(sclc$donor_id,sclc$SCLC_subtype))
colnames(samType1) <- c("Sample","cellType","Freq")
samType1$Value <- samType1$Freq / ave(samType1$Freq, samType1$Sample, FUN = sum) * 100
samType1$fraction_type <- "subtype"
colnames(samType1)[2] <- "group"

df1 <- rbind(samType,samType1)

## get main subtype
samType2 <- as.data.frame(table(sclc$donor_id,sclc$SCLC_major_subtype_of_sample))
samType2 <- samType2[which(samType2$Freq >0),1:2]
colnames(samType2) <- c("Sample","Subtype")

df <- merge(df1,samType2)


levels(df$Sample)[(levels(df$Sample)=="PleuralEffusion")] <- "Pleural"

reordered_data <- df %>%
  filter(group == "SCLC-A") %>%
  arrange(Value) %>%
  mutate(Sample = factor(Sample, levels = Sample))

df$Sample <- factor(df$Sample,levels=reordered_data$Sample)

df$fraction_type <- factor(df$fraction_type,levels=c("subtype","DLL3"))

df$group <- factor(df$group,levels=c("SCLC-A","SCLC-N","SCLC-P","neg","pos"))


p1 <- ggplot(df[df$Subtype=="SCLC-P",], aes(x = fraction_type, y = Value, fill = group)) + geom_bar(stat = "identity", position = "stack") + facet_grid(~ factor(Sample), switch = "x") + theme(strip.placement = "outside", strip.background = element_rect(fill = NA, color = "white"), strip.text.x.bottom=element_text(angle = 90), panel.spacing = unit(-.01,"cm"), plot.margin = unit(c(0, 0, 0, 0), "cm"),axis.text.x = element_text(angle = 45, hjust = 1),axis.text.y=element_blank(),axis.ticks.y=element_blank()) + labs(x="", y="",title="-P") +  
  scale_fill_manual(values = c("pos"= "red",
                                "neg" = "blue",
                                "SCLC-A" = "sandybrown", 
                               "SCLC-N" = "#56B4E9", 
                               "SCLC-P" = "forestgreen" 
                               )) + 
  scale_y_continuous(expand = expansion(mult = 0, add = 0)) 

p2 <- ggplot(df[df$Subtype=="SCLC-N",], aes(x = fraction_type, y = Value, fill = group)) + geom_bar(stat = "identity", position = "stack") + facet_grid(~ factor(Sample), switch = "x") + theme(strip.placement = "outside", strip.background = element_rect(fill = NA, color = "white"), strip.text.x.bottom=element_text(angle = 90), panel.spacing = unit(-.01,"cm"), axis.text.x = element_text(angle = 45, hjust = 1),plot.margin = unit(c(0, 0, 0, 0), "cm"),axis.text.y=element_blank(),axis.ticks.y=element_blank()) + labs(x="", y="",title="SCLC-N") +  
  scale_fill_manual(values = c("pos"= "red",
                                "neg" = "blue",
                                "SCLC-A" = "sandybrown", 
                               "SCLC-N" = "#56B4E9", 
                               "SCLC-P" = "forestgreen" 
                               )) + 
  scale_y_continuous(expand = expansion(mult = 0, add = 0)) 


p3 <- ggplot(df[df$Subtype=="SCLC-A",], aes(x = fraction_type, y = Value, fill = group)) + geom_bar(stat = "identity", position = "stack") + facet_grid(~ factor(Sample), switch = "x") + theme(strip.placement = "outside", strip.background = element_rect(fill = NA, color = "white"), strip.text.x.bottom=element_text(angle = 90), panel.spacing = unit(-.01,"cm"), axis.text.x = element_text(angle = 45, hjust = 1),plot.margin = unit(c(0, 0, 0, 0), "cm")) + labs(x="", title="SCLC-A", y="percent") +  
  scale_fill_manual(values = c("pos"= "red",
                                "neg" = "blue",
                                "SCLC-A" = "sandybrown", 
                               "SCLC-N" = "#56B4E9", 
                               "SCLC-P" = "forestgreen" 
                               )) + 
  scale_y_continuous(expand = expansion(mult = 0, add = 0)) 

ggarrange(p3,p2,p1, ncol=3,common.legend=TRUE, legend="bottom",widths=c(2.4,1,0.3))
#plot_grid(p3, p2, p1,labels = c("SCLC-A","SCLC-N", "SCLC-P"), nrow = 1)

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_DLL3_and_subtype_per_sample_orderBySCLCA_splitByMainSubtype.pdf")))


### 2D specify y scale and increase dot size



df_cc <- read.xlsx("../../tables/2025-07-07_coexpression_summary_at_sample_level.xlsx")
df_cc$Pdll3 <- df_cc$Pdll3*100
ggboxplot(df_cc, x = "Treatment", y = "Pdll3", color = "Treatment", palette=c("blue","red"),
          add = "jitter",add.params = list(size = 2.5,jitter=0.3)) + 
          labs(x="",y="%DLL3+") + theme(legend.position = "right") +
          stat_compare_means(label.x = 1.5,label.y = 90)  +
          scale_y_continuous(breaks=c(0,25,50,75,100), limits=c(0,100),expand = expansion(mult = c(0, 0)))

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_per_sample_by_stage.pdf")))




```



## Step 18: Pseudobulk comparison of target genes


```{r}
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

### all cells in
## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(sclc, assays = "RNA", return.seurat=T,group.by = c("donor_id"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("donor_id","DLL3","SEZ6","CD276"))

## add treatment info

dtreat <- as.data.frame(unclass(table(sclc$donor_id,sclc$treatment2)))

dtreat$treatment <- ifelse(dtreat$Naive >0, "naive","treated")

dtreat$donor_id <- rownames(dtreat)

dtreat <- dtreat[,c("donor_id","treatment")]

## merge

df <- merge(df,dtreat)

### compare DLL3

ggboxplot(df, x = "treatment", y = "DLL3",
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average DLL3") + theme(legend.position = "right") +
          stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_pseudobulk_by_treatment_allCell.pdf")))

### compare SEZ6

ggboxplot(df, x = "treatment", y = "SEZ6",
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average SEZ6") + theme(legend.position = "right") +
          stat_compare_means(label.x=1.4, label.y= 0.6) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_pseudobulk_by_treatment_allCell.pdf")))

### compare B7H3

ggboxplot(df, x = "treatment", y = "CD276",
          color = "treatment", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average CD276") + theme(legend.position = "right") +
          stat_compare_means(label.x=1.4) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_B7H3_pseudobulk_by_treatment_allCell.pdf")))


### add main subtype info

dsubtype <- as.data.frame(unclass(table(sclc$donor_id,sclc$SCLC_major_subtype_of_sample)))

dsubtype$subtype <- ifelse(dsubtype$`SCLC-A` >0, "SCLC-A",ifelse(dsubtype$`SCLC-N`>0,"SCLC-N","SCLC-P"))

dsubtype$donor_id <- rownames(dsubtype)

dsubtype <- dsubtype[,c("donor_id","subtype")]

## merge

df <- merge(df,dsubtype)

### -P is the lowest in all three genes

df2 <- df[which(!df$subtype=="SCLC-P"),]

### compare DLL3 (higher in -A than -N with p ~ 0.1)

ggboxplot(df2, x = "subtype", y = "DLL3",
          color = "subtype", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average DLL3") + theme(legend.position = "right") +
          stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_pseudobulk_by_subtype_allCell.pdf")))

### compare SEZ6

ggboxplot(df2, x = "subtype", y = "SEZ6",
          color = "subtype", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average SEZ6") + theme(legend.position = "right") +
          stat_compare_means(label.x=1.4, label.y= 0.6) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_pseudobulk_by_subtype_allCell.pdf")))

### compare B7H3

ggboxplot(df2, x = "subtype", y = "CD276",
          color = "subtype", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average CD276") + theme(legend.position = "right") +
          stat_compare_means(label.x=1.4) 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_B7H3_pseudobulk_by_subtype_allCell.pdf")))



#### positive cells only ###

## cannot use a loop

## DLL3
  
tmp <- subset(sclc,subset=DLL3_group=="pos")

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(tmp, assays = "RNA", return.seurat=T,group.by = c("donor_id"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("donor_id","DLL3"))

## merge

df <- merge(df,dtreat)

### compare target genes

ggboxplot(df, x = "treatment", y = "DLL3",
        color = "treatment", palette = c("blue","red"),
        add = "jitter") + labs(x="",y=paste0("log-normalized average ","DLL3 (positive only)")) + theme(legend.position = "right") +
        stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_","DLL3","_pseudobulk_by_treatment_posCellOnly.pdf")))

## subtype

df <- merge(df,dsubtype)

### -P is the lowest in all three genes

df2 <- df[which(!df$subtype=="SCLC-P"),]


ggboxplot(df2, x = "subtype", y = "DLL3",
          color = "subtype", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average DLL3  (positive only)") + theme(legend.position = "right") +
          stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_DLL3_pseudobulk_by_subtype_posCellOnly.pdf")))



## SEZ6
  
tmp <- subset(sclc,subset=SEZ6_group=="pos")

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(tmp, assays = "RNA", return.seurat=T,group.by = c("donor_id"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("donor_id","SEZ6"))

## merge

df <- merge(df,dtreat)

### compare target genes

ggboxplot(df, x = "treatment", y = "SEZ6",
        color = "treatment", palette = c("blue","red"),
        add = "jitter") + labs(x="",y=paste0("log-normalized average ","SEZ6 (positive only)")) + theme(legend.position = "right") +
        stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_","SEZ6","_pseudobulk_by_treatment_posCellOnly.pdf")))

## subtype

df <- merge(df,dsubtype)

### -P is the lowest in all three genes

df2 <- df[which(!df$subtype=="SCLC-P"),]


ggboxplot(df2, x = "subtype", y = "SEZ6",
          color = "subtype", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average SEZ6  (positive only)") + theme(legend.position = "right") +
          stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_SEZ6_pseudobulk_by_subtype_posCellOnly.pdf")))


## CD276
  
tmp <- subset(sclc,subset=CD276_group=="pos")

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(tmp, assays = "RNA", return.seurat=T,group.by = c("donor_id"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("donor_id","CD276"))

## merge

df <- merge(df,dtreat)

### compare target genes

ggboxplot(df, x = "treatment", y = "CD276",
        color = "treatment", palette = c("blue","red"),
        add = "jitter") + labs(x="",y=paste0("log-normalized average ","CD276 (positive only)")) + theme(legend.position = "right") +
        stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_","CD276","_pseudobulk_by_treatment_posCellOnly.pdf")))

## subtype

df <- merge(df,dsubtype)

### -P is the lowest in all three genes

df2 <- df[which(!df$subtype=="SCLC-P"),]


ggboxplot(df2, x = "subtype", y = "CD276",
          color = "subtype", palette = c("blue","red"),
          add = "jitter") + labs(x="",y="log-normalized average CD276  (positive only)") + theme(legend.position = "right") +
          stat_compare_means() 

ggsave(here(paste0("../../plots/",Sys.Date(),"_boxplot_by_CD276_pseudobulk_by_subtype_posCellOnly.pdf")))

### no significant difference in either comparison

```

## Step 19: figures for coexpression

```{r}
df_cc <- read.xlsx("../../tables/2025-07-07_coexpression_summary_at_sample_level.xlsx")
## plot two stacked bar side by side
df_cc$`SEZ6+` <- df_cc$Psez6
df_cc$`SEZ6-` <- 1-df_cc$Psez6

df_cc$`B7H3+` <- df_cc$Pb7h3
df_cc$`B7H3-` <- 1-df_cc$Pb7h3


df <- df_cc[,c(1,3,(ncol(df_cc)-3):ncol(df_cc))]

## convert to long

long_df <- melt(df, id.vars = c("Sample","Subtype"), 
                variable.name = "group", 
                value.name = "fraction")
long_df$Sample[long_df$Sample=="PleuralEffusion"] <- "PL"
long_df$Sample <- gsub("RU","",long_df$Sample)

long_df$group <- factor(long_df$group,levels=c("SEZ6-","SEZ6+","B7H3-","B7H3+"))
long_df$gene <- ifelse(startsWith(as.character(long_df$group),"SEZ6"),"SEZ6","B7H3")
long_df$gene <- factor(long_df$gene,levels=c("SEZ6","B7H3"))
### merge two data.frame
## SCLC-A

p1 <- ggplot(long_df[long_df$Subtype=="SCLC-P",], aes(x = gene, y = fraction, fill = group)) + geom_bar(stat = "identity", position = "stack") + facet_grid(~ Sample) + theme(strip.placement = "outside", strip.background = element_rect(fill = NA, color = "white"), strip.text.x.bottom=element_text(angle = 90), panel.spacing = unit(-.01,"cm"), plot.margin = unit(c(0, 0, 0, 0), "cm"),axis.text.x = element_text(angle = 45, hjust = 1),axis.text.y=element_blank(),axis.ticks.y=element_blank()) + labs(x="", y="",title="-P") +
  scale_fill_manual(values = c("SEZ6-"= "springgreen3",
                                "SEZ6+" = "#63B8FF",
                               "B7H3-" = "forestgreen", 
                               "B7H3+" = "#9E59AD")) + 
  scale_y_continuous(expand = expansion(mult = 0, add = 0)) 

p2 <- ggplot(long_df[long_df$Subtype=="SCLC-N",], aes(x = gene, y = fraction, fill = group)) + geom_bar(stat = "identity", position = "stack") + facet_grid(~ Sample) + theme(strip.placement = "outside", strip.background = element_rect(fill = NA, color = "white"), strip.text.x.bottom=element_text(angle = 90), panel.spacing = unit(-.01,"cm"), plot.margin = unit(c(0, 0, 0, 0), "cm"),axis.text.x = element_text(angle = 45, hjust = 1),axis.text.y=element_blank(),axis.ticks.y=element_blank()) + labs(x="", y="",title="SCLC-N") +
  scale_fill_manual(values = c("SEZ6-"= "springgreen3",
                                "SEZ6+" = "#63B8FF",
                               "B7H3-" = "forestgreen", 
                               "B7H3+" = "#9E59AD")) + 
  scale_y_continuous(expand = expansion(mult = 0, add = 0))

p3 <- ggplot(long_df[long_df$Subtype=="SCLC-A",], aes(x = gene, y = fraction, fill = group)) + geom_bar(stat = "identity", position = "stack") + facet_grid(~ Sample) + theme(strip.placement = "outside", strip.background = element_rect(fill = NA, color = "white"), strip.text.x.bottom=element_text(angle = 90), panel.spacing = unit(-.01,"cm"), plot.margin = unit(c(0, 0, 0, 0), "cm"),axis.text.x = element_text(angle = 45, hjust = 1),axis.text.y=element_blank(),axis.ticks.y=element_blank()) + labs(x="", y="",title="SCLC-A") +
  scale_fill_manual(values = c("SEZ6-"= "springgreen3",
                                "SEZ6+" = "#63B8FF",
                               "B7H3-" = "forestgreen", 
                               "B7H3+" = "#9E59AD")) + 
  scale_y_continuous(expand = expansion(mult = 0, add = 0))

ggarrange(p3,p2,p1, ncol=3,common.legend=TRUE, legend="bottom",widths=c(2.4,1,0.3))
#plot_grid(p3, p2, p1,labels = c("SCLC-A","SCLC-N", "SCLC-P"), nrow = 1)

ggsave(here(paste0("../../plots/",Sys.Date(),"_stackedbar_by_SEZ6_B7H3_subtype_per_sample_orderBySCLCA_splitByMainSubtype.pdf")))

### venn diagram for pooled data

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

DLL3_pos <- WhichCells(sclc, expression=DLL3_group=="pos")
SEZ6_pos <- WhichCells(sclc, expression=SEZ6_group=="pos")
B7H3_pos <- WhichCells(sclc, expression=CD276_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)

## use euler 
p <- plot(euler(a, shape = "circle"), quantities = list(type = c("counts", "percent")),fill = list(fill = c("#FFA500", "#63B8FF","#9E59AD"))) 
g <- grid.arrange(grobs=list(p),top=textGrob("Cohort C", gp = gpar(fontface = "bold")))
ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_bothCP_differentSize.pdf")),g)


### for each individual
plot.list <- list()
for (samp in unique(sclc$donor_id)){
  epi_tmp <- subset(sclc,subset=donor_id==samp)
  SEZ6_pos <- WhichCells(epi_tmp, expression=SEZ6_group=="pos")
  B7H3_pos <- WhichCells(epi_tmp, expression=CD276_group=="pos")
  DLL3_pos <- WhichCells(epi_tmp, expression=DLL3_group=="pos") 
  
  a <- list(`DLL3+`=DLL3_pos,`SEZ6+`=SEZ6_pos,`B7H3+`=B7H3_pos)
    
  p <- plot(euler(a, shape = "circle"), quantities = list(type = c("counts", "percent"), cex=0.5),labels = list(cex = 0.6),fill = list(fill = c("#FFA500", "#63B8FF","#9E59AD"))) 
  g <- grid.arrange(grobs=list(p),top=textGrob(samp, gp = gpar(fontsize = 8,fontface = "bold")))

  plot.list[[samp]] <- g
  ggsave((paste0("../../plots/",Sys.Date(),"_venn_SEZ6_DLL3_B7H3_euler_",samp,".pdf")),g)
  
}

## plot all in one 
ggarrange(plotlist = plot.list, ncol = 5, nrow = 4)

ggsave(here(paste0("../../plots/",Sys.Date(),"_three_target_by_individual_allIn1.pdf")))



```

## Step 20: GSEA using known NE signatures


```{r}

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

epi_sub <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))


data1 <- FetchData(epi_sub, vars = c("DLL3","SEZ6","CD276"))


##set default as triple negative
epi_sub$group_gsea <- "triple_negative"

### for DLL3+ only

dll3.posOnly <- WhichCells(epi_sub, expression=DLL3_group=="pos" & SEZ6_group=="neg" & CD276_group=="neg")
epi_sub$group_gsea[dll3.posOnly] <- "DLL3+Only"

### for SEZ6+ only

sez6.posOnly <- WhichCells(epi_sub, expression=DLL3_group=="neg" & SEZ6_group=="pos" & CD276_group=="neg")
epi_sub$group_gsea[sez6.posOnly] <- "SEZ6+Only"

### for CD276+ only
cd276.posOnly <- WhichCells(epi_sub, expression=DLL3_group=="neg" & SEZ6_group=="neg" & CD276_group=="pos")
epi_sub$group_gsea[cd276.posOnly] <- "B7H3+Only"

## for coexpression of two or three

epi_sub$group_gsea[which(rowSums(data1>0)>1)] <-"2or3+"

## get NE signatures

### NE Alshalafa

NE_Alshalalfa <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Alshalalfa",colNames = FALSE)

NE_Alshalalfa_All <- NE_Alshalalfa$X1

NE_Alshalalfa_UP <- NE_Alshalalfa$X1[which(NE_Alshalalfa$X2=="UP")]

NE_Alshalalfa_DOWN <- NE_Alshalalfa$X1[which(NE_Alshalalfa$X2=="DOWN")]

NE_Alshalalfa <- list(Alshalalfa_All=NE_Alshalalfa_All,Alshalalfa_UP=NE_Alshalalfa_UP,Alshalalfa_DOWN=NE_Alshalalfa_DOWN)


## NE Balanis

NE_Balanis <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Balanis",colNames = FALSE)[,1]


## NE Beltran

NE_Beltran <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Beltran",colNames = FALSE)

NE_Beltran_All <- NE_Beltran$X1

NE_Beltran_UP <- NE_Beltran$X1[which(NE_Beltran$X2=="UP")]

NE_Beltran_DOWN <- NE_Beltran$X1[which(NE_Beltran$X2=="DOWN")]



## NE Kumar

NE_Kumar <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Kumar",colNames = FALSE)[,1]


## NE Tsai

NE_Tsai69 <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Tsai69",colNames = FALSE)

NE_Tsai306 <- read.xlsx("/Users/me73/Partners HealthCare Dropbox/Maoxuan Lin/projects/shihbo/fromShihBo/NEPC_genesigs.xlsx",sheet="Tsai306",colNames = FALSE)

NE_Tsai <- rbind(NE_Tsai69,NE_Tsai306)

NE_Tsai_All <- NE_Tsai$X1

NE_Tsai_UP <- NE_Tsai$X1[which(NE_Tsai$X2=="UP")]

NE_Tsai_DOWN <- NE_Tsai$X1[which(NE_Tsai$X2=="DOWN")]


### put all list together

NE <- list(Alshalalfa_All=NE_Alshalalfa_All,Alshalalfa_UP=NE_Alshalalfa_UP,Alshalalfa_DOWN=NE_Alshalalfa_DOWN,
           Balanis=NE_Balanis,
           Beltran_All=NE_Beltran_All,Beltran_UP=NE_Beltran_UP,Beltran_DOWN=NE_Beltran_DOWN,
           Kumar=NE_Kumar,
           Tsai_All=NE_Tsai_All,Tsai_UP=NE_Tsai_UP,Tsai_DOWN=NE_Tsai_DOWN)
### save NE signatures with the object name of NE
save(NE, file="/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/NE.pathway.list.RData")

epi_sub <- AddModuleScore_UCell(epi_sub, features = NE, ncores=4, name = "")

epi_sub$group_gsea <- factor(epi_sub$group_gsea,levels=c("DLL3+Only","SEZ6+Only","B7H3+Only","2or3+","triple_negative"))

VlnPlot(epi_sub, features=names(NE), group.by="group_gsea",pt.size = 0, ncol=4) &
  theme(
    axis.text.x = element_text(size = 6), 
    axis.text.y = element_text(size = 6), 
    axis.title.x = element_blank(),
    plot.title = element_text(size = 8, face = "bold", hjust = 0.5) 
  )

ggsave((paste0("../../plots/",Sys.Date(),"_NE_signature_GSEA1.pdf")))


## GSEA II

data1 <- FetchData(epi_sub, vars = c("DLL3","EPCAM"))

epi_sub$epcam_gsea <- "otherTumorCell"

epi_sub$epcam_gsea[which(data1$DLL3 >0 & data1$EPCAM==0)] <- "DLL3+Only"

epi_sub$epcam_gsea[which(data1$DLL3 ==0 & data1$EPCAM>0)] <- "EPCAM+Only"

epi_sub$epcam_gsea[which(data1$DLL3 >0 & data1$EPCAM>0)] <- "EPCAM&DLL3+"


epi_sub$epcam_gsea <- factor(epi_sub$epcam_gsea,levels=c("EPCAM+Only","DLL3+Only","EPCAM&DLL3+","otherTumorCell"))

VlnPlot(epi_sub, features=names(NE), group.by="epcam_gsea",pt.size = 0, ncol=4) &
  theme(
    axis.text.x = element_text(size = 6), 
    axis.text.y = element_text(size = 6), 
    axis.title.x = element_blank(),
    plot.title = element_text(size = 8, face = "bold", hjust = 0.5) 
  )

ggsave((paste0("../../plots/",Sys.Date(),"_NE_signature_GSEA2_EPCAM.pdf")))



```


## Step 21: revision

```{r}
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))


dir.create("../../manuscript/revision")

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(sclc, assays = "RNA", return.seurat=T,group.by = c("donor_id"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("donor_id","DLL3","ASCL1","NEUROD1","POU2F3"))

## add subtype info

dst <- as.data.frame(unclass(table(sclc$donor_id,sclc$SCLC_major_subtype_of_sample)))

dst$subtype <- apply(dst, 1, function(x) {
  cols <- colnames(dst)
  nz <- cols[x != 0]
  if (length(nz) == 0) NA else nz[1]   # take the first non-zero column name
})

dst$donor_id <- rownames(dst)

dst <- dst[,c("donor_id","subtype")]

## merge

df <- merge(df,dst)

### compare DLL3

p1 <- ggboxplot(df, x = "subtype", y = "DLL3",
          color = "subtype", palette = c("royalblue2", "#00B0F6","skyblue"),
          add = "jitter") + labs(x="",y="pseudobulk log-normalized DLL3") + theme(legend.position = "none") + theme(text = element_text(size = 7)) +
          stat_compare_means(label.x=2, size=3) 

#ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_boxplot_by_DLL3_pseudobulk_by_subtype_allCell.pdf")))


### ASCL1

p2 <- ggboxplot(df, x = "subtype", y = "ASCL1",
          color = "subtype", palette = c("royalblue2", "#00B0F6","skyblue"),
          add = "jitter") + labs(x="",y="pseudobulk log-normalized ASCL1") + theme(legend.position = "none") +theme(text = element_text(size = 7)) +
          stat_compare_means(label.x=2, size=3) 

#ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_boxplot_by_ASCL1_pseudobulk_by_subtype_allCell.pdf")))


### NEUROD1

p3 <- ggboxplot(df, x = "subtype", y = "NEUROD1",
          color = "subtype", palette = c("royalblue2", "#00B0F6","skyblue"),
          add = "jitter") + labs(x="",y="pseudobulk log-normalized NEUROD1") + theme(legend.position = "none") +theme(text = element_text(size = 7)) +
          stat_compare_means(label.x=2,size=3) 

#ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_boxplot_by_NEUROD1_pseudobulk_by_subtype_allCell.pdf")))


### POU2F3

p4 <- ggboxplot(df, x = "subtype", y = "POU2F3",
          color = "subtype", palette = c("royalblue2", "#00B0F6","skyblue"),
          add = "jitter") + labs(x="",y="pseudobulk log-normalized POU2F3") + theme(legend.position = "none") +theme(text = element_text(size = 7)) +
          stat_compare_means(label.x=2,size=3) 

#ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_boxplot_by_POU2F3_pseudobulk_by_subtype_allCell.pdf")))

ggarrange(p1,p2,p3,p4, ncol = 2, nrow = 2)

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_boxplot_4in1_pseudobulk_by_subtype.pdf")))

#### get treatment information

dtreat <- as.data.frame(unique(sclc@meta.data[,c("donor_id","treatment2")]))


dfinal <- merge(df,dtreat)

## add subtype fraction at single cell level

dsubtype <- as.data.frame(unclass(table(sclc$donor_id,sclc$SCLC_subtype)/rowSums(table(sclc$donor_id,sclc$SCLC_subtype))))
dsubtype$donor_id <- rownames(dsubtype)
dfinal2 <- merge(dfinal,dsubtype)

colnames(dfinal2) <- gsub("-","_",colnames(dfinal2))


## DLL3 and SCLC_A
ggscatter(dfinal2, x = "SCLC_A", y = "DLL3", color = "subtype",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab("%SCLC-A") + ylab("DLL3 expression")+ stat_cor(method = "spearman")

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_pseudobulk_DLL3_Atype_cor_spearman_all_sample.pdf")))

## only samples with SCLC-A as main type

dfinalA <- dfinal2[which(dfinal2$subtype=="SCLC-A"),]

ggscatter(dfinalA, x = "SCLC_A", y = "DLL3", color = "subtype",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab("%SCLC-A") + ylab("DLL3 expression")+ stat_cor(method = "spearman")

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_pseudobulk_DLL3_Atype_cor_spearman_Atype_sample.pdf")))


## ASCL1 and SCLC_A
ggscatter(dfinal2, x = "SCLC_A", y = "ASCL1", color = "subtype",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab("%SCLC-A") + ylab("ASCL1 expression")+ stat_cor(method = "spearman")

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_pseudobulk_ASCL1_Atype_cor_spearman_all_sample.pdf")))


## only samples with SCLC-A as main type

dfinalA <- dfinal2[which(dfinal2$subtype=="SCLC-A"),]

ggscatter(dfinalA, x = "SCLC_A", y = "ASCL1", color = "subtype",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab("%SCLC-A") + ylab("ASCL1 expression")+ stat_cor(method = "spearman")

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_pseudobulk_ASCL1_Atype_cor_spearman_Atype_sample.pdf")))


## SCLC marker and DLL3 correlation
smarker <- c("ASCL1","NEUROD1","POU2F3")
for(gene in smarker){
  ggscatter(dfinal2, x = gene, y = "DLL3", color = "subtype",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab(paste0(gene, " expression")) + ylab("DLL3 expression")+ stat_cor(method = "spearman")

  ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_pseudobulk_",gene,"_DLL3_cor_spearman_all_sample.pdf")))
}

## SCLC marker and DLL3 with only Atype

dfinalA <- dfinal2[which(dfinal2$subtype=="SCLC-A"),]

for(gene in smarker){
  ggscatter(dfinalA, x = gene, y = "DLL3", color = "subtype",
                  add = "reg.line",
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, title="Spearman correlation (pseudobulk)"
  ) + xlab(paste0(gene, " expression")) + ylab("DLL3 expression")+ stat_cor(method = "spearman")

  ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_pseudobulk_",gene,"_DLL3_cor_spearman_Atype_sample.pdf")))
}



#### split into low, medium, and high

## all cells 

mlevels <- c("SCLC-A","SCLC-N","SCLC-P")
sclc$SCLC_subtype <- factor(sclc$SCLC_subtype,levels=mlevels)

VlnPlot2(
  sclc,
  features = "DLL3",
  group.by = "SCLC_subtype",
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_allCells.pdf")))

### DLL3 positive cells only

cells <- WhichCells(sclc, expression=DLL3_group=="pos")

VlnPlot2(
  sclc,
  features = "DLL3",
  group.by = "SCLC_subtype",
  cells=cells,
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_posCells.pdf")))


### split DLL3+ cells into low-medium-high categories

## get gene expression of DLL3+ cells

expr <- FetchData(sclc, vars = "DLL3")[,1]
expr_pos <- expr[expr > 0]
q25_pos <- quantile(expr_pos, 0.25)
q75_pos <- quantile(expr_pos, 0.75)

#> quantile(expr_pos)
#       0%       25%       50%       75%      100% 
#0.2361021 1.1145579 1.6177686 2.2033682 5.0686421 

sclc$DLL3_pos_subgroup <- "neg"

sclc$DLL3_pos_subgroup[expr>0] <- cut(expr[expr>0], breaks=c(min(expr_pos), q25_pos, q75_pos, max(expr_pos)),include.lowest=TRUE,labels=c("Low","Medium","High"))

sclc$DLL3_pos_subgroup <- factor(sclc$DLL3_pos_subgroup,
                        levels = c("1", "2", "3","neg"),
                        labels = c("Low", "Medium", "High","neg"))

### DLL3 positive cells only

## Low
cells <- WhichCells(sclc, expression=(DLL3_pos_subgroup=="Low"))

VlnPlot2(
  sclc,
  features = "DLL3",
  group.by = "SCLC_subtype",
  cells=cells,
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_posCellsLow.pdf")))


## Medium
cells <- WhichCells(sclc, expression=(DLL3_pos_subgroup=="Medium"))

VlnPlot2(
  sclc,
  features = "DLL3",
  group.by = "SCLC_subtype",
  cells=cells,
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_posCellsMed.pdf")))


## High

cells <- WhichCells(sclc, expression=(DLL3_pos_subgroup=="High"))

VlnPlot2(
  sclc,
  features = "DLL3",
  group.by = "SCLC_subtype",
  cells=cells,
  pt=F,
  width=0.6,
  cols=c("royalblue2", "#00B0F6","skyblue"),
  stat.method = "wilcox.test") 

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_violinplot_by_DLL3_by_SCLC_subtype_withBox_posCellsHigh.pdf")))


### DLL3&ASCL1 correlation

FeatureScatter(sclc,feature1="ASCL1",feature2="DLL3",group.by="DLL3_pos_subgroup") + labs(color = "DLL3")

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3_and_ASCL1.pdf")))


### only positive

cells <- WhichCells(sclc, expression=(DLL3>0 & ASCL1>0))

FeatureScatter(sclc,feature1="ASCL1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3_and_ASCL1_bothPos.pdf")))

### only positive Low DLL3

cells <- WhichCells(sclc, expression=(DLL3>0 & ASCL1>0 & DLL3_pos_subgroup=="Low"))

FeatureScatter(sclc,feature1="ASCL1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3Low_and_ASCL1_bothPos.pdf")))


### only positive Medium DLL3

cells <- WhichCells(sclc, expression=(DLL3>0 & ASCL1>0 & DLL3_pos_subgroup=="Medium"))

FeatureScatter(sclc,feature1="ASCL1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3Med_and_ASCL1_bothPos.pdf")))

### only positive High DLL3

cells <- WhichCells(sclc, expression=(DLL3>0 & ASCL1>0 & DLL3_pos_subgroup=="High"))

FeatureScatter(sclc,feature1="ASCL1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3High_and_ASCL1_bothPos.pdf")))


### DLL3&NEUROD1 correlation

FeatureScatter(sclc,feature1="NEUROD1",feature2="DLL3",group.by="DLL3_pos_subgroup") + labs(color = "DLL3") 

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3_and_NEUROD1.pdf")))


### only positive

cells <- WhichCells(sclc, expression=(DLL3>0 & NEUROD1>0))

FeatureScatter(sclc,feature1="NEUROD1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3_and_NEUROD1_bothPos.pdf")))

### only positive Low DLL3

cells <- WhichCells(sclc, expression=(DLL3>0 & NEUROD1>0 & DLL3_pos_subgroup=="Low"))

FeatureScatter(sclc,feature1="NEUROD1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3Low_and_NEUROD1_bothPos.pdf")))


### only positive Medium DLL3

cells <- WhichCells(sclc, expression=(DLL3>0 & NEUROD1>0 & DLL3_pos_subgroup=="Medium"))

FeatureScatter(sclc,feature1="NEUROD1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3Med_and_NEUROD1_bothPos.pdf")))

### only positive High DLL3

cells <- WhichCells(sclc, expression=(DLL3>0 & NEUROD1>0 & DLL3_pos_subgroup=="High"))

FeatureScatter(sclc,feature1="NEUROD1",feature2="DLL3",cells=cells,group.by="DLL3_pos_subgroup") + labs(color = "DLL3")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_scatterplot_DLL3High_and_NEUROD1_bothPos.pdf")))


### get correlation for different categories


df_st <- data.frame(unclass(table(sclc$donor_id, sclc$SCLC_subtype)/rowSums(table(sclc$donor_id, sclc$SCLC_subtype))))
## order 
colnames(df_st) <- gsub("\\.","_",colnames(df_st))

df_st$donor_id <- rownames(df_st)


## get fractions of DLL3 positive cell per donor_id by three levels: low-med-high

df_dl <- data.frame(unclass(table(sclc$donor_id, sclc$DLL3_pos_subgroup)/rowSums(table(sclc$donor_id, sclc$DLL3_pos_subgroup))))

df_dl$donor_id <- rownames(df_dl)



## join

df <- merge(df_st,df_dl)

df$MedandHigh <- df$Medium+df$High

# List of x variables
x_vars <- c("SCLC_A", "SCLC_N", "SCLC_P")
y_vars <- c("Low","Medium","High","MedandHigh")

# Create an empty list to store manuscript
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  for (y in y_vars){
    sp <- ggscatter(df, x = x, y = y,
                    add = "reg.line",
                    add.params = list(color = "blue", fill = "lightgray"),
                    conf.int = TRUE
    ) + xlab(paste0("%", x)) + ylab(paste0("%DLL3+(",y,")")) + theme(text = element_text(size = 7))
    
    # Add correlation
    p <- sp + stat_cor(method = "pearson",label.x.npc=0.01, size=3)
    
    # Store plot in list with name
    plot_list[[paste0(x,"-",y)]] <- p
  }
}
 
ggarrange(plotlist = plot_list, ncol = 4, nrow = 3)

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_pearson_corr_DLL3_by_subtype.pdf")))

### use spearman 

# Create an empty list to store manuscript
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  for (y in y_vars){
    sp <- ggscatter(df, x = x, y = y,
                    add = "reg.line",
                    add.params = list(color = "blue", fill = "lightgray"),
                    conf.int = TRUE
    ) + xlab(paste0("%", x)) + ylab(paste0("%DLL3+(",y,")"))+ theme(text = element_text(size = 7))
    
    # Add correlation
    p <- sp + stat_cor(method = "spearman",label.x.npc=0.01,size=3)
    
    # Store plot in list with name
    plot_list[[paste0(x,"-",y)]] <- p
  }
}
 
ggarrange(plotlist = plot_list, ncol = 4, nrow = 3)

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_spearman_corr_DLL3_by_subtype.pdf")))


## get treatment information

dtreat <- as.data.frame(unique(sclc@meta.data[,c("donor_id","treatment2","SCLC_major_subtype_of_sample")]))


dfinal <- merge(df,dtreat)

write.xlsx(dfinal,paste0("../../manuscript/revision/",Sys.Date(),"_DLL3_3tier_subtype_treatment_summary.xlsx"))

### only subtype_A

df2 <- dfinal[which(dfinal$SCLC_major_subtype_of_sample=="SCLC-A"),]


# List of x variables
x_vars <- c("SCLC_A", "SCLC_N", "SCLC_P")
y_vars <- c("Low","Medium","High","MedandHigh")

# Create an empty list to store manuscript
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  for (y in y_vars){
    sp <- ggscatter(df2, x = x, y = y,
                    add = "reg.line",
                    add.params = list(color = "blue", fill = "lightgray"),
                    conf.int = TRUE
    ) + xlab(paste0("%", x)) + ylab(paste0("%DLL3+(",y,")")) + theme(text = element_text(size = 7))
    
    # Add correlation
    p <- sp + stat_cor(method = "pearson",label.x.npc=0.01, size=3)
    
    # Store plot in list with name
    plot_list[[paste0(x,"-",y)]] <- p
  }
}
 
ggarrange(plotlist = plot_list, ncol = 4, nrow = 3)

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_pearson_corr_DLL3_by_subtype_onlyAsample.pdf")))

### use spearman 

# Create an empty list to store manuscript
plot_list <- list()

# Loop through each variable
for (x in x_vars) {
  for (y in y_vars){
    sp <- ggscatter(df2, x = x, y = y,
                    add = "reg.line",
                    add.params = list(color = "blue", fill = "lightgray"),
                    conf.int = TRUE
    ) + xlab(paste0("%", x)) + ylab(paste0("%DLL3+(",y,")"))+ theme(text = element_text(size = 7))
    
    # Add correlation
    p <- sp + stat_cor(method = "spearman",label.x.npc=0.01,size=3)
    
    # Store plot in list with name
    plot_list[[paste0(x,"-",y)]] <- p
  }
}
 
ggarrange(plotlist = plot_list, ncol = 4, nrow = 3)

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_spearman_corr_DLL3_by_subtype_onlyAsample.pdf")))




### compare naive vs treated for each tier

# List of x variables
y_vars <- c("Low","Medium","High","MedandHigh")

# Create an empty list to store plots
plot_list <- list()

# Loop through each variable
for (y in y_vars){
  p<- ggboxplot(dfinal, x = "treatment2", y = y,
          color = "treatment2", palette = c("blue","red"),
          add = "jitter") + labs(x="",y=paste0("%DLL3+(",y,")")) + theme(legend.position = "none") +
          stat_compare_means(label.x.npc="middle") 
  
  # Store plot in list with name
  plot_list[[y]] <- p
}
 
ggarrange(plotlist = plot_list, ncol = 2, nrow = 2)

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_boxplot_DLL3_3tier_by_treatment.pdf")))

saveRDS(sclc,paste0(inDir,"/sclc_epithelial_cells_high_capture_v2.0.rds"))



###

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

## subset only samples included in final analysis


###heterogeneity among the different keratin genes in the 10X RNA seq of primary SCLC tumors?

FeaturePlot(sclc,features=c("KRT8","KRT18","KRT19","EPCAM"))

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_UMAP_EPCAM_CK.pdf")))



UMAPPlot(sclc,group.by="donor_id")
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_UMAP_by_sampleID.pdf")))

## EPCAM/CK pos vs neg

expr_matrix <- FetchData(sclc, vars = c("KRT8","KRT18","KRT19","EPCAM"))
sclc$EPCAMCK_positive <- rowSums(expr_matrix > 0) > 0

UMAPPlot(sclc, group.by="EPCAMCK_positive",cols=c("blue","red"))
ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_UMAP_any_EPCAM_CK_positive.pdf")))

### 

inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

epi <- readRDS(paste0(inDir,"/epithelial_cells_with_gene_as_rowname.rds"))

samples <- unique(sclc$donor_id)
epi_sub <- subset(epi, subset = donor_id %in% samples)
## plot for DLL3, EPCAM, CKs
celltype_colors <- c(
  "#1f77b4", # blue
  "#ff7f0e", # orange
  "#2ca02c", # green
  "#d62728", # red
  "#9467bd", # purple
  "#8c564b", # brown
  "#e377c2", # pink
  "#7f7f7f", # gray
  "#bcbd22", # olive
  "#17becf", # teal
  "#aec7e8", # light blue
  "#ffbb78", # light orange
  "#98df8a", # light green
  "#ff9896"  # light red
)

UMAPPlot(epi_sub,group.by="author_cell_type",cols=celltype_colors)

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_UMAP_all_cell_types.pdf")), width = 6, height = 6)
       
for (ck in c("DLL3", "EPCAM","KRT8","KRT18","KRT19")){
  FeaturePlot(epi_sub,features=ck)
  ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_UMAP_",ck,"_in_all_cell_types.pdf")), width = 6, height = 6)
  
  VlnPlot(epi_sub,features=ck,group.by="author_cell_type",cols=celltype_colors)
  ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_violin_",ck,"_in_all_cell_types.pdf")), width = 6, height = 6)
}


### plot EPCAM, DLL3, CK in one plot

## assign EPCAM pos/neg group
sclc$EPCAM_group <- "neg"
 
sclc$EPCAM_group[which(sclc[['RNA']]$data["EPCAM",] > 0)] <- "pos" 
sclc$EPCAM_group <- factor(sclc$EPCAM_group)


## define CK +/-
sclc$CK_group <- "neg"
expr_matrix <- FetchData(sclc, vars = c("KRT8","KRT18","KRT19"))
sclc$CK_group[which(rowSums(expr_matrix > 0) > 0)] <- "pos"
sclc$CK_group <- factor(sclc$CK_group)

## first plot dots with integrated color
sclc$DLL3_ <- ifelse(sclc$DLL3_group=="neg",0,1)
sclc$EPCAM_ <- ifelse(sclc$EPCAM_group=="neg",0,1)
sclc$CK_ <- ifelse(sclc$CK_group=="neg",0,1)

df <- as.data.frame(FetchData(sclc,c("UMAP_1","UMAP_2","DLL3_","EPCAM_","CK_","donor_id")))

df_grouped <- df %>%
  group_by(DLL3_, EPCAM_, CK_) %>%
  mutate(group = cur_group_id()) %>%
  ungroup() %>% as.data.frame()

## unique(df_grouped[,c(3,4,5,7)])
### old in MGH cohort
#     DLL3_ EPCAM_ CK_ group
# 1       0     0     0     1
# 3       1     0     0     5
# 6       0     0     1     2
# 8       0     1     1     4
# 19      1     0     1     6
# 23      0     1     0     3
# 47      1     1     0     7
# 100     1     1     1     8

## in CC cohort
#     DLL3_ EPCAM_ CK_ group
# 1       1      0   0     5
# 2       1      1   1     8
# 3       1      1   0     7
# 9       0      1   0     3
# 14      0      0   0     1
# 41      0      1   1     4
# 80      0      0   1     2
# 210     1      0   1     6

## colors


#rgb(1,0.647,0) "#FFA500"  ## DLL3
#rgb(1,1,0)  "#FFFF00" ## CK
#rgb(0.39,0.72,1) "#63B8FF" ## EPCAM
#rgb(1,1,1) "#FFFFFF"
#rgb(0,0,0) "#000000"
#rgb(0.5,0.5,0.5) "#808080"


### try different color V: 

r_d <- c(1,0.647,0) #"#FFA500"
r_c <- c(0.62,0.35,0.68) # "#9E59AD"
r_e <- c(0.39,0.72,1) #"#63B8FF"


ref_col <- data.frame(group=1:8,col_val=c(rgb(1,1,1), # group 1
                                         rgb(0.62,0.35,0.68), # 2
                                         rgb(0.39,0.72,1), # 3
                                         rgb(0.505, 0.535, 0.84), #4: colMeans(rbind(r_s,r_b))
                                         rgb(1,0.647,0), # 5
                                         rgb(0.81, 0.4985, 0.34), #6
                                         rgb(0.6950, 0.6835, 0.5), #7
                                         rgb(0.67, 0.57, 0.56) #8
                                         ))

df_col <- merge(df_grouped,ref_col)

data_col <- c("1"="#FFFFFF","2"="#9E59AD","3"="#63B8FF","4"="#8188D6","5"="#FFA500", "6"="#CF7F57","7"="#B1AE80","8"= "#AB918F")

max_x <- max(df_grouped$UMAP_1)
max_y <- max(df_grouped$UMAP_2)

legend_df <- data.frame(
  x = max_x,
  y = c(max_y, max_y-1.5,max_y-3),
  label = c('DLL3', 'EPCAM', 'CK'),
  color=c("#FFA500", "#63B8FF","#9E59AD") # The desired colors for the legend
)

p <- ggplot(df_col,aes(x=UMAP_1,y=UMAP_2)) + geom_point(aes(color=factor(group)),size=0.5,stroke=0.1) + 
  scale_colour_manual(values=data_col,guide="none") + theme_minimal() +
  theme(panel.border = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.line = element_line(color = "black"))  

 
p+
  geom_point(data = legend_df,
             aes(x = x-2, y = y, color = label ),
             size = 3) +
  geom_text(data = legend_df,
            aes(x = x -1, y = y+0.1, label = label),
            hjust = 0,size=3) +# get all elements but need to change the color for new legend at this layer
  scale_color_manual(name="",values = c(data_col,setNames(legend_df$color, legend_df$label)) ,guide="none") 

ggsave(here(paste0("../../manuscript/revision/",Sys.Date(),"_UMAP_EPCAM_DLL3_CK_binary_tumorOnly_withLabel_newColor_5.pdf")),width=11,height=11)

### draw Venn diagram for these three gene or sets as well

DLL3_pos <- WhichCells(sclc, expression=DLL3_group=="pos")
EPCAM_pos <- WhichCells(sclc, expression=EPCAM_group=="pos")
CK_pos <- WhichCells(sclc, expression=CK_group=="pos")

a <- list(`DLL3+`=DLL3_pos,`EPCAM+`=EPCAM_pos,`CK+`=CK_pos)

## use euler 
p <- plot(euler(a, shape = "ellipse"), quantities = list(type = c("counts", "percent")),fill = list(fill = c("#FFA500", "#63B8FF","#9E59AD"))) 
g <- grid.arrange(grobs=list(p),top=textGrob("Cohort C", gp = gpar(fontface = "bold")))
ggsave((paste0("../../manuscript/revision/",Sys.Date(),"_venn_EPCAM_DLL3_CK_euler_bothCP_differentSize.pdf")),g)

## fix the 99% issue instead 100% as th
## try different plot
library(VennDiagram)
venn.diagram(
x = list(`DLL3+`=DLL3_pos,`EPCAM+`=EPCAM_pos,`CK+`=CK_pos),
category.names = c("DLL3+", "EPCAM+", "CK+"),
col = "transparent",
fill = c("#FFA500", "#63B8FF","#9E59AD"),
alpha = 0.30,
print.mode=c("raw","percent"),
filename = paste0("../../manuscript/revision/",Sys.Date(),"_venn_EPCAM_DLL3_CK_euler_bothCP_sameSize.tiff"),
imagetype="tiff",
output=TRUE
)


### GSEA DLL3+ vs DLL3-

## load tumor aggressiveness pathway with name of gene_sets
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/tumor.aggressiveness.list.RData")


sclc <- AddModuleScore_UCell(sclc, features = gene_sets, ncores=4, name = "")

VlnPlot2(sclc, features=names(gene_sets), group.by = "DLL3_group", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("blue","red"), label="p.format",label.size=2.2)

ggsave((paste0("../../manuscript/revision/",Sys.Date(),"_SCLC_SMpathway_score_epi.pdf")))


### load NE signatures with the object name of NE
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/NE.pathway.list.RData")

sclc <- AddModuleScore_UCell(sclc, features = NE, ncores=4, name = "")

VlnPlot2(sclc, features=names(NE), group.by = "DLL3_group", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("blue","red"), label="p.format",label.size=1.8)

ggsave((paste0("../../manuscript/revision/",Sys.Date(),"_SCLC_NE_pathway_score_epi.pdf")))


### additional GSEA
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

## add DLL3 at the beginning of DLL3_pos

sclc$DLL3_group <- factor(paste0("DLL3_",sclc$DLL3_group))

##### assign EPCAM high/low group
sclc$EPCAM_group <- "EPCAM_low"
epcam_median <- median(sclc[['RNA']]$data["EPCAM",])
sclc$EPCAM_group[which(sclc[['RNA']]$data["EPCAM",] > epcam_median)] <- "EPCAM_high" 
sclc$EPCAM_group <- factor(sclc$EPCAM_group)

### assign combine group

sclc$DLL3_EPCAM <- paste0(sclc$DLL3_group,"_",sclc$EPCAM_group)

sclc$DLL3_EPCAM  <- factor(sclc$DLL3_EPCAM, levels=c("DLL3_pos_EPCAM_low","DLL3_pos_EPCAM_high","DLL3_neg_EPCAM_high","DLL3_neg_EPCAM_low"))


## load tumor aggressiveness pathway with name of gene_sets
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/tumor.aggressiveness.list.RData")

sclc <- AddModuleScore_UCell(sclc, features = gene_sets, ncores=4, name = "")

VlnPlot2(sclc, features=names(gene_sets), group.by = "DLL3_EPCAM", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2.2)+ theme(axis.text.x = element_blank())

ggsave((paste0("../../manuscript/revision/",Sys.Date(),"_SCLC_SMpathway_score_epi_by_DLL3_EPCAM.pdf")))
### NE

### load NE signatures with the object name of NE
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/NE.pathway.list.RData")

sclc <- AddModuleScore_UCell(sclc, features = NE, ncores=4, name = "")

VlnPlot2(sclc, features=names(NE), group.by = "DLL3_EPCAM", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2) + theme(axis.text.x = element_blank())

ggsave((paste0("../../manuscript/revision/",Sys.Date(),"_SCLC_NE_pathway_score_epi_by_DLL3_EPCAM.pdf")))


```


## step 22: revision-merge tumor and CTCs

```{r}
## cancer cell data
inDir <- "/Volumes/rama/labMembers/mao20/MGH2024/CTC/SCLC/data/public/chan_cc_2021"

sclc <- readRDS(paste0(inDir,"/sclc_epithelial_cells_high_capture_v1.0.rds"))

## add DLL3 at the beginning of DLL3_pos

sclc$DLL3_group <- factor(paste0("DLL3_",sclc$DLL3_group))

##### assign EPCAM high/low group
sclc$EPCAM_group <- "EPCAM_low"
epcam_median <- median(sclc[['RNA']]$data["EPCAM",])
sclc$EPCAM_group[which(sclc[['RNA']]$data["EPCAM",] > epcam_median)] <- "EPCAM_high" 
sclc$EPCAM_group <- factor(sclc$EPCAM_group)

### assign combine group

sclc$DLL3_EPCAM <- paste0(sclc$DLL3_group,"_",sclc$EPCAM_group)

sclc$DLL3_EPCAM  <- factor(sclc$DLL3_EPCAM, levels=c("DLL3_pos_EPCAM_low","DLL3_pos_EPCAM_high","DLL3_neg_EPCAM_high","DLL3_neg_EPCAM_low"))

## keep only necessary meta columns

sclc$sample <- sclc$donor_id
sclc$percent.mito <- 100*sclc$mito_frac
sclc$cohort <- "CancerCell"

new_meta <- sclc@meta.data[,c("sample","nFeature_RNA","nCount_RNA","percent.mito","DLL3_group","EPCAM_group","DLL3_EPCAM","cohort","SCLC_subtype")]

#sclc@meta.data <- new_meta

tmp <- sclc[["RNA"]]@counts

cc <- CreateSeuratObject(counts = tmp, project = "CancerCell", assay = "RNA")

cc@meta.data <- new_meta

#sclc$cohort <- "CancerCell"
## MGH cohort

epi_sub <- readRDS("/Volumes/rama/labMembers/mao20/MGH2024/CTC/moshe/rdata/scrna.unintegrated.seurat.final.rds")

epi_sub$DLL3_group <- factor(paste0("DLL3_",epi_sub$DLL3_group))

##### assign EPCAM high/low group
epi_sub$EPCAM_group <- "EPCAM_low"
epcam_median <- median(epi_sub[['RNA']]$data["EPCAM",])
epi_sub$EPCAM_group[which(epi_sub[['RNA']]$data["EPCAM",] > epcam_median)] <- "EPCAM_high" 
epi_sub$EPCAM_group <- factor(epi_sub$EPCAM_group)

### assign combine group

epi_sub$DLL3_EPCAM <- paste0(epi_sub$DLL3_group,"_",epi_sub$EPCAM_group)

epi_sub$DLL3_EPCAM  <- factor(epi_sub$DLL3_EPCAM, levels=c("DLL3_pos_EPCAM_low","DLL3_pos_EPCAM_high","DLL3_neg_EPCAM_high","DLL3_neg_EPCAM_low"))

epi_sub$cohort <- "MGH"


new_meta <- epi_sub@meta.data[,c("sample","nFeature_RNA","nCount_RNA","percent.mito","DLL3_group","EPCAM_group","DLL3_EPCAM","cohort","SCLC_subtype2")]

colnames(new_meta)[ncol(new_meta)] <- "SCLC_subtype"
#epi_sub@meta.data <- new_meta

tmp <- epi_sub@assays$RNA$counts

mgh <- CreateSeuratObject(counts = tmp, project = "MGH", assay = "RNA")

mgh@meta.data <- new_meta


## CTC 037

scrna.037<- readRDS(paste0("/Volumes/rama/labMembers/mao20/MGH2024/CTC/sh/results/MGHSCLC002031036037/Rdata/MGHSCLC037.scrna.seurat.with.anno.rds"))

scrna.037 <- subset(scrna.037,subset=final_anno=="epithelial")

scrna.037$DLL3_group <- "DLL3_neg"
 
scrna.037$DLL3_group[which(scrna.037[['RNA']]$data["DLL3",] > 0)] <- "DLL3_pos" 
scrna.037$DLL3_group <- factor(scrna.037$DLL3_group)

##### assign EPCAM high/low group
scrna.037$EPCAM_group <- "EPCAM_low"
epcam_median <- median(scrna.037[['RNA']]$data["EPCAM",])
scrna.037$EPCAM_group[which(scrna.037[['RNA']]$data["EPCAM",] > epcam_median)] <- "EPCAM_high" 
scrna.037$EPCAM_group <- factor(scrna.037$EPCAM_group)

### assign combine group

scrna.037$DLL3_EPCAM <- paste0(scrna.037$DLL3_group,"_",scrna.037$EPCAM_group)

scrna.037$DLL3_EPCAM  <- factor(scrna.037$DLL3_EPCAM, levels=c("DLL3_pos_EPCAM_low","DLL3_pos_EPCAM_high","DLL3_neg_EPCAM_high","DLL3_neg_EPCAM_low"))

scrna.037$cohort <- "CTC037"

### assign subtype at single cell level
# Define your gene sets (example with immune cell markers)
gene_sets <- list(
  SCLC_A = c("ASCL1", "SOX4", "STMN2", "DOC"),
  SCLC_N = c("NEUROD1", "ADCYAP1","NRXN1", "SSTR2", "ID1", "ID3", "SST", "DLK1"),
  SCLC_P = c("POU2F3", "ASCL2", "CD44", "MYC", "KIT", "YBX1"), 
  pInflamed= c("CD274","PDCD1","CD80","CD86","CTLA4","CD38","IDO1","TIGIT","ICOS","LAG3","CCL5","CXCL10",
              "HLA-A","HLA-B","HLA-C","HLA-E","HLA-F","HLA-G","HLA-DRA","HLA-DRB1","HLA-DQA1","HLA-DMA","HLA-DPA1",
              "HLA-DMB","HLA-DPB1","HLA-DQB1","HLA-DOA","HLA-DOB")
)


# Calculate scores for all gene sets
scrna.037 <- AddModuleScore_UCell(scrna.037, features = gene_sets, name = "")


score_matrix <- FetchData(scrna.037, vars = names(gene_sets))
colnames(score_matrix) <- c("SCLC-A","SCLC-N","SCLC-P","SCLC-I")
# Assign cell type based on highest score
scrna.037$SCLC_subtype <- colnames(score_matrix)[max.col(score_matrix)]


new_meta <- scrna.037@meta.data[,c("sample","nFeature_RNA","nCount_RNA","percent.mito","DLL3_group","EPCAM_group","DLL3_EPCAM","cohort","SCLC_subtype")]

#scrna.037@meta.data <- new_meta

tmp <- scrna.037@assays$RNA$counts

ctc <- CreateSeuratObject(counts = tmp, project = "CTC037", assay = "RNA")

ctc@meta.data <- new_meta

merged_seurat <- merge(x = cc, y = c(mgh,ctc), project = "SCLC")

### consume too much memory; let's use only count matrix

### and it works

dir.create("../../manuscript/revision/merge")

### check QC

# metadata variable

metadata_postfilter <- merged_seurat@meta.data


# Visualize the number of cell counts per sample
metadata_postfilter %>%
  	ggplot(aes(x=sample, fill=sample)) +
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("merged NCells ")

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_NCell.pdf"))


## mito simple
metadata_postfilter %>% 
  	ggplot(aes(color=sample, x=percent.mito, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 10) + 
    ggtitle("merged mito percent Per Cell")

ggsave(paste0("../../manuscript/revision/merge",Sys.Date(),"_merged_mito_simple.pdf"))

## get boxplot
metadata_postfilter %>% 
  	ggboxplot(x="sample", y="percent.mito",color="sample") +
  	theme_classic() +
  	ylab("mito percent") +
    ggtitle("merged mito percent Per Cell") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),legend.position = "none") 

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_mito_simple_boxplot.pdf"))


## UMI

metadata_postfilter %>% 
  	ggplot(aes(color=sample, x=nCount_RNA, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 1000) + 
    ggtitle("merged UMIs Per Cell")

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_UMI.pdf"))

## get boxplot
metadata_postfilter %>% 
  	ggboxplot(x="sample", y="nCount_RNA",color="sample") +
  	theme_classic() +
  	ylab("#UMI") +
    ggtitle("merged UMIs Per Cell") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),legend.position = "none") 

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_UMI_boxplot.pdf"))

## NGene

metadata_postfilter %>% 
  	ggplot(aes(color=sample, x=nFeature_RNA, fill=sample )) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 500) + 
    ggtitle("merged Genes Per Cell")

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_NGene.pdf"))

## get boxplot
metadata_postfilter %>% 
  	ggboxplot(x="sample", y="nFeature_RNA",color="sample") +
  	theme_classic() +
  	ylab("#Gene") +
    ggtitle("merged Genes Per Cell") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),legend.position = "none") 

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_NGene_boxplot.pdf"))


### normalize

merged_seurat <- NormalizeData(merged_seurat)
merged_seurat <- FindVariableFeatures(merged_seurat)
merged_seurat <- ScaleData(merged_seurat,vars.to.regress = c("percent.mito"))
merged_seurat <- RunPCA(merged_seurat, npcs = 30, verbose = F)

# Plot the PCA colored by cohort
DimPlot(merged_seurat,
        reduction = "pca",
        group.by= "cohort")

ggsave("../../manuscript/revision/merge/PCA_by_cohort.pdf") 

# Plot the PCA colored by SCLC_subtype
DimPlot(merged_seurat,
        reduction = "pca",
        group.by= "SCLC_subtype")

ggsave("../../manuscript/revision/merge/PCA_by_SCLC_subtype.pdf")

### by sample

DimPlot(merged_seurat,
        reduction = "pca",
        group.by= "sample")

ggsave("../../manuscript/revision/merge/PCA_by_sample.pdf")

## check DLL3 distribution across cohorts

merged_seurat$cohort <- factor(merged_seurat$cohort,levels=c("CancerCell","MGH","CTC037"))

VlnPlot2(
  merged_seurat,
  features = "DLL3",
  group.by = "cohort",
  pt=F,
  width=0.6,
  stat.method = "wilcox.test") +scale_fill_manual(values = c("blue","cyan","red") )


ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_violin_by_DLL3_by_cohort_with_box.pdf")))

## check EPCAM distribution across cohorts

VlnPlot2(
  merged_seurat,
  features = "EPCAM",
  group.by = "cohort",
  pt=F,
  width=0.6,
  stat.method = "wilcox.test") +scale_fill_manual(values = c("blue","cyan","red") )


ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_violin_by_EPCAM_by_cohort_with_box.pdf")))


### check UMAP

merged_seurat <- FindNeighbors(merged_seurat, reduction = "pca", dims = 1:30)
merged_seurat <- FindClusters(merged_seurat, resolution = 0.2, cluster.name = "clusters")

merged_seurat <- RunUMAP(merged_seurat, reduction = "pca",dims = 1:30, reduction.name = "umap")


DimPlot(
  merged_seurat,
  reduction = "umap",
  group.by = c("sample"),
  combine = FALSE, label.size = 2
)

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_UMAP_by_sample.pdf"))

DimPlot(
  merged_seurat,
  reduction = "umap",
  group.by = c("clusters"),
  combine = FALSE, label.size = 2
)

ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_UMAP_by_clusters.pdf")))

### 
DimPlot(
  merged_seurat,
  reduction = "umap",
  group.by = "cohort",
  combine = FALSE, label.size = 2
)

ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_UMAP_by_cohort.pdf")))


DimPlot(
  merged_seurat,
  reduction = "umap",
  group.by = "SCLC_subtype",
  combine = FALSE, label.size = 2
)

ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_UMAP_by_SCLC_subtype.pdf")))


### get pseudobulk expression

## pseudobulk the counts based on sample
pseudo_epi <- AggregateExpression(merged_seurat, assays = "RNA", return.seurat=T,group.by = c("sample"))

### fetch target genes in data slot (default)

df <- FetchData(pseudo_epi, vars = c("sample","DLL3","EPCAM"))

dcohort <- data.frame(unique(merged_seurat@meta.data[,c("cohort","sample")]))

df <- merge(df,dcohort)

df$cohort <- factor(df$cohort,levels=c("CancerCell","MGH","CTC037"))

### compare DLL3

ggboxplot(df, x = "cohort", y = "DLL3",
          color = "cohort", palette = c("blue","cyan","red"),
          add = "jitter") + labs(x="",y="pseudobulk DLL3") + theme(legend.position = "right") +
          stat_compare_means() 

ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_boxplot_by_DLL3_pseudobulk_by_sample_cohort.pdf")))


### compare EPCAM

ggboxplot(df, x = "cohort", y = "EPCAM",
          color = "cohort", palette = c("blue","cyan","red"),
          add = "jitter") + labs(x="",y="pseudobulk EPCAM") + theme(legend.position = "right") +
          stat_compare_means() 

ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_boxplot_by_EPCAM_pseudobulk_by_sample_cohort.pdf")))

saveRDS(merged_seurat,"../../manuscript/revision/merge/merged_cc_mgh_037.seurat.without.filtering.rds")


## Harmony (method=HarmonyIntegration)

scrna.integrated <- IntegrateLayers(
  object = merged_seurat, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "integrated.harmony", id.variable = "cohort",
  verbose = FALSE
)

scrna.integrated <- FindNeighbors(scrna.integrated, reduction = "integrated.harmony", dims = 1:30)
scrna.integrated <- FindClusters(scrna.integrated, resolution = 0.2, cluster.name = "harmony_clusters")

scrna.integrated <- RunUMAP(scrna.integrated, reduction = "integrated.harmony",dims = 1:30, reduction.name = "umap.harmony")


DimPlot(
  scrna.integrated,
  reduction = "umap.harmony",
  group.by = c("sample"),
  combine = FALSE, label.size = 2
)

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_integration_harmony_by_sample.pdf"))

DimPlot(
  scrna.integrated,
  reduction = "umap.harmony",
  group.by = c("harmony_clusters"),
  combine = FALSE, label.size = 2
)

ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_integration_harmony_by_clusters.pdf")))

DimPlot(
  scrna.integrated,
  reduction = "umap.harmony",
  group.by = c("cohort"),
  combine = FALSE, label.size = 2
)

ggsave(paste0("../../manuscript/revision/merge/",Sys.Date(),"_integration_harmony_by_cohort.pdf"))

DimPlot(
  scrna.integrated,
  reduction = "umap.harmony",
  group.by = c("SCLC_subtype"),
  combine = FALSE, label.size = 2
)

ggsave(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_integration_harmony_by_sclc_subtype.pdf")))

##Quantify variance explained by batch vs patient

pca_mat <- Embeddings(merged_seurat, "pca")[, 1:20]
cohort <- factor(merged_seurat$cohort)
sample <- factor(merged_seurat$sample)

r2_batch <- sapply(1:20, function(i) summary(lm(pca_mat[,i] ~ cohort))$r.squared)
r2_patient <- sapply(1:20, function(i) summary(lm(pca_mat[,i] ~ sample))$r.squared)

pdf(here(paste0("../../manuscript/revision/merge/",Sys.Date(),"_variance_explanation_by_cohort_vs_patient.pdf")))
barplot(
  rbind(r2_batch, r2_patient),
  beside = TRUE,
  col = c("tomato", "skyblue"),
  legend.text = c("Batch", "Patient"),
  names.arg = paste0("PC", 1:20),   #  adds PC1PC20 labels
  las = 2,                          # rotate labels for readability
  cex.names = 0.8,                  # adjust label size
  main = "Variance explained by batch vs patient",
  ylab = "R"
)
dev.off()

### re-group EPCAM and DLL3_EPCAM
#STOP HERE: need to move to ERIS due to memory limitation
merged_seurat <- readRDS("../../manuscript/revision/merge/merged_cc_mgh_037.seurat.without.filtering.rds")

##### assign EPCAM high/low group
merged_seurat <- JoinLayers(merged_seurat)

saveRDS(merged_seurat,"../../manuscript/revision/merge/merged_cc_mgh_037.seurat.without.filtering.joinLayers.rds")

merged_seurat$EPCAM_group_m <- "EPCAM_low"
epcam_median <- median(merged_seurat[['RNA']]$data["EPCAM",])
merged_seurat$EPCAM_group_m[which(merged_seurat[['RNA']]$data["EPCAM",] > epcam_median)] <- "EPCAM_high" 
merged_seurat$EPCAM_group_m <- factor(merged_seurat$EPCAM_group_m)

### assign combine group

merged_seurat$DLL3_EPCAM_m <- paste0(merged_seurat$DLL3_group,"_",merged_seurat$EPCAM_group_m)

merged_seurat$DLL3_EPCAM_m  <- factor(merged_seurat$DLL3_EPCAM_m, levels=c("DLL3_pos_EPCAM_low","DLL3_pos_EPCAM_high","DLL3_neg_EPCAM_high","DLL3_neg_EPCAM_low"))


## load tumor aggressiveness pathway with name of gene_sets
load("/data/rama/labMembers/mao20/MGH2024/CTC/pathways/tumor.aggressiveness.list.RData")

merged_seurat <- AddModuleScore_UCell(merged_seurat, features = gene_sets, ncores=4, name = "")

VlnPlot2(merged_seurat, features=names(gene_sets), group.by = "DLL3_EPCAM_m", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2.2)+ theme(axis.text.x = element_blank())
ggsave((paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_seurat_SMpathway_score_epi_by_DLL3_EPCAM_megred.pdf")))
### check cohort

VlnPlot2(merged_seurat, features=names(gene_sets), group.by = "cohort", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2.2)+ theme(axis.text.x = element_blank())

ggsave((paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_seurat_SMpathway_score_epi_by_cohort_megred.pdf")))



### NE

### load NE signatures with the object name of NE
load("/data/rama/labMembers/mao20/MGH2024/CTC/pathways/NE.pathway.list.RData")

merged_seurat <- AddModuleScore_UCell(merged_seurat, features = NE, ncores=4, name = "")

VlnPlot2(merged_seurat, features=names(NE), group.by = "DLL3_EPCAM_m", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2) + theme(axis.text.x = element_blank())

ggsave((paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_seurat_NE_pathway_score_epi_by_DLL3_EPCAM.pdf")))

## by cohort

VlnPlot2(merged_seurat, features=names(NE), group.by = "cohort", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2) + theme(axis.text.x = element_blank())

ggsave((paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_seurat_NE_pathway_score_epi_by_cohort.pdf")))


### get pseudobulk

pseudo_sclc <- AggregateExpression(merged_seurat, assays = "RNA", return.seurat = T, group.by = c("sample", "DLL3_EPCAM_m"))

# each 'cell' is a donor-celltype-dll3level pseudobulk profile
tail(Cells(pseudo_sclc))

Idents(pseudo_sclc) <- "DLL3_EPCAM_m"

bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-pos-EPCAM-low", 
                         ident.2 = "DLL3-pos-EPCAM-high",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)

bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-pos-EPCAM-low", 
                         ident.2 = "DLL3-neg-EPCAM-high",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)


bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-pos-EPCAM-low", 
                         ident.2 = "DLL3-neg-EPCAM-low",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)


bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-pos-EPCAM-high", 
                         ident.2 = "DLL3-neg-EPCAM-high",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)

bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-pos-EPCAM-high", 
                         ident.2 = "DLL3-neg-EPCAM-low",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)



bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-neg-EPCAM-high", 
                         ident.2 = "DLL3-neg-EPCAM-low",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)


### compare each subpopulation by cohort

## DLL3_pos_EPCAM_low
for(pop in unique(merged_seurat$DLL3_EPCAM_m)){

  cells <- rownames(merged_seurat@meta.data)[merged_seurat$DLL3_EPCAM_m == pop]
  tmp <- subset(merged_seurat,cells=cells)

  VlnPlot2(tmp, features=names(gene_sets), group.by = "cohort", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan"), label="p.format",label.size=2.2)+ theme(axis.text.x = element_blank()) + ggtitle(pop)

  ggsave((paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_seurat_SMpathway_score_epi_by_cohort_",pop,".pdf")))


  VlnPlot2(tmp, features=names(NE), group.by = "cohort", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan"), label="p.format",label.size=2) + theme(axis.text.x = element_blank())+ ggtitle(pop)

  ggsave((paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_seurat_NE_pathway_score_epi_by_cohort_",pop,".pdf")))

}
########################
### let apply low vs high to DLL3 too

merged_seurat <- readRDS("../../manuscript/revision/merge/merged_cc_mgh_037.seurat.without.filtering.joinLayers.rds")

## EPCAM high vs low

merged_seurat$EPCAM_group_m <- "EPCAM_low"
epcam_median <- median(merged_seurat[['RNA']]$data["EPCAM",])
merged_seurat$EPCAM_group_m[which(merged_seurat[['RNA']]$data["EPCAM",] > epcam_median)] <- "EPCAM_high" 
merged_seurat$EPCAM_group_m <- factor(merged_seurat$EPCAM_group_m)

## DLL3 high vs low

merged_seurat$DLL3_group_m <- "DLL3_low"
DLL3_median <- median(merged_seurat[['RNA']]$data["DLL3",])
merged_seurat$DLL3_group_m[which(merged_seurat[['RNA']]$data["DLL3",] > DLL3_median)] <- "DLL3_high" 
merged_seurat$DLL3_group_m <- factor(merged_seurat$DLL3_group_m)


### assign combine group

merged_seurat$DLL3_EPCAM_m <- paste0(merged_seurat$DLL3_group_m,"_",merged_seurat$EPCAM_group_m)

merged_seurat$DLL3_EPCAM_m  <- factor(merged_seurat$DLL3_EPCAM_m, levels=c("DLL3_high_EPCAM_low","DLL3_high_EPCAM_high","DLL3_low_EPCAM_high","DLL3_low_EPCAM_low"))

#table(merged_seurat$DLL3_EPCAM_m)
# DLL3_high_EPCAM_low DLL3_high_EPCAM_high  DLL3_low_EPCAM_high   DLL3_low_EPCAM_low 
#               14560                20560                14560                20561 

#merged_seurat$DLL3_EPCAM <- paste0(merged_seurat$DLL3_group,"_",merged_seurat$EPCAM_group_m)
#table(merged_seurat$DLL3_EPCAM)

#DLL3_neg_EPCAM_high  DLL3_neg_EPCAM_low DLL3_pos_EPCAM_high  DLL3_pos_EPCAM_low 
#              12872               18373               22248               16748 

## load tumor aggressiveness pathway with name of gene_sets
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/tumor.aggressiveness.list.RData")

merged_seurat <- AddModuleScore_UCell(merged_seurat, features = gene_sets, ncores=4, name = "")

VlnPlot2(merged_seurat, features=names(gene_sets), group.by = "DLL3_EPCAM_m", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2.2)+ theme(axis.text.x = element_blank())
ggsave((paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_seurat_SMpathway_score_epi_by_DLL3_EPCAM_megred_highLow.pdf")))


### NE

### load NE signatures with the object name of NE
load("/Volumes/rama/labMembers/mao20/MGH2024/CTC/pathways/NE.pathway.list.RData")

merged_seurat <- AddModuleScore_UCell(merged_seurat, features = NE, ncores=4, name = "")

VlnPlot2(merged_seurat, features=names(NE), group.by = "DLL3_EPCAM_m", pt=F,show.mean=TRUE,
  width=0.5, stat.method = "wilcox.test", ncol=2,cols=c("red","orange","cyan","blue"), label="p.format",label.size=2) + theme(axis.text.x = element_blank())

ggsave((paste0("../../manuscript/revision/merge/",Sys.Date(),"_merged_seurat_NE_pathway_score_epi_by_DLL3_EPCAM_highLow.pdf")))


### get pseudobulk

pseudo_sclc <- AggregateExpression(merged_seurat, assays = "RNA", return.seurat = T, group.by = c("sample", "DLL3_EPCAM_m"))

# each 'cell' is a donor-celltype-dll3level pseudobulk profile
tail(Cells(pseudo_sclc))

Idents(pseudo_sclc) <- "DLL3_EPCAM_m"

bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-high-EPCAM-low", 
                         ident.2 = "DLL3-high-EPCAM-high",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)

bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-high-EPCAM-low", 
                         ident.2 = "DLL3-low-EPCAM-high",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)


bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-high-EPCAM-low", 
                         ident.2 = "DLL3-low-EPCAM-low",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)


bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-high-EPCAM-high", 
                         ident.2 = "DLL3-low-EPCAM-high",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)

bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-high-EPCAM-high", 
                         ident.2 = "DLL3-low-EPCAM-low",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)



bulk.sclc.de <- FindMarkers(object = pseudo_sclc, 
                         ident.1 = "DLL3-low-EPCAM-high", 
                         ident.2 = "DLL3-low-EPCAM-low",
                         test.use = "DESeq2")
head(bulk.sclc.de, n = 15)




```









